// phantasma-sdk-ts/dist/cjs/core/ledger/Ledger-Commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetBalanceFromMnemonic = exports.GetBalanceFromPrivateKey = exports.SendTransactionLedger = exports.GetAddressFromLedeger = exports.GetBalanceFromLedger = exports.GetLedgerSignerData = exports.GetLedgerAccountSigner = exports.GetLedgerDeviceInfo = exports.ToWholeNumber = exports.LeftPad = void 0;
const index_js_1 = require("../index.js");
const logger_js_1 = require("../utils/logger.js");
const index_js_2 = require("../tx/index.js");
const index_js_3 = require("../types/index.js");
const Address_Transcode_js_1 = require("./Address-Transcode.js");
const Transaction_Sign_js_1 = require("./Transaction-Sign.js");
const Transaction_Transcode_js_1 = require("./Transaction-Transcode.js");
const Ledger_Utils_js_1 = require("./Ledger-Utils.js");
/**
 *
 * @param number
 * @param length
 * @returns
 */
const LeftPad = (number, length) => {
    let str = '' + number;
    while (str.length < length) {
        str = '0' + str;
    }
    return str;
};
exports.LeftPad = LeftPad;
/**
 *
 * @param balance
 * @param decimals
 * @returns
 */
const ToWholeNumber = (balance, decimals) => {
    if (balance === undefined) {
        throw Error('balance is a required parameter.');
    }
    if (decimals === undefined) {
        throw Error('decimals is a required parameter.');
    }
    // console.log('toWholeNumber', 'balance', balance);
    const paddedBalance = (0, exports.LeftPad)(balance, decimals + 1);
    // console.log('toWholeNumber', 'paddedBalance', paddedBalance);
    const prefixLength = paddedBalance.length - decimals;
    // console.log('toWholeNumber', 'prefixLength', prefixLength);
    const prefix = paddedBalance.slice(0, prefixLength);
    // console.log('toWholeNumber', 'prefix', prefix);
    const suffix = paddedBalance.slice(-decimals);
    // console.log('toWholeNumber', 'suffix', suffix);
    return `${prefix}.${suffix}`;
};
exports.ToWholeNumber = ToWholeNumber;
/**
 * Get the device info from the ledger.
 * @param config
 * @returns
 */
const GetLedgerDeviceInfo = async (config) => {
    if (config == undefined) {
        throw Error('config is a required parameter.');
    }
    const version = await (0, Ledger_Utils_js_1.GetVersion)(config.Transport);
    const applicationName = await (0, Ledger_Utils_js_1.GetApplicationName)(config.Transport);
    return {
        version: version,
        applicationName: applicationName,
    };
};
exports.GetLedgerDeviceInfo = GetLedgerDeviceInfo;
/**
 * Get Ledger Account Signer
 * @param config
 * @param accountIx
 * @returns
 */
const GetLedgerAccountSigner = async (config, accountIx) => {
    /* istanbul ignore if */
    if (config === undefined) {
        throw Error('config is a required parameter.');
    }
    /* istanbul ignore if */
    if (accountIx === undefined) {
        throw Error('accountIx is a required parameter.');
    }
    const paths = await config.Transport.list();
    logger_js_1.logger.log('paths', paths);
    if (paths.length == 0) {
        alert('NUmber of devices found:' + paths.length);
        return;
    }
    const accountData = await GetLedgerSignerData(config, {
        verifyOnDevice: false,
        debug: true,
    });
    let signer = {
        GetPublicKey: () => {
            return accountData.publicKey;
        },
        GetAccount: () => {
            return accountData.address;
        },
    };
    return signer;
};
exports.GetLedgerAccountSigner = GetLedgerAccountSigner;
/**
 * GetLedgerSignerData
 * @param config
 * @param options
 * @returns
 */
async function GetLedgerSignerData(config, options) {
    if (config == undefined) {
        throw Error('config is a required parameter.');
    }
    if (options == undefined) {
        throw Error('options is a required parameter.');
    }
    const msg = await (0, Ledger_Utils_js_1.GetPublicKey)(config.Transport, options);
    let response = {
        address: index_js_3.Address.Null,
        publicKey: '',
        success: false,
        message: '',
    };
    response.success = false;
    response.message = msg.message;
    if (!msg.success) {
        return response;
    }
    const publicKey = msg.publicKey;
    const address = (0, Address_Transcode_js_1.GetAddressPublicKeyFromPublicKey)(publicKey);
    response.success = true;
    response.message = 'success';
    response.address = address;
    response.publicKey = publicKey;
    return response;
}
exports.GetLedgerSignerData = GetLedgerSignerData;
/**
 * GetBalanceFromLedger
 * @param config
 * @param options
 * @returns
 */
const GetBalanceFromLedger = async (config, options) => {
    /* istanbul ignore if */
    if (config == undefined) {
        throw Error('config is a required parameter.');
    }
    /* istanbul ignore if */
    if (options == undefined) {
        throw Error('options is a required parameter.');
    }
    const msg = await (0, Ledger_Utils_js_1.GetPublicKey)(config.Transport, options);
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('getBalanceFromLedger', 'msg', msg);
    }
    let response = {
        address: index_js_3.Address.Null,
        publicKey: '',
        balances: new Map(),
        success: false,
        message: '',
    };
    response.message = msg.message;
    if (!msg.success) {
        return response;
    }
    const publicKey = msg.publicKey;
    const address = (0, Address_Transcode_js_1.GetAddressPublicKeyFromPublicKey)(publicKey);
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('address', address);
        logger_js_1.logger.log('rpc', config.RPC);
    }
    logger_js_1.logger.log('rpcAwait', await config.RPC.getAccount(address.Text));
    const rpcResponse = await config.RPC.getAccount(address.Text);
    if (config.Debug) {
        logger_js_1.logger.log('rpcResponse', rpcResponse);
    }
    response.balances = new Map();
    if (rpcResponse.balances !== undefined) {
        rpcResponse.balances.forEach((balanceElt) => {
            response.balances[balanceElt.symbol] = (0, exports.ToWholeNumber)(balanceElt.amount, balanceElt.decimals);
        });
    }
    response.address = address;
    response.publicKey = publicKey;
    response.success = true;
    return response;
};
exports.GetBalanceFromLedger = GetBalanceFromLedger;
/**
 * Get Addres from Ledger
 * @param config
 * @param options
 * @returns
 */
const GetAddressFromLedeger = async (config, options) => {
    /* istanbul ignore if */
    if (config == undefined) {
        throw Error('config is a required parameter.');
    }
    /* istanbul ignore if */
    if (options == undefined) {
        throw Error('options is a required parameter.');
    }
    const msg = await (0, Ledger_Utils_js_1.GetPublicKey)(config.Transport, options);
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('getBalanceFromLedger', 'msg', msg);
    }
    if (msg.success) {
        const publicKey = msg.publicKey;
        const address = (0, Address_Transcode_js_1.GetAddressFromPublicKey)(publicKey);
        return address;
    }
    else {
        return msg;
    }
};
exports.GetAddressFromLedeger = GetAddressFromLedeger;
/**
 *
 * @param encodedTx
 * @param config
 * @returns
 */
async function SignEncodedTx(encodedTx, config) {
    const response = await (0, Ledger_Utils_js_1.SignLedger)(config.Transport, encodedTx);
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('sendAmountUsingLedger', 'signCallback', 'response', response);
    }
    if (response.success) {
        return response.signature;
    }
    else {
        throw Error(response.message);
    }
}
/**
 * SendTransactionLedger
 * @param config
 * @param script
 * @returns
 */
async function SendTransactionLedger(config, script) {
    if (config == undefined) {
        throw Error('config is a required parameter.');
    }
    const options = { verifyOnDevice: false };
    const msg_publicKey = await (0, Ledger_Utils_js_1.GetPublicKey)(config.Transport, options);
    if (!msg_publicKey.success) {
        if (config.Debug) {
            logger_js_1.logger.log('SendTransactionLedger', 'error ', msg_publicKey);
        }
        return msg_publicKey;
    }
    const addr = (0, Address_Transcode_js_1.GetAddressPublicKeyFromPublicKey)(msg_publicKey.publicKey);
    const publicKey = msg_publicKey.publicKey;
    const nexusName = config.NexusName;
    const chainName = config.ChainName;
    const gasPrice = config.GasPrice;
    const gasLimit = config.GasLimit;
    const expirationDate = (0, Transaction_Transcode_js_1.GetExpirationDate)();
    // no payload, could be a message.
    const payload = config.Payload;
    const myTransaction = new index_js_2.Transaction(nexusName, // Nexus Name
    chainName, // Chain
    script, // In string format
    expirationDate, // Expiration Date
    payload); // Extra Info to attach to Transaction in Serialized Hex
    const encodedTx = index_js_3.Base16.encodeUint8Array(myTransaction.ToByteAray(false));
    try {
        if (config.Debug) {
            logger_js_1.logger.log('sendAmountUsingCallback', 'encodedTx', encodedTx);
        }
        const signature = await SignEncodedTx(encodedTx, config);
        if (config.Debug) {
            logger_js_1.logger.log('sendAmountUsingCallback', 'signature', signature);
        }
        if (config.VerifyResponse) {
            const verifyResponse = (0, Transaction_Sign_js_1.Verify)(encodedTx, signature, publicKey);
            if (verifyResponse == false) {
                throw Error(`invalidSignature encodedTx:'${encodedTx}', publicKey:'${publicKey}' signature:'${signature}'`);
            }
            if (config.Debug) {
                logger_js_1.logger.log('verifyResponse', verifyResponse);
            }
        }
        let signatureBytes = index_js_3.Base16.decodeUint8Array(signature);
        let mySignature = new index_js_3.Ed25519Signature(signatureBytes);
        let myNewSignaturesArray = [];
        myNewSignaturesArray.push(mySignature);
        myTransaction.signatures = myNewSignaturesArray;
        if (config.Debug) {
            logger_js_1.logger.log('signedTx', myTransaction);
        }
        const encodedSignedTx = index_js_3.Base16.encodeUint8Array(myTransaction.ToByteAray(true));
        logger_js_1.logger.log('encoded signed tx: ', encodedSignedTx);
        const txHash = await config.RPC.sendRawTransaction(encodedSignedTx);
        if (config.Debug) {
            logger_js_1.logger.log('sendAmountUsingCallback', 'txHash', txHash);
        }
        const response = {
            success: true,
            message: txHash,
        };
        /* istanbul ignore if */
        if (config.Debug) {
            logger_js_1.logger.log('response', response);
        }
        return response;
    }
    catch (error) {
        if (config.Debug) {
            logger_js_1.logger.log('error', error);
        }
        const errorResponse = {
            success: false,
            message: error.message,
        };
        return errorResponse;
    }
}
exports.SendTransactionLedger = SendTransactionLedger;
/**
 *
 * @param config
 * @param privateKey
 * @returns
 */
const GetBalanceFromPrivateKey = async (config, privateKey) => {
    /* istanbul ignore if */
    if (config == undefined) {
        throw Error('config is a required parameter.');
    }
    /* istanbul ignore if */
    if (privateKey == undefined) {
        throw Error('privateKey is a required parameter.');
    }
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('privateKey', privateKey);
    }
    // https://github.com/phantasma-io/phantasma-ts/blob/7d04aaed839851ae5640f68ab223ca7d92c42016/core/tx/utils.js
    const publicKey = (0, Transaction_Sign_js_1.GetPublicFromPrivate)(privateKey);
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('publicKey', publicKey);
    }
    const address = (0, Address_Transcode_js_1.GetAddressFromPublicKey)(publicKey);
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('address', address);
    }
    // const path = `/address/${address}`;
    // const response = await httpRequestUtil.get(config, path);
    const rpcResponse = await config.RPC.getAccount(address);
    if (config.Debug) {
        logger_js_1.logger.log('rpcResponse', rpcResponse);
    }
    let response;
    response.balances = new Map();
    if (rpcResponse.balances !== undefined) {
        rpcResponse.balances.forEach((balanceElt) => {
            response.balances[balanceElt.symbol] = (0, exports.ToWholeNumber)(balanceElt.amount, balanceElt.decimals);
        });
    }
    response.address = index_js_3.Address.FromText(address);
    response.success = true;
    // const lastRefPath = `/transaction/last-ref/${address}`;
    // const lastRefResponse = await httpRequestUtil.get(config, lastRefPath);
    // response.lastRef = lastRefResponse;
    return response;
};
exports.GetBalanceFromPrivateKey = GetBalanceFromPrivateKey;
/**
 *
 * @param config
 * @param mnemonic
 * @param index
 * @returns
 */
const GetBalanceFromMnemonic = async (config, mnemonic, index) => {
    /* istanbul ignore if */
    if (config == undefined) {
        throw Error('config is a required parameter.');
    }
    /* istanbul ignore if */
    if (mnemonic == undefined) {
        throw Error('mnemonic is a required parameter.');
    }
    /* istanbul ignore if */
    if (index == undefined) {
        throw Error('index is a required parameter.');
    }
    /* istanbul ignore if */
    if (config.Debug) {
        logger_js_1.logger.log('mnemonic', mnemonic);
    }
    const privateKey = (0, index_js_1.GetPrivateKeyFromMnemonic)(config, mnemonic, index);
    return await (0, exports.GetBalanceFromPrivateKey)(config, privateKey);
};
exports.GetBalanceFromMnemonic = GetBalanceFromMnemonic;
