// phantasma-sdk-ts/dist/cjs/core/ledger/Transaction-Sign.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetPublicFromPrivate = exports.Verify = exports.Sign = exports.GetHash = exports.SignBytes = exports.PublicToPem = exports.PublicToDer = exports.PrivateToDer = void 0;
const crypto = __importStar(require("crypto"));
const logger_js_1 = require("../utils/logger.js");
const PUBLIC_KEY_PREFIX = '302A300506032B6570032100';
const DEBUG = false;
const PrivateToDer = (privateKeyHex) => {
    if (DEBUG) {
        logger_js_1.logger.log('privateToDer', 'privateKeyHex', privateKeyHex);
    }
    const derHex = `302e020100300506032b657004220420${privateKeyHex}`;
    if (DEBUG) {
        logger_js_1.logger.log('privateToDer', 'derHex', derHex);
    }
    return Buffer.from(derHex, 'hex');
};
exports.PrivateToDer = PrivateToDer;
const PublicToDer = (publicKeyHex) => {
    const publicKeyDerHex = `${PUBLIC_KEY_PREFIX}${publicKeyHex}`;
    return Buffer.from(publicKeyDerHex, 'hex');
};
exports.PublicToDer = PublicToDer;
const PublicToPem = (publicKeyHex) => {
    const publicKeyDer = (0, exports.PublicToDer)(publicKeyHex);
    const publicKeyDerBase64 = publicKeyDer.toString('base64');
    return `-----BEGIN PUBLIC KEY-----\n${publicKeyDerBase64}\n-----END PUBLIC KEY-----`;
};
exports.PublicToPem = PublicToPem;
const SignBytes = (hash, privateKey) => {
    if (DEBUG) {
        logger_js_1.logger.log('signBytes.hash', hash);
        logger_js_1.logger.log('signBytes.privateKey', privateKey);
    }
    const privateKeyDer = (0, exports.PrivateToDer)(privateKey.toString('hex'));
    if (DEBUG) {
        logger_js_1.logger.log('signBytes.privateKeyDer', privateKeyDer);
    }
    const privateKeyObj = crypto.createPrivateKey({
        key: privateKeyDer,
        format: 'der',
        type: 'pkcs8',
    });
    const signature = crypto.sign(undefined, hash, privateKeyObj);
    const signatureHex = signature.toString('hex');
    if (DEBUG) {
        logger_js_1.logger.log('signatureHex', signatureHex);
    }
    return signatureHex;
};
exports.SignBytes = SignBytes;
const GetHash = (encodedTx, debug) => {
    return Buffer.from(encodedTx, 'hex');
};
exports.GetHash = GetHash;
const Sign = (encodedTx, privateKeyHex) => {
    if (DEBUG) {
        logger_js_1.logger.log('sign', 'encodedTx', encodedTx);
    }
    const privateKey = Buffer.from(privateKeyHex, 'hex');
    if (DEBUG) {
        logger_js_1.logger.log('sign', 'privateKey', privateKey.toString('hex'));
    }
    const hash = (0, exports.GetHash)(encodedTx);
    if (DEBUG) {
        logger_js_1.logger.log('sign', 'hash', hash.toString('hex'));
    }
    const signature = (0, exports.SignBytes)(hash, privateKey);
    if (DEBUG) {
        logger_js_1.logger.log('sign', 'signature', signature);
    }
    return signature.toLowerCase();
};
exports.Sign = Sign;
const Verify = (encodedTx, signatureHex, publicKeyHex) => {
    if (DEBUG) {
        logger_js_1.logger.log('verify', 'encodedTx', encodedTx);
        logger_js_1.logger.log('verify', 'signatureHex', signatureHex);
        logger_js_1.logger.log('verify', 'publicKeyHex', publicKeyHex);
    }
    const publicKeyPem = (0, exports.PublicToPem)(publicKeyHex);
    if (DEBUG) {
        logger_js_1.logger.log('verify', 'publicKeyPem', publicKeyPem);
    }
    const publicKeyObj = crypto.createPublicKey({
        key: publicKeyPem,
        format: 'pem',
        type: 'spki',
    });
    const signature = Buffer.from(signatureHex, 'hex');
    const hash = (0, exports.GetHash)(encodedTx);
    if (DEBUG) {
        logger_js_1.logger.log('verify', 'hash', hash.toString('hex'));
    }
    return crypto.verify(undefined, hash, publicKeyObj, signature);
};
exports.Verify = Verify;
const GetPublicFromPrivate = (privateKey) => {
    const privateKeyDer = (0, exports.PrivateToDer)(privateKey);
    const privateKeyObj = crypto.createPrivateKey({
        key: privateKeyDer,
        format: 'der',
        type: 'pkcs8',
    });
    const privateKeyString = privateKeyObj.export({ format: 'der', type: 'pkcs8' });
    /*const publicKeyObj = crypto.createPublicKey({
      key: privateKeyObj,
      format: 'pem',
      type: 'sec1',
    });*/
    const publicKeyObj = crypto.createPublicKey({
        key: privateKeyString,
        format: 'pem',
        type: 'spki',
    });
    const encodedHex = publicKeyObj
        .export({ format: 'der', type: 'spki' })
        .toString('hex')
        .toUpperCase();
    if (encodedHex.startsWith(PUBLIC_KEY_PREFIX)) {
        return encodedHex.substring(PUBLIC_KEY_PREFIX.length);
    }
    else {
        throw new Error(`unknown prefix, expecting '${PUBLIC_KEY_PREFIX}' cannot decode public key '${encodedHex}'`);
    }
};
exports.GetPublicFromPrivate = GetPublicFromPrivate;
