// phantasma-sdk-ts/dist/esm/core/tx/ContractTxHelper.js
import { ProofOfWork } from '../link/interfaces/ProofOfWork.js';
import { DomainSettings } from '../types/DomainSettings.js';
import { bytesToHex } from '../utils/index.js';
import { ScriptBuilder } from '../vm/index.js';
import { Transaction } from './Transaction.js';
import { coerceContractBytes, normalizeContractName } from './ContractArtifacts.js';
export class ContractTxHelper {
    static buildDeployScript(params) {
        return this.buildContractLifecycleScript('Runtime.DeployContract', params);
    }
    static buildUpgradeScript(params) {
        return this.buildContractLifecycleScript('Runtime.UpgradeContract', params);
    }
    static buildDeployTransaction(params) {
        return this.buildContractLifecycleTransaction(this.buildDeployScript(params), params);
    }
    static buildUpgradeTransaction(params) {
        return this.buildContractLifecycleTransaction(this.buildUpgradeScript(params), params);
    }
    static buildDeployTransactionAndEncode(params) {
        const tx = this.buildDeployTransaction(params);
        this.signTransaction(tx, params.signer, params.proofOfWork);
        return tx.ToStringEncoded(true).toUpperCase();
    }
    static buildUpgradeTransactionAndEncode(params) {
        const tx = this.buildUpgradeTransaction(params);
        this.signTransaction(tx, params.signer, params.proofOfWork);
        return tx.ToStringEncoded(true).toUpperCase();
    }
    static buildDeployScriptFromBundle(bundle, from, gasPrice, gasLimit) {
        return this.buildDeployScript({
            from,
            contractName: bundle.contractName,
            script: bundle.script,
            abi: bundle.abi,
            gasPrice,
            gasLimit,
        });
    }
    static buildUpgradeScriptFromBundle(bundle, from, gasPrice, gasLimit) {
        return this.buildUpgradeScript({
            from,
            contractName: bundle.contractName,
            script: bundle.script,
            abi: bundle.abi,
            gasPrice,
            gasLimit,
        });
    }
    static encodePayloadText(text) {
        return bytesToHex(new TextEncoder().encode(text));
    }
    static buildContractLifecycleScript(interopName, params) {
        // Keep deploy/upgrade on the same legacy VM script path so CLI and wallet flows generate identical transactions.
        const fromAddress = this.normalizeAddress(params.from);
        const contractName = normalizeContractName(params.contractName);
        const scriptBytes = coerceContractBytes(params.script, 'script');
        const abiBytes = coerceContractBytes(params.abi, 'abi');
        const gasPrice = this.normalizeSmallInteger(params.gasPrice ?? this.DefaultGasPrice, 'gasPrice');
        const gasLimit = this.normalizeSmallInteger(params.gasLimit ?? this.DefaultGasLimit, 'gasLimit');
        const nullAddress = new ScriptBuilder().NullAddress;
        return new ScriptBuilder()
            .BeginScript()
            .AllowGas(fromAddress, nullAddress, gasPrice, gasLimit)
            .CallInterop(interopName, [fromAddress, contractName, scriptBytes, abiBytes])
            .SpendGas(fromAddress)
            .EndScript();
    }
    static buildContractLifecycleTransaction(scriptHex, params) {
        // Transaction construction stays separate from signing so callers can inspect or override the unsigned tx first.
        const nexus = params.nexus.trim();
        const chain = (params.chain ?? DomainSettings.RootChainName).trim();
        if (!nexus) {
            throw new Error('nexus cannot be empty');
        }
        if (!chain) {
            throw new Error('chain cannot be empty');
        }
        const payloadHex = params.payloadHex?.trim() ?? '';
        return new Transaction(nexus, chain, scriptHex, params.expiration ?? new Date(Date.now() + 5 * 60 * 1000), payloadHex);
    }
    static signTransaction(tx, signer, proofOfWork = ProofOfWork.Minimal) {
        // Mining/signing is the only mutation stage; everything above should remain deterministic and side-effect free.
        if (proofOfWork > 0) {
            tx.mineTransaction(proofOfWork);
        }
        if (typeof signer === 'string') {
            const trimmedWif = signer.trim();
            if (!trimmedWif) {
                throw new Error('signer WIF cannot be empty');
            }
            tx.sign(trimmedWif);
            return;
        }
        tx.signWithKeys(signer);
    }
    static normalizeAddress(address) {
        if (typeof address === 'string') {
            const trimmed = address.trim();
            if (!trimmed) {
                throw new Error('from cannot be empty');
            }
            return trimmed;
        }
        return address;
    }
    static normalizeSmallInteger(value, label) {
        const numeric = typeof value === 'bigint' ? Number(value) : value;
        if (!Number.isSafeInteger(numeric) || numeric <= 0) {
            throw new Error(`${label} must be a positive safe integer`);
        }
        return numeric;
    }
}
ContractTxHelper.DefaultGasPrice = DomainSettings.DefaultMinimumGasFee;
ContractTxHelper.DefaultGasLimit = 100000;
