// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/Modules/SeriesInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeriesInfo = void 0;
const Bytes32_js_1 = require("../../Bytes32.js");
const VmStructSchema_js_1 = require("../Vm/VmStructSchema.js");
class SeriesInfo {
    constructor(init) {
        this.maxMint = 0;
        this.maxSupply = 0;
        this.owner = new Bytes32_js_1.Bytes32();
        this.metadata = new Uint8Array();
        this.rom = new VmStructSchema_js_1.VmStructSchema();
        this.ram = new VmStructSchema_js_1.VmStructSchema();
        Object.assign(this, init);
    }
    write(w) {
        w.write4u(this.maxMint);
        w.write4u(this.maxSupply);
        this.owner.write(w);
        w.writeArray(this.metadata);
        this.rom.write(w);
        this.ram.write(w);
    }
    read(r) {
        this.maxMint = r.read4u();
        this.maxSupply = r.read4u();
        this.owner = Bytes32_js_1.Bytes32.read(r);
        this.metadata = r.readArray();
        this.rom = r.readBlob(VmStructSchema_js_1.VmStructSchema);
        this.ram = r.readBlob(VmStructSchema_js_1.VmStructSchema);
    }
    static read(r) {
        const v = new SeriesInfo();
        v.read(r);
        return v;
    }
}
exports.SeriesInfo = SeriesInfo;
