// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/Modules/StandardMeta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardMeta = void 0;
const SmallString_js_1 = require("../../SmallString.js");
exports.StandardMeta = {
    id: new SmallString_js_1.SmallString('_i'),
    Chain: {
        address: new SmallString_js_1.SmallString('_a'),
        name: new SmallString_js_1.SmallString('_n'),
        nexus: new SmallString_js_1.SmallString('_x'),
        tokenomics: new SmallString_js_1.SmallString('_t'),
    },
    Token: {
        staking_org_id: new SmallString_js_1.SmallString('_soi'),
        staking_org_threshold: new SmallString_js_1.SmallString('_sot'),
        staking_reward_token: new SmallString_js_1.SmallString('_srt'),
        staking_reward_period: new SmallString_js_1.SmallString('_srp'),
        staking_reward_mul: new SmallString_js_1.SmallString('_srm'),
        staking_reward_div: new SmallString_js_1.SmallString('_srd'),
        staking_lock: new SmallString_js_1.SmallString('_sl'),
        staking_booster_token: new SmallString_js_1.SmallString('_sbt'),
        staking_booster_mul: new SmallString_js_1.SmallString('_sbm'),
        staking_booster_div: new SmallString_js_1.SmallString('_sbd'),
        staking_booster_limit: new SmallString_js_1.SmallString('_sbl'),
        phantasma_script: new SmallString_js_1.SmallString('_phs'),
        phantasma_abi: new SmallString_js_1.SmallString('_phb'),
        pre_burn: new SmallString_js_1.SmallString('_brn'),
    },
};
