// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/Modules/TokenListing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenListing = exports.ListingType = void 0;
const Bytes32_js_1 = require("../../Bytes32.js");
const IntX_js_1 = require("../../IntX.js");
var ListingType;
(function (ListingType) {
    ListingType[ListingType["FixedPrice"] = 0] = "FixedPrice";
})(ListingType = exports.ListingType || (exports.ListingType = {}));
class TokenListing {
    constructor(type = ListingType.FixedPrice, seller = new Bytes32_js_1.Bytes32(), quoteTokenId = 0n, price = new IntX_js_1.IntX(), startDate = 0n, endDate = 0n) {
        this.type = type;
        this.seller = seller;
        this.quoteTokenId = quoteTokenId;
        this.price = price;
        this.startDate = startDate;
        this.endDate = endDate;
    }
    write(w) {
        w.write1(this.type);
        this.seller.write(w);
        w.write8u(this.quoteTokenId);
        this.price.write(w);
        w.write8(this.startDate);
        w.write8(this.endDate);
    }
    read(r) {
        this.type = r.read1();
        this.seller = Bytes32_js_1.Bytes32.read(r);
        this.quoteTokenId = r.read8u();
        this.price = r.readBlob(IntX_js_1.IntX);
        this.startDate = r.read8();
        this.endDate = r.read8();
    }
    static read(r) {
        const v = new TokenListing();
        v.read(r);
        return v;
    }
}
exports.TokenListing = TokenListing;
