// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/Modules/TokenInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenInfo = void 0;
const IntX_js_1 = require("../../IntX.js");
const Bytes32_js_1 = require("../../Bytes32.js");
const SmallString_js_1 = require("../../SmallString.js");
const CarbonTokenFlags_js_1 = require("../CarbonTokenFlags.js");
class TokenInfo {
    constructor(init) {
        this.maxSupply = new IntX_js_1.IntX();
        this.flags = CarbonTokenFlags_js_1.CarbonTokenFlags.None;
        this.decimals = 0;
        this.owner = new Bytes32_js_1.Bytes32();
        this.symbol = new SmallString_js_1.SmallString();
        this.metadata = new Uint8Array();
        this.tokenSchemas = undefined;
        Object.assign(this, init);
    }
    write(w) {
        this.maxSupply.write(w);
        w.write1(this.flags & 0xff);
        w.write1(this.decimals & 0xff);
        this.owner.write(w);
        this.symbol.write(w);
        w.writeArray(this.metadata);
        if ((this.flags & CarbonTokenFlags_js_1.CarbonTokenFlags.NonFungible) !== 0) {
            w.writeArray(this.tokenSchemas ?? new Uint8Array());
        }
    }
    read(r) {
        this.maxSupply = r.readBlob(IntX_js_1.IntX);
        this.flags = r.read1();
        this.decimals = r.read1();
        this.owner = Bytes32_js_1.Bytes32.read(r);
        this.symbol = r.readBlob(SmallString_js_1.SmallString);
        this.metadata = r.readArray();
        if ((this.flags & CarbonTokenFlags_js_1.CarbonTokenFlags.NonFungible) !== 0) {
            this.tokenSchemas = r.readArray();
        }
        else {
            this.tokenSchemas = undefined;
        }
    }
    static read(r) {
        const v = new TokenInfo();
        v.read(r);
        return v;
    }
}
exports.TokenInfo = TokenInfo;
