// phantasma-sdk-ts/dist/cjs/core/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Int2Buffer = exports.hex2ascii = exports.bigIntToByteArray = exports.numberToByteArray = exports.uint8ArrayToBytes = exports.arrayNumberToUint8Array = exports.hexStringToUint8Array = exports.stringToUint8Array = exports.uint8ArrayToNumberArray = exports.uint8ArrayToStringDefault = exports.uint8ArrayToString = exports.encodeBase16 = exports.decodeBase16 = exports.getDifficulty = exports.reverseHex = void 0;
const Hex_js_1 = require("./Hex.js");
__exportStar(require("./Hex.js"), exports);
__exportStar(require("./ValidationUtils.js"), exports);
__exportStar(require("./logger.js"), exports);
function reverseHex(hex) {
    let out = '';
    for (let i = hex.length - 2; i >= 0; i -= 2) {
        out += hex.substr(i, 2);
    }
    return out;
}
exports.reverseHex = reverseHex;
function getDifficulty(transactionHash) {
    let bytes = (0, Hex_js_1.hexToBytes)(transactionHash).reverse();
    let result = 0;
    for (let i = 0; i < bytes.length; i++) {
        var n = bytes[i];
        for (let j = 0; j < 8; j++) {
            if ((bytes[i] & (1 << j)) != 0) {
                result = 1 + (i << 3) + j;
            }
        }
    }
    return 256 - result;
}
exports.getDifficulty = getDifficulty;
function decodeBase16(hex) {
    let str = '';
    for (let i = 0; i < hex.length; i += 2) {
        str += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
    }
    return str;
}
exports.decodeBase16 = decodeBase16;
function encodeBase16(str) {
    return str
        .split('')
        .map((c) => c.charCodeAt(0).toString(16).padStart(2, '0'))
        .join('')
        .toUpperCase();
}
exports.encodeBase16 = encodeBase16;
function uint8ArrayToString(array) {
    let result = '';
    for (let i = 0; i < array.length; i++) {
        result += String.fromCharCode(array[i]);
    }
    return result;
}
exports.uint8ArrayToString = uint8ArrayToString;
function uint8ArrayToStringDefault(array) {
    let result = '';
    for (let i = 0; i < array.length; i++) {
        result += array[i].toString(16);
    }
    return result;
}
exports.uint8ArrayToStringDefault = uint8ArrayToStringDefault;
function uint8ArrayToNumberArray(array) {
    let result = [];
    for (let i = 0; i < array.length; i++) {
        result.push(array[i]);
    }
    return result;
}
exports.uint8ArrayToNumberArray = uint8ArrayToNumberArray;
function stringToUint8Array(str) {
    let result = new Uint8Array(str.length);
    for (let i = 0; i < str.length; i++) {
        result[i] = str.charCodeAt(i);
    }
    return result;
}
exports.stringToUint8Array = stringToUint8Array;
function hexStringToUint8Array(str) {
    let result = new Uint8Array(str.length);
    for (let i = 0; i < str.length; i++) {
        result[i] = str.charCodeAt(i).toString(16).charCodeAt(0);
    }
    return result;
}
exports.hexStringToUint8Array = hexStringToUint8Array;
function arrayNumberToUint8Array(arr) {
    let result = new Uint8Array(arr.length);
    for (let i = 0; i < arr.length; i++) {
        result[i] = arr[i];
    }
    return result;
}
exports.arrayNumberToUint8Array = arrayNumberToUint8Array;
function uint8ArrayToBytes(array) {
    let result = [];
    for (let i = 0; i < array.length; i++) {
        result.push(array[i]);
    }
    return result;
}
exports.uint8ArrayToBytes = uint8ArrayToBytes;
function numberToByteArray(num, size) {
    if (size === undefined) {
        if (num < 0xfd) {
            size = 1;
        }
        else if (num <= 0xfffd) {
            size = 2;
        }
        else if (num <= 0xffffd) {
            size = 3;
        }
        else if (num <= 0xfffffffd) {
            size = 4;
        }
        else if (num <= 0xffffffff) {
            size = 5;
        }
        else if (num <= 0xffffffffffffffff) {
            size = 9;
        }
    }
    var bytes = new Uint8Array(size);
    var i = 0;
    do {
        bytes[i++] = num & 0xff;
        num = num >> 8;
    } while (num);
    return bytes;
}
exports.numberToByteArray = numberToByteArray;
function bigIntToByteArray(bigint) {
    // Get a big-endian byte representation of the bigint
    var bytes = bigint.toString(16).padStart(64, '0');
    var byteArray = new Uint8Array(bytes.length / 2);
    for (var i = 0; i < bytes.length; i += 2) {
        byteArray[i / 2] = parseInt(bytes.substring(i, i + 2), 16);
    }
    return byteArray;
}
exports.bigIntToByteArray = bigIntToByteArray;
const hex2ascii = (hexx) => {
    const hex = hexx.toString();
    let str = '';
    for (let i = 0; i < hex.length; i += 2) {
        const char = hex.substr(i, 2);
        const charCode = String.fromCharCode(parseInt(char, 16));
        // console.log('char', char, 'charCode', charCode);
        str += charCode;
    }
    return str;
};
exports.hex2ascii = hex2ascii;
const Int2Buffer = (i) => {
    let hex = i.toString(16).toUpperCase();
    if (hex.length % 2 === 1) {
        hex = '0' + hex;
    }
    return Buffer.from(hex, 'hex');
};
exports.Int2Buffer = Int2Buffer;
