// phantasma-sdk-ts/dist/esm/core/types/Consensus.js
import { bytesToHex, stringToUint8Array } from '../utils/index.js';
import { Base16 } from './Extensions/index.js';
import { Timestamp } from './Timestamp.js';
export var ConsensusMode;
(function (ConsensusMode) {
    ConsensusMode[ConsensusMode["Unanimity"] = 0] = "Unanimity";
    ConsensusMode[ConsensusMode["Majority"] = 1] = "Majority";
    ConsensusMode[ConsensusMode["Popularity"] = 2] = "Popularity";
    ConsensusMode[ConsensusMode["Ranking"] = 3] = "Ranking";
})(ConsensusMode || (ConsensusMode = {}));
export var PollState;
(function (PollState) {
    PollState[PollState["Inactive"] = 0] = "Inactive";
    PollState[PollState["Active"] = 1] = "Active";
    PollState[PollState["Consensus"] = 2] = "Consensus";
    PollState[PollState["Failure"] = 3] = "Failure";
})(PollState || (PollState = {}));
export class PollChoice {
    constructor(value) {
        if (value instanceof Array)
            this.value = Base16.decode(bytesToHex(value));
        else
            this.value = value;
    }
    SerializeData(writer) {
        writer.writeByteArray(stringToUint8Array(this.value));
    }
    UnserializeData(reader) {
        this.value = Base16.decode(reader.readByteArray());
    }
    static Unserialize(reader) {
        let pollChoice = new PollChoice('');
        pollChoice.UnserializeData(reader);
        return pollChoice;
    }
}
export class PollValue {
    SerializeData(writer) {
        writer.writeByteArray(stringToUint8Array(this.value));
        writer.writeBigInteger(this.ranking);
        writer.writeBigInteger(this.votes);
    }
    UnserializeData(reader) {
        this.value = Base16.decode(reader.readByteArray());
        this.ranking = reader.readBigInteger();
        this.votes = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let pollValue = new PollValue();
        pollValue.UnserializeData(reader);
        return pollValue;
    }
}
export class PollVote {
    SerializeData(writer) {
        writer.writeBigInteger(this.index);
        writer.writeBigInteger(this.percentage);
    }
    UnserializeData(reader) {
        this.index = reader.readBigInteger();
        this.percentage = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let pollVote = new PollVote();
        pollVote.UnserializeData(reader);
        return pollVote;
    }
}
export class ConsensusPoll {
    constructor() {
        this.subject = '';
        this.organization = '';
        this.mode = ConsensusMode.Unanimity;
        this.state = PollState.Inactive;
        this.entries = [];
        this.round = BigInt(0);
        this.startTime = Timestamp.null;
        this.endTime = Timestamp.null;
        this.choicesPerUser = BigInt(0);
        this.totalVotes = BigInt(0);
    }
    SerializeData(writer) {
        writer.writeString(this.subject);
        writer.writeString(this.organization);
        writer.writeByte(this.mode);
        writer.writeByte(this.state);
        writer.writeByte(this.entries.length);
        this.entries.forEach((entry) => {
            entry.SerializeData(writer);
        });
        writer.writeBigInteger(this.round);
        writer.writeTimestamp(this.startTime);
        writer.writeTimestamp(this.endTime);
        writer.writeBigInteger(this.choicesPerUser);
        writer.writeBigInteger(this.totalVotes);
    }
    UnserializeData(reader) {
        this.subject = reader.readString();
        this.organization = reader.readString();
        this.mode = reader.readByte();
        this.state = reader.readByte();
        this.entries = [];
        const entriesLength = reader.readUnsignedInt();
        for (let i = 0; i < entriesLength; i++) {
            this.entries.push(PollValue.Unserialize(reader));
        }
        this.round = reader.readBigInteger();
        this.startTime = reader.readTimestamp();
        this.endTime = reader.readTimestamp();
        this.choicesPerUser = reader.readBigInteger();
        this.totalVotes = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let consensusPoll = new ConsensusPoll();
        consensusPoll.UnserializeData(reader);
        return consensusPoll;
    }
}
export class PollPresence {
    SerializeData(writer) {
        writer.writeString(this.subject);
    }
    UnserializeData(reader) {
        this.subject = reader.readString();
        this.round = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let pollPresence = new PollPresence();
        pollPresence.UnserializeData(reader);
        return pollPresence;
    }
}
