// phantasma-sdk-ts/dist/esm/core/types/Ed25519Signature.js
import { SignatureKind } from '../interfaces/Signature.js';
import pkg from 'elliptic';
import { stringToUint8Array, bytesToHex } from '../utils/index.js';
import { PBinaryWriter } from './Extensions/index.js';
const { eddsa } = pkg;
const ed25519 = new eddsa('ed25519');
export class Ed25519Signature {
    constructor(bytes) {
        this.Kind = SignatureKind.Ed25519;
        this.Bytes = bytes;
    }
    Verify(message, address) {
        return this.VerifyMultiple(message, [address]);
    }
    VerifyMultiple(message, addresses) {
        for (const address of addresses) {
            if (!address.IsUser) {
                continue;
            }
            const pubKey = address.ToByteArray().slice(2);
            const msgBytes = Buffer.from(message);
            const sigHex = bytesToHex(this.Bytes);
            const pubKeyHex = bytesToHex(pubKey);
            if (ed25519.verify(msgBytes, sigHex, pubKeyHex)) {
                return true;
            }
        }
        return false;
    }
    SerializeData(writer) {
        //writer.writeString(uint8ArrayToString(this.Bytes));
        writer.writeByteArray(this.Bytes);
    }
    UnserializeData(reader) {
        this.Bytes = stringToUint8Array(reader.readString());
    }
    ToByteArray() {
        const stream = new Uint8Array(64);
        const writer = new PBinaryWriter(stream);
        this.SerializeData(writer);
        return new Uint8Array(stream);
    }
    static Generate(keypair, message) {
        const msgHashHex = Buffer.from(bytesToHex(message), 'hex');
        //const msgHashHex = uint8ArrayToString(message);
        const privateKeyBuffer = Buffer.from(bytesToHex(keypair.PrivateKey), 'hex');
        //const privateKeyBuffer = uint8ArrayToString(keypair.PrivateKey);
        const sign = ed25519.sign(msgHashHex, privateKeyBuffer);
        return new Ed25519Signature(sign.toBytes());
    }
}
