// phantasma-sdk-ts/dist/esm/core/types/PhantasmaKeys.js
import { Address } from './Address.js';
import base58 from 'bs58';
import WIF from 'wif';
import { bytesToHex } from '../utils/index.js';
import { Ed25519Signature } from './Ed25519Signature.js';
import pkg from 'elliptic';
import { Entropy } from './Entropy.js';
import { getWifFromPrivateKey, } from '../tx/index.js';
const { eddsa } = pkg;
const ed25519 = new eddsa('ed25519');
export class PhantasmaKeys {
    get PrivateKey() {
        return this._privateKey;
    }
    get PublicKey() {
        return this._publicKey;
    }
    constructor(privateKey) {
        if (privateKey.length == 64) {
            privateKey = privateKey.slice(0, 32);
        }
        if (privateKey.length != PhantasmaKeys.PrivateKeyLength) {
            throw new Error(`privateKey should have length ${PhantasmaKeys.PrivateKeyLength} but has ${privateKey.length}`);
        }
        this._privateKey = new Uint8Array(PhantasmaKeys.PrivateKeyLength);
        this._privateKey.set(privateKey);
        let wif = getWifFromPrivateKey(bytesToHex(privateKey));
        const privateKeyString = bytesToHex(this._privateKey);
        const privateKeyBuffer = Buffer.from(privateKeyString, 'hex');
        const publicKey = ed25519.keyFromSecret(privateKeyBuffer).getPublic();
        this._publicKey = publicKey;
        this.Address = Address.FromKey(this);
    }
    toString() {
        return this.Address.Text;
    }
    static generate() {
        const privateKey = Entropy.GetRandomBytes(PhantasmaKeys.PrivateKeyLength);
        const pair = new PhantasmaKeys(privateKey);
        return pair;
    }
    static fromWIF(wif) {
        if (!wif) {
            throw new Error('WIF required');
        }
        let data = base58.decode(wif); // checkdecode
        if (data.length == 38) {
            data = data.slice(0, 34);
        }
        if (data.length != 34 || data[0] != 0x80 || data[33] != 0x01) {
            throw new Error('Invalid WIF format');
        }
        const privateKey = data.slice(1, 33);
        return new PhantasmaKeys(privateKey);
    }
    toWIF() {
        const privateKeyString = bytesToHex(this._privateKey);
        const privatekeyBuffer = Buffer.from(privateKeyString, 'hex');
        const wif = WIF.encode(128, privatekeyBuffer, true); //uint8ArrayToHex(data); // .base58CheckEncode();
        return wif;
    }
    static xor(x, y) {
        if (x.length != y.length) {
            throw new Error('x and y should have the same length');
        }
        const result = new Uint8Array(x.length);
        for (let i = 0; i < x.length; i++) {
            result[i] = x[i] ^ y[i];
        }
        return result;
    }
    Sign(msg, customSignFunction) {
        return Ed25519Signature.Generate(this, msg);
    }
}
PhantasmaKeys.PrivateKeyLength = 32;
