// phantasma-sdk-ts/dist/esm/core/types/Serialization.js
import { PBinaryReader, PBinaryWriter } from "./Extensions/index.js";
import { Timestamp } from "./Timestamp.js";
import { stringToUint8Array } from "../utils/index.js";
//function CustomWriter(writer: PBinaryWriter, obj: any): void;
//function CustomReader(reader: PBinaryReader): any;
export class CustomSerializer {
    constructor(reader, writer) {
        this.Read = reader;
        this.Write = writer;
    }
}
export class Serialization {
    static RegisterType(type, reader, writer) {
        Serialization._customSerializers[typeof type] = new CustomSerializer(reader, writer);
    }
    static SerializeEnum(obj) {
        if (!obj) {
            return new Uint8Array();
        }
        if (obj instanceof Uint8Array) {
            return obj;
        }
        let writer = new PBinaryWriter();
        writer.writeEnum(obj);
        return writer.toUint8Array();
    }
    static Serialize(obj) {
        if (!obj) {
            return new Uint8Array();
        }
        if (obj instanceof Uint8Array) {
            return obj;
        }
        //let jsonString = JSON.stringify(obj);
        let writer = new PBinaryWriter();
        this.SerializeObject(writer, obj, typeof obj);
        //let jsonAsUint8 = new TextEncoder().encode(jsonString);
        return writer.toUint8Array();
    }
    static SerializeObject(writer, obj, type) {
        if (type == null || type == undefined) {
            type = typeof obj;
            this.SerializeObject(writer, obj, type);
            return;
        }
        if (Serialization._customSerializers.has(typeof type)) {
            var serializer = Serialization._customSerializers[typeof type];
            serializer.Write(writer, obj);
            return;
        }
        /*if (typeof obj == "void") {
          return;
        }*/
        const structType = Object.getPrototypeOf(obj).constructor.name;
        const result = Object.apply(structType);
        let localType = Object.apply(typeof type);
        //console.log(localType);
        if (obj instanceof Boolean || typeof obj == "boolean") {
            writer.writeByte(obj ? 1 : 0);
            return;
        }
        else if (obj instanceof Number || typeof obj == "number") {
            writer.writeByte(stringToUint8Array(obj.toString()).length);
            writer.writeVarInt(obj);
            return;
        }
        else if (obj instanceof BigInt || typeof obj == "bigint") {
            writer.writeBigInteger(obj);
        }
        else if (obj instanceof String || typeof obj == "string") {
            writer.writeString(obj);
            return;
        }
        else if (obj instanceof Timestamp) {
            writer.writeTimestamp(obj);
            return;
        }
        else if (obj instanceof Date) {
            writer.writeDateTime(obj);
            return;
        }
        else if (typeof obj.UnserializeData === "function" &&
            typeof obj.SerializeData === "function") {
            obj.SerializeData(writer);
            return;
        }
        else if (Array.isArray(obj)) {
            writer.writeVarInt(obj.length);
            obj.forEach((entry) => {
                this.SerializeObject(writer, entry, typeof entry);
            });
            return;
        }
        else if (Object.getPrototypeOf(type) == "enum") {
            writer.writeByte(obj);
            return;
        }
        else if (obj instanceof Uint8Array) {
            writer.writeByteArray(Array.from(obj));
        }
        else {
            // TODO: Add support for other types
            // Get the keys of the object
            const fields = Object.keys(obj);
            fields.forEach((field) => {
                let value = obj[field];
                this.SerializeObject(writer, value, typeof value);
            });
        }
    }
    static Unserialize(bytesOrBytes, type) {
        if (bytesOrBytes instanceof PBinaryReader) {
            return Serialization.UnserializeObject(bytesOrBytes, type);
        }
        if (!bytesOrBytes || bytesOrBytes.length === 0) {
            return null;
        }
        //let type = Object.prototype.propertyIsEnumerable(T);
        let stream = new PBinaryReader(bytesOrBytes);
        return Serialization.UnserializeObject(stream, type);
    }
    static UnserializeObject(reader, type) {
        if (Serialization._customSerializers.has(typeof type)) {
            var serializer = Serialization._customSerializers[typeof type];
            return serializer.Read(reader);
        }
        if (type == null || type == undefined) {
            return null;
        }
        let localType; //: typeof type;
        if (type.name != undefined) {
            let className = type.name;
            localType = Object.apply(className);
        }
        else {
            localType = new type();
        }
        if (localType instanceof Boolean ||
            typeof localType == "boolean" ||
            type.name == "Boolean") {
            return reader.readBoolean();
        }
        else if (localType instanceof Number || typeof localType == "number") {
            return reader.readVarInt();
        }
        else if (localType instanceof BigInt || typeof localType == "bigint") {
            return reader.readBigInteger();
        }
        else if (localType instanceof String || typeof localType == "string") {
            return reader.readVarString();
        }
        else if (localType instanceof Timestamp) {
            return new Timestamp(reader.readVarInt());
        }
        else if (typeof localType.UnserializeData === "function" &&
            typeof localType.SerializeData === "function") {
            let obj = localType;
            obj.UnserializeData(reader);
            return obj;
        }
        else if (Array.isArray(type)) {
            var len = reader.readByte();
            var arr = new Array(len);
            for (let i = 0; i < len; i++) {
                arr[i] = this.UnserializeObject(reader, type[i]);
            }
            return arr;
        }
        else if (Object.getPrototypeOf(type) == "enum") {
            return reader.readByte();
        }
        else {
            const fields = Object.keys(localType);
            /*console.log(fields);
            fields.forEach((field) => {
              localType[field] = this.UnserializeObject(
                reader,
                typeof localType[field]
              );
            });
            return localType as T;*/
        }
    }
}
Serialization._customSerializers = new Map(); //: { [key: string]: CustomSerializer };
