// phantasma-sdk-ts/dist/esm/core/types/Timestamp.js
export class Timestamp {
    constructor(value) {
        this.value = value;
    }
    toString() {
        return new Date(this.value * 1000).toUTCString();
    }
    toStringFormat(format) {
        return new Date(this.value * 1000).toUTCString();
    }
    compareTo(other) {
        if (other.value === this.value) {
            return 0;
        }
        if (this.value < other.value) {
            return -1;
        }
        return 1;
    }
    equals(obj) {
        if (!(obj instanceof Timestamp)) {
            return false;
        }
        return this.value === obj.value;
    }
    getHashCode() {
        return this.value;
    }
    getSize() {
        return 4;
    }
    static equal(A, B) {
        return A.value === B.value;
    }
    static notEqual(A, B) {
        return !(A.value === B.value);
    }
    static lessThan(A, B) {
        return A.value < B.value;
    }
    static greaterThan(A, B) {
        return A.value > B.value;
    }
    static lessThanOrEqual(A, B) {
        return A.value <= B.value;
    }
    static greaterThanOrEqual(A, B) {
        return A.value >= B.value;
    }
    static subtract(A, B) {
        return A.value - B.value;
    }
    static fromNumber(ticks) {
        return new Timestamp(ticks);
    }
    static fromDate(time) {
        return new Timestamp(time.getTime() / 1000);
    }
    static addTimeSpan(A, B) {
        return A.value + B;
    }
    static subtractTimeSpan(A, B) {
        return A.value - B;
    }
    static Serialize() { }
    static Unserialize() { }
}
Timestamp.now = Date.now();
Timestamp.null = new Timestamp(0);
