// csharp-binary-stream/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endianness = exports.Encoding = exports.EncodingError = exports.InvalidArgumentError = exports.OutOfBoundsError = exports.InvalidUtf8CharacterError = exports.EndOfStreamError = exports.BinaryWriter = exports.BinaryReader = void 0;
const BinaryReader_1 = require("./BinaryReader");
Object.defineProperty(exports, "BinaryReader", { enumerable: true, get: function () { return BinaryReader_1.BinaryReader; } });
const BinaryWriter_1 = require("./BinaryWriter");
Object.defineProperty(exports, "BinaryWriter", { enumerable: true, get: function () { return BinaryWriter_1.BinaryWriter; } });
const EndOfStreamError_1 = require("./errors/EndOfStreamError");
Object.defineProperty(exports, "EndOfStreamError", { enumerable: true, get: function () { return EndOfStreamError_1.EndOfStreamError; } });
const InvalidUtf8CharacterError_1 = require("./errors/InvalidUtf8CharacterError");
Object.defineProperty(exports, "InvalidUtf8CharacterError", { enumerable: true, get: function () { return InvalidUtf8CharacterError_1.InvalidUtf8CharacterError; } });
const OutOfBoundsError_1 = require("./errors/OutOfBoundsError");
Object.defineProperty(exports, "OutOfBoundsError", { enumerable: true, get: function () { return OutOfBoundsError_1.OutOfBoundsError; } });
const InvalidArgumentError_1 = require("./errors/InvalidArgumentError");
Object.defineProperty(exports, "InvalidArgumentError", { enumerable: true, get: function () { return InvalidArgumentError_1.InvalidArgumentError; } });
const EncodingError_1 = require("./errors/EncodingError");
Object.defineProperty(exports, "EncodingError", { enumerable: true, get: function () { return EncodingError_1.EncodingError; } });
const Encoding_1 = require("./Encoding");
Object.defineProperty(exports, "Encoding", { enumerable: true, get: function () { return Encoding_1.Encoding; } });
const Endianness_1 = require("./Endianness");
Object.defineProperty(exports, "Endianness", { enumerable: true, get: function () { return Endianness_1.Endianness; } });
