// phantasma-sdk-ts/dist/esm/core/types/Extensions/PBinaryReader.js
import BigInteger from 'big-integer';
import { BinaryReader } from 'csharp-binary-stream';
import { SignatureKind } from '../../interfaces/index.js';
import { bytesToHex, stringToUint8Array } from '../../utils/index.js';
import { VMType } from '../../vm/index.js';
import { Ed25519Signature } from '../Ed25519Signature.js';
import { Timestamp } from '../Timestamp.js';
export class PBinaryReader {
    get length() {
        return this.reader.length;
    }
    get position() {
        return this.reader.position;
    }
    set position(value) {
        this.reader.position = value;
    }
    get isEndOfStream() {
        return this.reader.isEndOfStream;
    }
    readBoolean() {
        return this.reader.readBoolean();
    }
    readByte() {
        return this.reader.readByte();
    }
    readBytes(bytesToRead) {
        return this.reader.readBytes(bytesToRead);
    }
    readSignedByte() {
        return this.reader.readSignedByte();
    }
    readShort() {
        return this.reader.readShort();
    }
    readUnsignedShort() {
        return this.reader.readUnsignedShort();
    }
    readInt() {
        return this.reader.readInt();
    }
    readUnsignedInt() {
        return this.reader.readUnsignedInt();
    }
    readLongString() {
        return this.reader.readLongString();
    }
    readLong() {
        return this.reader.readLong();
    }
    readUnsignedLongString() {
        return this.reader.readUnsignedLongString();
    }
    readUnsignedLong() {
        return this.reader.readUnsignedLong();
    }
    readFloat() {
        return this.reader.readFloat();
    }
    readDouble() {
        return this.reader.readDouble();
    }
    readChar(encoding) {
        return this.reader.readChar(encoding);
    }
    readChars(charactersToRead, encoding) {
        return this.reader.readChars(charactersToRead, encoding);
    }
    readCharBytes(bytesToRead, encoding) {
        return this.reader.readCharBytes(bytesToRead, encoding);
    }
    constructor(arg1) {
        this.reader = new BinaryReader(arg1);
    }
    read(numBytes) {
        var res = bytesToHex(this.readBytes(numBytes)).substr(0, numBytes * 2);
        //this.position += numBytes;
        return res;
    }
    readString() {
        var len = this.readVarInt();
        return this.readStringBytes(len);
    }
    readStringBytes(numBytes) {
        var res = '';
        for (var i = 0; i < numBytes; ++i) {
            res += String.fromCharCode(this.readByte());
        }
        return res;
    }
    readBigInteger() {
        // TO DO: implement negative numbers
        var len = this.readVarInt();
        var res = 0;
        var stringBytes = this.read(len);
        [...stringBytes.match(/.{1,2}/g)]
            .reverse()
            .forEach((c) => (res = res * 256 + parseInt(c, 16)));
        let bigInt = BigInt(res);
        return bigInt;
    }
    readBigIntAccurate() {
        var len = this.readVarInt();
        var res = BigInteger();
        var stringBytes = this.read(len);
        [...stringBytes.match(/.{1,2}/g)].reverse().forEach((c) => {
            res = res.times(256).plus(parseInt(c, 16));
        });
        return res.toString();
    }
    readSignatureV2() {
        let kind = this.readByte();
        let curve;
        let signature = new Ed25519Signature();
        switch (kind) {
            case SignatureKind.None:
                return null;
            case SignatureKind.Ed25519:
                let len = this.readVarInt();
                signature.Bytes = new Uint8Array(this.readBytes(len));
                break;
            case SignatureKind.ECDSA:
                curve = this.readByte();
                signature.Bytes = stringToUint8Array(this.readString());
                break;
            default:
                throw 'read signature: ' + kind;
        }
        return signature;
    }
    readSignature() {
        let kind = this.readByte();
        let signature = new Ed25519Signature();
        let curve;
        signature.Kind = kind;
        switch (kind) {
            case SignatureKind.None:
                return null;
            case SignatureKind.Ed25519:
                let len = this.readVarInt();
                signature.Bytes = stringToUint8Array(this.read(len));
                break;
            case SignatureKind.ECDSA:
                curve = this.readByte();
                signature.Bytes = stringToUint8Array(this.readString());
                break;
            default:
                throw 'read signature: ' + kind;
        }
        return signature;
    }
    readByteArray() {
        var res;
        var length = this.readVarInt();
        if (length == 0)
            return [];
        res = this.read(length);
        return res;
    }
    readTimestamp() {
        //var len = this.readByte();
        let result = 0;
        let bytes = this.read(4);
        //[...(bytes.match(/.{1,2}/g) as any)];
        bytes
            .match(/.{1,2}/g)
            .reverse()
            .forEach((c) => (result = result * 256 + parseInt(c, 16)));
        let timestamp = new Timestamp(result);
        return timestamp;
    }
    readVarInt() {
        var len = this.readByte();
        var res = 0;
        if (len === 0xfd) {
            [...this.read(2).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        else if (len === 0xfe) {
            [...this.read(4).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        else if (len === 0xff) {
            [...this.read(8).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        return len;
    }
    readVarString() {
        var len = this.readVarInt();
        if (len == 0)
            return '';
        return this.readStringBytes(len);
    }
    readVmObject() {
        const type = this.readByte();
        switch (type) {
            case VMType.String:
                return this.readString();
            case VMType.Number:
                return this.readBigIntAccurate();
            case VMType.Bool:
                return this.readByte() != 0;
            case VMType.Struct:
                const numFields = this.readVarInt();
                let res = {};
                for (let i = 0; i < numFields; ++i) {
                    const key = this.readVmObject();
                    const value = this.readVmObject();
                    res[key] = value;
                }
                return res;
            case VMType.Enum:
                return this.readVarInt();
            case VMType.Object:
                const numBytes = this.readVarInt();
                return this.read(numBytes);
            default:
                return 'unsupported type ' + type;
        }
    }
}
