// phantasma-sdk-ts/dist/cjs/core/tx/ContractTxHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractTxHelper = void 0;
const ProofOfWork_js_1 = require("../link/interfaces/ProofOfWork.js");
const DomainSettings_js_1 = require("../types/DomainSettings.js");
const index_js_1 = require("../utils/index.js");
const index_js_2 = require("../vm/index.js");
const Transaction_js_1 = require("./Transaction.js");
const ContractArtifacts_js_1 = require("./ContractArtifacts.js");
class ContractTxHelper {
    static buildDeployScript(params) {
        return this.buildContractLifecycleScript('Runtime.DeployContract', params);
    }
    static buildUpgradeScript(params) {
        return this.buildContractLifecycleScript('Runtime.UpgradeContract', params);
    }
    static buildDeployTransaction(params) {
        return this.buildContractLifecycleTransaction(this.buildDeployScript(params), params);
    }
    static buildUpgradeTransaction(params) {
        return this.buildContractLifecycleTransaction(this.buildUpgradeScript(params), params);
    }
    static buildDeployTransactionAndEncode(params) {
        const tx = this.buildDeployTransaction(params);
        this.signTransaction(tx, params.signer, params.proofOfWork);
        return tx.ToStringEncoded(true).toUpperCase();
    }
    static buildUpgradeTransactionAndEncode(params) {
        const tx = this.buildUpgradeTransaction(params);
        this.signTransaction(tx, params.signer, params.proofOfWork);
        return tx.ToStringEncoded(true).toUpperCase();
    }
    static buildDeployScriptFromBundle(bundle, from, gasPrice, gasLimit) {
        return this.buildDeployScript({
            from,
            contractName: bundle.contractName,
            script: bundle.script,
            abi: bundle.abi,
            gasPrice,
            gasLimit,
        });
    }
    static buildUpgradeScriptFromBundle(bundle, from, gasPrice, gasLimit) {
        return this.buildUpgradeScript({
            from,
            contractName: bundle.contractName,
            script: bundle.script,
            abi: bundle.abi,
            gasPrice,
            gasLimit,
        });
    }
    static encodePayloadText(text) {
        return (0, index_js_1.bytesToHex)(new TextEncoder().encode(text));
    }
    static buildContractLifecycleScript(interopName, params) {
        // Keep deploy/upgrade on the same legacy VM script path so CLI and wallet flows generate identical transactions.
        const fromAddress = this.normalizeAddress(params.from);
        const contractName = (0, ContractArtifacts_js_1.normalizeContractName)(params.contractName);
        const scriptBytes = (0, ContractArtifacts_js_1.coerceContractBytes)(params.script, 'script');
        const abiBytes = (0, ContractArtifacts_js_1.coerceContractBytes)(params.abi, 'abi');
        const gasPrice = this.normalizeSmallInteger(params.gasPrice ?? this.DefaultGasPrice, 'gasPrice');
        const gasLimit = this.normalizeSmallInteger(params.gasLimit ?? this.DefaultGasLimit, 'gasLimit');
        const nullAddress = new index_js_2.ScriptBuilder().NullAddress;
        return new index_js_2.ScriptBuilder()
            .BeginScript()
            .AllowGas(fromAddress, nullAddress, gasPrice, gasLimit)
            .CallInterop(interopName, [fromAddress, contractName, scriptBytes, abiBytes])
            .SpendGas(fromAddress)
            .EndScript();
    }
    static buildContractLifecycleTransaction(scriptHex, params) {
        // Transaction construction stays separate from signing so callers can inspect or override the unsigned tx first.
        const nexus = params.nexus.trim();
        const chain = (params.chain ?? DomainSettings_js_1.DomainSettings.RootChainName).trim();
        if (!nexus) {
            throw new Error('nexus cannot be empty');
        }
        if (!chain) {
            throw new Error('chain cannot be empty');
        }
        const payloadHex = params.payloadHex?.trim() ?? '';
        return new Transaction_js_1.Transaction(nexus, chain, scriptHex, params.expiration ?? new Date(Date.now() + 5 * 60 * 1000), payloadHex);
    }
    static signTransaction(tx, signer, proofOfWork = ProofOfWork_js_1.ProofOfWork.Minimal) {
        // Mining/signing is the only mutation stage; everything above should remain deterministic and side-effect free.
        if (proofOfWork > 0) {
            tx.mineTransaction(proofOfWork);
        }
        if (typeof signer === 'string') {
            const trimmedWif = signer.trim();
            if (!trimmedWif) {
                throw new Error('signer WIF cannot be empty');
            }
            tx.sign(trimmedWif);
            return;
        }
        tx.signWithKeys(signer);
    }
    static normalizeAddress(address) {
        if (typeof address === 'string') {
            const trimmed = address.trim();
            if (!trimmed) {
                throw new Error('from cannot be empty');
            }
            return trimmed;
        }
        return address;
    }
    static normalizeSmallInteger(value, label) {
        const numeric = typeof value === 'bigint' ? Number(value) : value;
        if (!Number.isSafeInteger(numeric) || numeric <= 0) {
            throw new Error(`${label} must be a positive safe integer`);
        }
        return numeric;
    }
}
exports.ContractTxHelper = ContractTxHelper;
ContractTxHelper.DefaultGasPrice = DomainSettings_js_1.DomainSettings.DefaultMinimumGasFee;
ContractTxHelper.DefaultGasLimit = 100000;
