// phantasma-sdk-ts/dist/cjs/core/tx/ContractArtifacts.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildContractArtifactBundle = exports.buildContractArtifactManifest = exports.coerceContractBytes = exports.normalizeContractName = void 0;
const sha256_js_1 = __importDefault(require("crypto-js/sha256.js"));
const enc_hex_js_1 = __importDefault(require("crypto-js/enc-hex.js"));
const index_js_1 = require("../utils/index.js");
function sha256Hex(bytes) {
    return (0, sha256_js_1.default)(enc_hex_js_1.default.parse((0, index_js_1.bytesToHex)(bytes))).toString(enc_hex_js_1.default);
}
function buildFileEntry(path, bytes) {
    const trimmedPath = path.trim();
    if (!trimmedPath) {
        throw new Error('artifact path cannot be empty');
    }
    return {
        path: trimmedPath,
        size: bytes.length,
        sha256: sha256Hex(bytes),
    };
}
function normalizeContractName(contractName) {
    const trimmed = contractName.trim();
    if (!trimmed) {
        throw new Error('contractName cannot be empty');
    }
    return trimmed;
}
exports.normalizeContractName = normalizeContractName;
function coerceContractBytes(input, label) {
    if (input instanceof Uint8Array) {
        if (input.length === 0) {
            throw new Error(`${label} cannot be empty`);
        }
        // Copy mutable buffers so callers can keep reusing their original arrays safely.
        return new Uint8Array(input);
    }
    if (typeof input === 'string') {
        const trimmed = input.trim();
        if (!trimmed) {
            throw new Error(`${label} cannot be empty`);
        }
        const bytes = (0, index_js_1.hexToBytes)(trimmed);
        if (bytes.length === 0) {
            throw new Error(`${label} cannot be empty`);
        }
        return bytes;
    }
    throw new Error(`${label} must be Uint8Array or hex string`);
}
exports.coerceContractBytes = coerceContractBytes;
function buildContractArtifactManifest(params) {
    // The manifest is the stable handoff contract between compiler, CLI, frontend, and runbooks.
    const contractName = normalizeContractName(params.contractName);
    const compilerName = params.compilerName.trim();
    const compilerVersion = params.compilerVersion.trim();
    if (!compilerName) {
        throw new Error('compilerName cannot be empty');
    }
    if (!compilerVersion) {
        throw new Error('compilerVersion cannot be empty');
    }
    const scriptBytes = coerceContractBytes(params.script, 'script');
    const abiBytes = coerceContractBytes(params.abi, 'abi');
    let debugEntry;
    if (params.debugPath !== undefined || params.debug !== undefined) {
        if (!params.debugPath || params.debug === undefined) {
            throw new Error('debugPath and debug must be provided together');
        }
        debugEntry = buildFileEntry(params.debugPath, coerceContractBytes(params.debug, 'debug'));
    }
    const manifest = {
        format: 'pha.contract.artifacts/v1',
        contractName,
        createdAtUtc: (params.createdAtUtc ?? new Date().toISOString()).trim(),
        compiler: {
            name: compilerName,
            version: compilerVersion,
        },
        files: {
            script: buildFileEntry(params.scriptPath, scriptBytes),
            abi: buildFileEntry(params.abiPath, abiBytes),
            ...(debugEntry ? { debug: debugEntry } : {}),
        },
        ...(params.sourceFile ? { sourceFile: params.sourceFile } : {}),
    };
    if (!manifest.createdAtUtc) {
        throw new Error('createdAtUtc cannot be empty');
    }
    return manifest;
}
exports.buildContractArtifactManifest = buildContractArtifactManifest;
function buildContractArtifactBundle(params) {
    // Keep bundle materialization small and explicit so higher layers can choose their own IO/layout policy.
    const bundle = {
        contractName: normalizeContractName(params.contractName),
        script: coerceContractBytes(params.script, 'script'),
        abi: coerceContractBytes(params.abi, 'abi'),
        ...(params.debug !== undefined ? { debug: coerceContractBytes(params.debug, 'debug') } : {}),
        ...(params.manifest ? { manifest: params.manifest } : {}),
    };
    return bundle;
}
exports.buildContractArtifactBundle = buildContractArtifactBundle;
