// phantasma-sdk-ts/dist/cjs/core/tx/Transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const elliptic_1 = __importDefault(require("elliptic"));
const logger_js_1 = require("../utils/logger.js");
const { eddsa } = elliptic_1.default;
const index_js_1 = require("../vm/index.js");
const index_js_2 = require("../utils/index.js");
const enc_hex_js_1 = __importDefault(require("crypto-js/enc-hex.js"));
const sha256_js_1 = __importDefault(require("crypto-js/sha256.js"));
const index_js_3 = require("../types/index.js");
const utils_js_1 = require("./utils.js");
const curve = new eddsa('ed25519');
class Transaction {
    static FromBytes(serializedData) {
        let transaction = new Transaction('', '', '', new Date(), '');
        return transaction.unserialize(serializedData);
    }
    constructor(nexusName, chainName, script, // Should be HexString
    expiration, payload // Should be HexString
    ) {
        this.nexusName = nexusName;
        this.chainName = chainName;
        this.script = script;
        this.expiration = expiration;
        this.payload = payload == null || payload == '' ? '7068616e7461736d612d7473' : payload;
        this.signatures = [];
    }
    sign(wif) {
        let _keys = index_js_3.PhantasmaKeys.fromWIF(wif);
        const msg = this.ToByteAray(false);
        let sig = _keys.Sign(msg);
        let sigs = [];
        if (this.signatures != null && this.signatures.length > 0) {
            sigs = this.signatures;
        }
        sigs.push(sig);
        this.signatures = sigs;
        //const signature = this.getSign(this.toString(false), privateKey);
        //this.signatures.unshift({ signature, kind: 1 });
    }
    signWithPrivateKey(privateKey) {
        const msg = this.ToByteAray(false);
        let sig = index_js_3.PhantasmaKeys.fromWIF((0, utils_js_1.getWifFromPrivateKey)(privateKey)).Sign(msg);
        let sigs = [];
        if (this.signatures != null && this.signatures.length > 0) {
            sigs = this.signatures;
        }
        sigs.push(sig);
        this.signatures = sigs;
    }
    signWithKeys(keys) {
        const msg = this.ToByteAray(false);
        let sig = keys.Sign(msg);
        let sigs = [];
        if (this.signatures != null && this.signatures.length > 0) {
            sigs = this.signatures;
        }
        sigs.push(sig);
        this.signatures = sigs;
    }
    VerifySignature(address) {
        // Verify that at least one stored signature matches the given address for the unsigned tx bytes.
        if (!this.signatures || this.signatures.length === 0) {
            return false;
        }
        const addr = typeof address === 'string' ? index_js_3.Address.FromText(address) : address;
        const message = this.ToByteAray(false);
        for (const sig of this.signatures) {
            if (sig.Verify(message, addr)) {
                return true;
            }
        }
        return false;
    }
    VerifySignatures(addresses) {
        // Verify which of the provided addresses signed this transaction (no public-key recovery).
        if (!this.signatures || this.signatures.length === 0 || !addresses || addresses.length === 0) {
            return { ok: false, matched: [] };
        }
        const message = this.ToByteAray(false);
        const matched = new Set();
        for (const address of addresses) {
            const addr = typeof address === 'string' ? index_js_3.Address.FromText(address) : address;
            for (const sig of this.signatures) {
                if (sig.Verify(message, addr)) {
                    matched.add(addr.Text);
                    break;
                }
            }
        }
        const result = Array.from(matched);
        return { ok: result.length > 0, matched: result };
    }
    GetUnsignedBytes() {
        // Expose unsigned bytes for diagnostics and SDK-level verification helpers.
        return this.ToByteAray(false);
    }
    GetSignatureInfo() {
        // Return signature metadata without exposing signature contents.
        if (!this.signatures || this.signatures.length === 0) {
            return [];
        }
        return this.signatures.map((sig) => ({
            kind: sig.Kind,
            length: sig.Bytes ? sig.Bytes.length : 0,
        }));
    }
    ToByteAray(withSignature) {
        let writer = new index_js_3.PBinaryWriter();
        writer.writeString(this.nexusName);
        writer.writeString(this.chainName);
        writer.AppendHexEncoded(this.script);
        writer.writeDateTime(this.expiration);
        writer.AppendHexEncoded(this.payload);
        if (withSignature) {
            writer.writeVarInt(this.signatures.length);
            this.signatures.forEach((sig) => {
                writer.writeSignature(sig);
                //writer.writeByte(sig.kind);
                //writer.writeByteArray(stringToUint8Array(sig.signature));
            });
        }
        return writer.toUint8Array();
    }
    UnserializeData(reader) {
        this.nexusName = reader.readString();
        this.chainName = reader.readString();
        this.script = (0, index_js_2.uint8ArrayToStringDefault)(reader.readByteArray());
        let time = reader.readTimestamp();
        this.expiration = new Date(time.toString());
        this.payload = (0, index_js_2.uint8ArrayToStringDefault)(reader.readByteArray());
        let sigCount = reader.readVarInt();
        for (let i = 0; i < sigCount; i++) {
            let sig = reader.readSignatureV2();
            this.signatures.push(sig);
        }
    }
    SerializeData(writer) {
        writer.writeString(this.nexusName);
        writer.writeString(this.chainName);
        writer.writeByteArray(index_js_3.Base16.decodeUint8Array(this.script));
        writer.writeDateTime(this.expiration);
        writer.writeByteArray(index_js_3.Base16.decodeUint8Array(this.payload));
        writer.writeVarInt(this.signatures.length);
        this.signatures.forEach((sig) => {
            writer.writeSignature(sig);
        });
    }
    toString(withSignature) {
        /*const utc = Date.UTC(
          this.expiration.getUTCFullYear(),
          this.expiration.getUTCMonth(),
          this.expiration.getUTCDate(),
          this.expiration.getUTCHours(),
          this.expiration.getUTCMinutes(),
          this.expiration.getUTCSeconds()
        );
        let num = utc / 1000;
    
        let a = (num & 0xff000000) >> 24;
        let b = (num & 0x00ff0000) >> 16;
        let c = (num & 0x0000ff00) >> 8;
        let d = num & 0x000000ff;
    
        let expirationBytes = [d, c, b, a];*/
        let sb = new index_js_1.ScriptBuilder()
            .EmitVarString(this.nexusName)
            .EmitVarString(this.chainName)
            .EmitVarInt(this.script.length / 2)
            .AppendHexEncoded(this.script)
            .EmitTimestamp(this.expiration)
            .EmitVarInt(this.payload.length / 2)
            .AppendHexEncoded(this.payload);
        if (withSignature) {
            sb.EmitVarInt(this.signatures.length);
            this.signatures.forEach((sig) => {
                logger_js_1.logger.log('adding signature ', sig);
                if (sig.Kind == 1) {
                    sb.AppendByte(1); // Signature Type
                    sb.EmitVarInt(sig.Bytes.length / 2);
                    sb.AppendHexEncoded((0, index_js_2.bytesToHex)(sig.Bytes));
                }
                else if (sig.Kind == 2) {
                    sb.AppendByte(2); // ECDSA Signature
                    sb.AppendByte(1); // Curve type secp256k1
                    sb.EmitVarInt(sig.Bytes.length / 2);
                    sb.AppendHexEncoded((0, index_js_2.bytesToHex)(sig.Bytes));
                }
            });
        }
        return sb.str;
    }
    ToStringEncoded(withSignature) {
        return index_js_3.Base16.encodeUint8Array(this.ToByteAray(withSignature));
    }
    getHash() {
        let generatedHash = (0, sha256_js_1.default)(enc_hex_js_1.default.parse(this.toString(false)));
        this.hash = (0, index_js_2.bytesToHex)((0, index_js_2.hexToBytes)(generatedHash.toString(enc_hex_js_1.default)).reverse());
        return this.hash;
    }
    mineTransaction(difficulty) {
        if (difficulty < 0 || difficulty > 256) {
            logger_js_1.logger.log('Error adding difficulty');
            return;
        }
        let nonce = 0;
        let deepCopy = new Transaction(JSON.parse(JSON.stringify(this.nexusName)), JSON.parse(JSON.stringify(this.chainName)), JSON.parse(JSON.stringify(this.script)), this.expiration, JSON.parse(JSON.stringify(this.payload)));
        let payload = Buffer.alloc(4);
        while (true) {
            if ((0, index_js_2.getDifficulty)(deepCopy.getHash()) >= difficulty) {
                this.payload = deepCopy.payload;
                logger_js_1.logger.log('It took ' + nonce + ' iterations to get a difficulty of >' + difficulty);
                return;
            }
            nonce++;
            payload[0] = (nonce >> 0) & 0xff;
            payload[1] = (nonce >> 8) & 0xff;
            payload[2] = (nonce >> 16) & 0xff;
            payload[3] = (nonce >> 24) & 0xff;
            deepCopy.payload = (0, index_js_2.bytesToHex)(payload);
        }
    }
    getSign(msgHex, privateKey) {
        const msgHashHex = Buffer.from(msgHex, 'hex');
        const privateKeyBuffer = Buffer.from(privateKey, 'hex');
        const sig = curve.sign(msgHashHex, privateKeyBuffer);
        return sig.toHex();
    }
    unserialize(serializedData) {
        let dec = new index_js_1.Decoder(serializedData);
        let nexusName = dec.readString();
        let chainName = dec.readString();
        let scriptLength = dec.readVarInt();
        let script = dec.read(scriptLength);
        let date = new Date(dec.readTimestamp() * 1000);
        let payloadLength = dec.readVarInt();
        let payload = dec.read(payloadLength);
        let nTransaction = new Transaction(nexusName, chainName, script, date, payload);
        let signatureCount = dec.readVarInt();
        /*for (let i = 0; i < signatureCount; i++) {
          nTransaction.signatures.push(dec.readSignature());
        }*/
        return nTransaction;
    }
    static Unserialize(serialized) {
        let reader = new index_js_3.PBinaryReader(serialized);
        let tx = new Transaction('', '', '', new Date(), '');
        tx.UnserializeData(reader);
        return tx;
    }
}
exports.Transaction = Transaction;
