// phantasma-sdk-ts/dist/cjs/core/tx/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyData = exports.signData = exports.getWifFromPrivateKey = exports.generateNewWif = exports.generateNewSeedWords = exports.generateNewSeed = exports.getPublicKeyFromPrivateKey = exports.getAddressFromWif = exports.getPrivateKeyFromWif = void 0;
const wif_1 = __importDefault(require("wif"));
const elliptic_1 = __importDefault(require("elliptic"));
const bs58_1 = __importDefault(require("bs58"));
const bip39 = __importStar(require("bip39"));
const crypto_1 = __importDefault(require("crypto"));
const { eddsa } = elliptic_1.default;
const curve = new eddsa('ed25519');
function ab2hexstring(arr) {
    if (typeof arr !== 'object') {
        throw new Error(`ab2hexstring expects an array.Input was ${arr}`);
    }
    let result = '';
    const intArray = new Uint8Array(arr);
    for (const i of intArray) {
        let str = i.toString(16);
        str = str.length === 0 ? '00' : str.length === 1 ? '0' + str : str;
        result += str;
    }
    return result;
}
function getPrivateKeyFromWif(wif) {
    return ab2hexstring(wif_1.default.decode(wif, 128).privateKey);
}
exports.getPrivateKeyFromWif = getPrivateKeyFromWif;
function getAddressFromWif(wif) {
    const curve = new eddsa('ed25519');
    const privateKey = getPrivateKeyFromWif(wif);
    const privateKeyBuffer = Buffer.from(privateKey, 'hex');
    const publicKey = curve.keyFromSecret(privateKeyBuffer).getPublic('hex');
    var addressHex = Buffer.from('0100' + publicKey, 'hex');
    return 'P' + bs58_1.default.encode(addressHex);
}
exports.getAddressFromWif = getAddressFromWif;
function getPublicKeyFromPrivateKey(privateKey) {
    const privateKeyBuffer = Buffer.from(privateKey, 'hex');
    const publicKey = curve.keyFromSecret(privateKeyBuffer).getPublic('hex');
    return publicKey;
}
exports.getPublicKeyFromPrivateKey = getPublicKeyFromPrivateKey;
function generateNewSeed() {
    let buffer = new Uint8Array(32);
    let privateKey = Buffer.alloc(32);
    crypto_1.default.getRandomValues(buffer);
    for (let i = 0; i < 32; ++i) {
        privateKey.writeUInt8(buffer[i], i);
    }
    const wif = wif_1.default.encode(128, privateKey, true);
    const mnemonic = bip39.generateMnemonic();
    return mnemonic;
}
exports.generateNewSeed = generateNewSeed;
function generateNewSeedWords() {
    let buffer = new Uint8Array(32);
    let privateKey = Buffer.alloc(32);
    crypto_1.default.getRandomValues(buffer);
    for (let i = 0; i < 32; ++i) {
        privateKey.writeUInt8(buffer[i], i);
    }
    const wif = wif_1.default.encode(128, privateKey, true);
    const mnemonic = bip39.generateMnemonic();
    const seedWords = mnemonic.split(' ');
    return seedWords;
}
exports.generateNewSeedWords = generateNewSeedWords;
function generateNewWif() {
    let buffer = new Uint8Array(32);
    let privateKey = Buffer.alloc(32);
    crypto_1.default.getRandomValues(buffer);
    for (let i = 0; i < 32; ++i) {
        privateKey.writeUInt8(buffer[i], i);
    }
    const wif = wif_1.default.encode(128, privateKey, true);
    return wif;
}
exports.generateNewWif = generateNewWif;
function getWifFromPrivateKey(privateKey) {
    const privateKeyBuffer = Buffer.from(privateKey, 'hex');
    const wif = wif_1.default.encode(128, privateKeyBuffer, true);
    return wif;
}
exports.getWifFromPrivateKey = getWifFromPrivateKey;
function signData(msgHex, privateKey) {
    const msgHashHex = Buffer.from(msgHex, 'hex');
    const privateKeyBuffer = Buffer.from(privateKey, 'hex');
    const sig = curve.sign(msgHashHex, privateKeyBuffer);
    const numBytes = sig.toBytes().length;
    return '01' + (numBytes < 16 ? '0' : '') + numBytes.toString(16) + sig.toHex();
}
exports.signData = signData;
function verifyData(msgHex, phaSig, address) {
    const msgBytes = Buffer.from(msgHex, 'hex');
    const realSig = phaSig.substring(4);
    const pubKey = bs58_1.default.decode(address.substring(1)).slice(2);
    return curve.verify(msgBytes, realSig, ab2hexstring(pubKey));
}
exports.verifyData = verifyData;
