// phantasma-sdk-ts/dist/cjs/core/types/Address.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Address = exports.AddressKind = void 0;
const bs58_1 = __importDefault(require("bs58"));
const index_js_1 = require("../utils/index.js");
const sha256_js_1 = __importDefault(require("crypto-js/sha256.js"));
const enc_hex_js_1 = __importDefault(require("crypto-js/enc-hex.js"));
const index_js_2 = require("../tx/index.js");
const elliptic_1 = __importDefault(require("elliptic"));
const { eddsa } = elliptic_1.default;
const curve = new eddsa('ed25519');
var AddressKind;
(function (AddressKind) {
    AddressKind[AddressKind["Invalid"] = 0] = "Invalid";
    AddressKind[AddressKind["User"] = 1] = "User";
    AddressKind[AddressKind["System"] = 2] = "System";
    AddressKind[AddressKind["Interop"] = 3] = "Interop";
})(AddressKind = exports.AddressKind || (exports.AddressKind = {}));
class Address {
    /*public get Kind(): AddressKind {
          return this.IsNull ? AddressKind.System : (this._bytes[0] >= 3) ? AddressKind.Interop
              : (AddressKind)this._bytes[0];
      }*/
    get Kind() {
        if (this.IsNull) {
            return AddressKind.System;
        }
        else if (this._bytes[0] >= 3) {
            return AddressKind.Interop;
        }
        else {
            return this._bytes[0];
        }
    }
    get IsSystem() {
        return this.Kind == AddressKind.System;
    }
    get IsInterop() {
        return this.Kind == AddressKind.Interop;
    }
    get IsUser() {
        return this.Kind == AddressKind.User;
    }
    /*public get TendermintAddress(): string {
          return encodeBase16(this._bytes.slice(2).SHA256().slice(0, 20));
      }
      public get TendermintAddress() {
          return SHA256(this._bytes.slice(2)).slice(0, 20).toString('hex');
      }
      */
    get IsNull() {
        if (this._bytes == null || this._bytes.length == 0) {
            return true;
        }
        for (let i = 1; i < this._bytes.length; i++) {
            if (this._bytes[i] != 0) {
                return false;
            }
        }
        return true;
    }
    get Text() {
        if (this.IsNull) {
            return Address.NullText;
        }
        if (!this._text) {
            if (Address._keyToTextCache.has(this._bytes)) {
                this._text = Address._keyToTextCache.get(this._bytes);
            }
            if (!this._text) {
                let prefix;
                switch (this.Kind) {
                    case AddressKind.User:
                        prefix = 'P';
                        break;
                    case AddressKind.Interop:
                        prefix = 'X';
                        break;
                    default:
                        prefix = 'S';
                        break;
                }
                this._text = prefix + bs58_1.default.encode(this._bytes);
                Address._keyToTextCache.set(this._bytes, this._text);
            }
        }
        return this._text;
    }
    constructor(publicKey) {
        let pkFromArray = Array.from(publicKey);
        if (publicKey.length != Address.LengthInBytes) {
            throw new Error(`publicKey length must be ${Address.LengthInBytes}, it was ${publicKey.length}}`);
        }
        this._bytes = new Uint8Array(Address.LengthInBytes);
        this._bytes.set(publicKey);
        this._text = null;
    }
    static FromPublickKey(publicKey) {
        publicKey = publicKey.slice(0, 34);
        return new Address(publicKey);
    }
    static FromText(text) {
        return Address.Parse(text);
    }
    static Parse(text) {
        if (text == null) {
            return Address.Null;
        }
        if (text == Address.NullText) {
            return Address.Null;
        }
        let addr;
        let originalText = text;
        let prefix = text[0];
        text = text.slice(1);
        var bytes = bs58_1.default.decode(text);
        addr = new Address(bytes);
        switch (prefix) {
            case 'P':
                if (addr.Kind != AddressKind.User) {
                    throw new Error(`Invalid address prefix. Expected 'P', got '${prefix}'`);
                }
                break;
            case 'S':
                if (addr.Kind != AddressKind.System) {
                    throw new Error(`Invalid address prefix. Expected 'S', got '${prefix}'`);
                }
                break;
            case 'X':
                if (addr.Kind < AddressKind.Interop) {
                    throw new Error(`Invalid address prefix. Expected 'X', got '${prefix}'`);
                }
                break;
            default:
                throw new Error(`Invalid address prefix. Expected 'P', 'S' or 'X', got '${prefix}'`);
        }
        /*this._keyToTextCache.values().forEach((value) => {
          if (value == text) {
            return Address.FromHash(this._keyToTextCache(value));
          }*/
        return addr;
    }
    static IsValidAddress(text) {
        try {
            Address.FromText(text);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    static FromBytes(bytes) {
        return new Address(bytes);
    }
    static FromKey(key) {
        const bytes = new Uint8Array(Address.LengthInBytes);
        bytes[0] = AddressKind.User;
        if (key.PublicKey.length == 32) {
            bytes.set(key.PublicKey, 2);
        }
        else if (key.PublicKey.length == 33) {
            bytes.set(key.PublicKey, 1);
        }
        else if (key.PublicKey.length == 64) {
            bytes.set(key.PublicKey.slice(0, 32), 1);
        }
        else {
            throw new Error('Invalid public key length: ' + key.PublicKey.length);
        }
        return new Address(bytes);
    }
    static FromHash(input) {
        let bytes;
        if (typeof input === 'string') {
            bytes = new TextEncoder().encode(input);
        }
        else {
            bytes = input;
        }
        const hash = (0, sha256_js_1.default)(enc_hex_js_1.default.parse((0, index_js_1.uint8ArrayToString)(bytes)));
        bytes = new Uint8Array(Address.LengthInBytes);
        bytes[0] = AddressKind.User;
        bytes.set(hash.words.slice(0, 32), 2);
        return new Address(bytes);
    }
    static FromWif(wif) {
        const privateKey = (0, index_js_2.getPrivateKeyFromWif)(wif);
        const publicKey = (0, index_js_2.getPublicKeyFromPrivateKey)(privateKey);
        var addressHex = Buffer.from('0100' + publicKey, 'hex');
        return this.FromBytes(addressHex);
    }
    compareTo(other) {
        for (let i = 0; i < Address.LengthInBytes; i++) {
            if (this._bytes[i] < other._bytes[i]) {
                return -1;
            }
            else if (this._bytes[i] > other._bytes[i]) {
                return 1;
            }
        }
        return 0;
    }
    equals(other) {
        if (!(other instanceof Address)) {
            return false;
        }
        const address = other;
        return this._bytes.toString() === address._bytes.toString();
    }
    toString() {
        if (this.IsNull) {
            return Address.NullText;
        }
        if (!this._text) {
            let prefix;
            switch (this.Kind) {
                case AddressKind.User:
                    prefix = 'P';
                    break;
                case AddressKind.Interop:
                    prefix = 'X';
                    break;
                default:
                    prefix = 'S';
                    break;
            }
            this._text = prefix + bs58_1.default.encode(this._bytes);
        }
        return this._text;
    }
    GetPublicKey() {
        if (!this._bytes || this._bytes.length !== Address.LengthInBytes) {
            throw new Error('invalid address byte length');
        }
        return this._bytes.slice(2, 34);
    }
    ToByteArray() {
        return this._bytes;
    }
    SerializeData(writer) {
        writer.writeByteArray(this._bytes);
    }
    UnserializeData(reader) {
        this._bytes = reader.readByteArray();
        this._text = null;
    }
}
exports.Address = Address;
Address.NullText = 'NULL';
Address.LengthInBytes = 34;
Address.MaxPlatformNameLength = 10;
Address.NullPublicKey = new Uint8Array(Address.LengthInBytes);
Address.Null = new Address(Address.NullPublicKey);
Address._keyToTextCache = new Map();
