// phantasma-sdk-ts/dist/cjs/core/types/Consensus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PollPresence = exports.ConsensusPoll = exports.PollVote = exports.PollValue = exports.PollChoice = exports.PollState = exports.ConsensusMode = void 0;
const index_js_1 = require("../utils/index.js");
const index_js_2 = require("./Extensions/index.js");
const Timestamp_js_1 = require("./Timestamp.js");
var ConsensusMode;
(function (ConsensusMode) {
    ConsensusMode[ConsensusMode["Unanimity"] = 0] = "Unanimity";
    ConsensusMode[ConsensusMode["Majority"] = 1] = "Majority";
    ConsensusMode[ConsensusMode["Popularity"] = 2] = "Popularity";
    ConsensusMode[ConsensusMode["Ranking"] = 3] = "Ranking";
})(ConsensusMode = exports.ConsensusMode || (exports.ConsensusMode = {}));
var PollState;
(function (PollState) {
    PollState[PollState["Inactive"] = 0] = "Inactive";
    PollState[PollState["Active"] = 1] = "Active";
    PollState[PollState["Consensus"] = 2] = "Consensus";
    PollState[PollState["Failure"] = 3] = "Failure";
})(PollState = exports.PollState || (exports.PollState = {}));
class PollChoice {
    constructor(value) {
        if (value instanceof Array)
            this.value = index_js_2.Base16.decode((0, index_js_1.bytesToHex)(value));
        else
            this.value = value;
    }
    SerializeData(writer) {
        writer.writeByteArray((0, index_js_1.stringToUint8Array)(this.value));
    }
    UnserializeData(reader) {
        this.value = index_js_2.Base16.decode(reader.readByteArray());
    }
    static Unserialize(reader) {
        let pollChoice = new PollChoice('');
        pollChoice.UnserializeData(reader);
        return pollChoice;
    }
}
exports.PollChoice = PollChoice;
class PollValue {
    SerializeData(writer) {
        writer.writeByteArray((0, index_js_1.stringToUint8Array)(this.value));
        writer.writeBigInteger(this.ranking);
        writer.writeBigInteger(this.votes);
    }
    UnserializeData(reader) {
        this.value = index_js_2.Base16.decode(reader.readByteArray());
        this.ranking = reader.readBigInteger();
        this.votes = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let pollValue = new PollValue();
        pollValue.UnserializeData(reader);
        return pollValue;
    }
}
exports.PollValue = PollValue;
class PollVote {
    SerializeData(writer) {
        writer.writeBigInteger(this.index);
        writer.writeBigInteger(this.percentage);
    }
    UnserializeData(reader) {
        this.index = reader.readBigInteger();
        this.percentage = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let pollVote = new PollVote();
        pollVote.UnserializeData(reader);
        return pollVote;
    }
}
exports.PollVote = PollVote;
class ConsensusPoll {
    constructor() {
        this.subject = '';
        this.organization = '';
        this.mode = ConsensusMode.Unanimity;
        this.state = PollState.Inactive;
        this.entries = [];
        this.round = BigInt(0);
        this.startTime = Timestamp_js_1.Timestamp.null;
        this.endTime = Timestamp_js_1.Timestamp.null;
        this.choicesPerUser = BigInt(0);
        this.totalVotes = BigInt(0);
    }
    SerializeData(writer) {
        writer.writeString(this.subject);
        writer.writeString(this.organization);
        writer.writeByte(this.mode);
        writer.writeByte(this.state);
        writer.writeByte(this.entries.length);
        this.entries.forEach((entry) => {
            entry.SerializeData(writer);
        });
        writer.writeBigInteger(this.round);
        writer.writeTimestamp(this.startTime);
        writer.writeTimestamp(this.endTime);
        writer.writeBigInteger(this.choicesPerUser);
        writer.writeBigInteger(this.totalVotes);
    }
    UnserializeData(reader) {
        this.subject = reader.readString();
        this.organization = reader.readString();
        this.mode = reader.readByte();
        this.state = reader.readByte();
        this.entries = [];
        const entriesLength = reader.readUnsignedInt();
        for (let i = 0; i < entriesLength; i++) {
            this.entries.push(PollValue.Unserialize(reader));
        }
        this.round = reader.readBigInteger();
        this.startTime = reader.readTimestamp();
        this.endTime = reader.readTimestamp();
        this.choicesPerUser = reader.readBigInteger();
        this.totalVotes = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let consensusPoll = new ConsensusPoll();
        consensusPoll.UnserializeData(reader);
        return consensusPoll;
    }
}
exports.ConsensusPoll = ConsensusPoll;
class PollPresence {
    SerializeData(writer) {
        writer.writeString(this.subject);
    }
    UnserializeData(reader) {
        this.subject = reader.readString();
        this.round = reader.readBigInteger();
    }
    static Unserialize(reader) {
        let pollPresence = new PollPresence();
        pollPresence.UnserializeData(reader);
        return pollPresence;
    }
}
exports.PollPresence = PollPresence;
