// phantasma-sdk-ts/dist/cjs/core/types/Contract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractEvent = exports.ContractMethod = exports.ContractInterface = exports.ContractParameter = void 0;
const index_js_1 = require("../utils/index.js");
const index_js_2 = require("./Extensions/index.js");
class ContractParameter {
    constructor(name, type) {
        this.name = name;
        this.type = type;
    }
}
exports.ContractParameter = ContractParameter;
class ContractInterface {
    Events() {
        return this._events;
    }
    EventCount() {
        return this._events.length;
    }
    newEmpty() {
        this._methods = new Map();
        this._events = [];
    }
    constructor(methods, events) {
        this._methods = new Map();
        this.Methods = Array.from(this._methods.values());
        this.MethodCount = this._methods.size;
        for (const entry of methods) {
            this._methods.set(entry.name, entry);
        }
        this._events = events;
    }
    get(name) {
        return this.FindMethod(name);
    }
    HasMethod(name) {
        return this._methods.has(name);
    }
    HasTokenTrigger(trigger) {
        const strName = trigger.toString();
        const name = strName[0].toLowerCase() + strName.slice(1);
        return this._methods.has(name);
    }
    FindMethod(name) {
        if (this._methods.has(name)) {
            return this._methods.get(name);
        }
        return null;
    }
    FindEvent(value) {
        for (const evt of this._events) {
            if (evt.value === value) {
                return evt;
            }
        }
        return null;
    }
    ImplementsEvent(evt) {
        for (const entry of this.Events()) {
            if (entry.name === evt.name &&
                entry.value === evt.value &&
                entry.returnType === evt.returnType) {
                return true;
            }
        }
        return false;
    }
    ImplementsMethod(method) {
        if (!this._methods.has(method.name)) {
            return false;
        }
        const thisMethod = this._methods.get(method.name);
        if (thisMethod.parameters.length !== method.parameters.length) {
            return false;
        }
        for (let i = 0; i < method.parameters.length; i++) {
            if (thisMethod.parameters[i].type !== method.parameters[i].type) {
                return false;
            }
        }
        return true;
    }
    ImplementsInterface(other) {
        for (const method of other.Methods) {
            if (!this.ImplementsMethod(method)) {
                return false;
            }
        }
        for (const evt of other.Events()) {
            if (!this.ImplementsEvent(evt)) {
                return false;
            }
        }
        return true;
    }
    UnserializeData(reader) {
        const len = reader.readByte();
        this._methods.clear();
        for (let i = 0; i < len; i++) {
            const method = ContractMethod.Unserialize(reader);
            this._methods.set(method.name, method);
        }
        const eventLen = reader.readByte();
        this._events = [];
        for (let i = 0; i < eventLen; i++) {
            this._events.push(ContractEvent.Unserialize(reader));
        }
    }
    SerializeData(writer) {
        writer.writeByte(this._methods.size);
        for (const [_, value] of this._methods) {
            value.Serialize(writer);
        }
        writer.writeByte(this._events.length);
        for (const entry of this._events) {
            entry.Serialize(writer);
        }
    }
}
exports.ContractInterface = ContractInterface;
ContractInterface.Empty = new ContractInterface([], []);
class ContractMethod {
    SerializeData(writer) {
        this.Serialize(writer);
    }
    UnserializeData(reader) {
        return ContractMethod.Unserialize(reader);
    }
    constructorOne(name, returnType, labels, parameters) {
        if (!labels.has(name)) {
            throw new Error(`Missing offset in label map for method ${name}`);
        }
        const offset = labels.get(name);
        this.name = name;
        this.offset = offset;
        this.returnType = returnType;
        this.parameters = parameters;
    }
    constructor(name, returnType, offset, parameters) {
        this.name = name;
        this.offset = offset;
        this.returnType = returnType;
        this.parameters = parameters;
    }
    isProperty() {
        if (this.name.length >= 4 &&
            this.name.startsWith('get') &&
            this.name[3] === this.name[3].toUpperCase()) {
            return true;
        }
        if (this.name.length >= 3 &&
            this.name.startsWith('is') &&
            this.name[2] === this.name[2].toUpperCase()) {
            return true;
        }
        return false;
    }
    isTrigger() {
        if (this.name.length >= 3 &&
            this.name.startsWith('on') &&
            this.name[2] === this.name[2].toUpperCase()) {
            return true;
        }
        return false;
    }
    toString() {
        return `${this.name} : ${this.returnType}`;
    }
    static fromBytes(bytes) {
        const stream = new Uint8Array(bytes);
        const reader = new index_js_2.PBinaryReader(stream);
        return ContractMethod.Unserialize(reader);
    }
    static Unserialize(reader) {
        const name = reader.readString();
        const returnType = reader.readByte();
        const offset = reader.readInt();
        const len = reader.readByte();
        const parameters = new Array(len);
        for (let i = 0; i < len; i++) {
            const pName = reader.readString();
            const pVMType = reader.readByte();
            parameters[i] = new ContractParameter(pName, pVMType);
        }
        return new ContractMethod(name, returnType, offset, parameters);
    }
    Serialize(writer) {
        writer.writeString(this.name);
        writer.writeByte(this.returnType);
        writer.writeInt(this.offset);
        writer.writeByte(this.parameters.length);
        this.parameters.forEach((entry) => {
            writer.writeString(entry.name);
            writer.writeByte(entry.type);
        });
    }
    toArray() {
        const stream = new Uint8Array();
        const writer = new index_js_2.PBinaryWriter(stream);
        this.Serialize(writer);
        return stream;
    }
}
exports.ContractMethod = ContractMethod;
class ContractEvent {
    constructor(value, name, returnType, description) {
        this.value = value;
        this.name = name;
        this.returnType = returnType;
        this.description = description;
    }
    SerializeData(writer) {
        this.Serialize(writer);
    }
    UnserializeData(reader) {
        return ContractEvent.Unserialize(reader);
    }
    toString() {
        return `${this.name} : ${this.returnType} => ${this.value}`;
    }
    static Unserialize(reader) {
        const value = reader.readByte();
        const name = reader.readString();
        const returnType = reader.readByte();
        const description = reader.readBytes(reader.readByte());
        return new ContractEvent(value, name, returnType, (0, index_js_1.arrayNumberToUint8Array)(description));
    }
    Serialize(writer) {
        writer.writeByte(this.value);
        writer.writeString(this.name);
        writer.writeByte(this.returnType);
        writer.writeByte(this.description.length);
        writer.writeBytes((0, index_js_1.uint8ArrayToNumberArray)(this.description));
    }
}
exports.ContractEvent = ContractEvent;
