// phantasma-sdk-ts/dist/cjs/core/types/Ed25519Signature.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ed25519Signature = void 0;
const Signature_js_1 = require("../interfaces/Signature.js");
const elliptic_1 = __importDefault(require("elliptic"));
const index_js_1 = require("../utils/index.js");
const index_js_2 = require("./Extensions/index.js");
const { eddsa } = elliptic_1.default;
const ed25519 = new eddsa('ed25519');
class Ed25519Signature {
    constructor(bytes) {
        this.Kind = Signature_js_1.SignatureKind.Ed25519;
        this.Bytes = bytes;
    }
    Verify(message, address) {
        return this.VerifyMultiple(message, [address]);
    }
    VerifyMultiple(message, addresses) {
        for (const address of addresses) {
            if (!address.IsUser) {
                continue;
            }
            const pubKey = address.ToByteArray().slice(2);
            const msgBytes = Buffer.from(message);
            const sigHex = (0, index_js_1.bytesToHex)(this.Bytes);
            const pubKeyHex = (0, index_js_1.bytesToHex)(pubKey);
            if (ed25519.verify(msgBytes, sigHex, pubKeyHex)) {
                return true;
            }
        }
        return false;
    }
    SerializeData(writer) {
        //writer.writeString(uint8ArrayToString(this.Bytes));
        writer.writeByteArray(this.Bytes);
    }
    UnserializeData(reader) {
        this.Bytes = (0, index_js_1.stringToUint8Array)(reader.readString());
    }
    ToByteArray() {
        const stream = new Uint8Array(64);
        const writer = new index_js_2.PBinaryWriter(stream);
        this.SerializeData(writer);
        return new Uint8Array(stream);
    }
    static Generate(keypair, message) {
        const msgHashHex = Buffer.from((0, index_js_1.bytesToHex)(message), 'hex');
        //const msgHashHex = uint8ArrayToString(message);
        const privateKeyBuffer = Buffer.from((0, index_js_1.bytesToHex)(keypair.PrivateKey), 'hex');
        //const privateKeyBuffer = uint8ArrayToString(keypair.PrivateKey);
        const sign = ed25519.sign(msgHashHex, privateKeyBuffer);
        return new Ed25519Signature(sign.toBytes());
    }
}
exports.Ed25519Signature = Ed25519Signature;
