// phantasma-sdk-ts/dist/cjs/core/types/PhantasmaKeys.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhantasmaKeys = void 0;
const Address_js_1 = require("./Address.js");
const bs58_1 = __importDefault(require("bs58"));
const wif_1 = __importDefault(require("wif"));
const index_js_1 = require("../utils/index.js");
const Ed25519Signature_js_1 = require("./Ed25519Signature.js");
const elliptic_1 = __importDefault(require("elliptic"));
const Entropy_js_1 = require("./Entropy.js");
const index_js_2 = require("../tx/index.js");
const { eddsa } = elliptic_1.default;
const ed25519 = new eddsa('ed25519');
class PhantasmaKeys {
    get PrivateKey() {
        return this._privateKey;
    }
    get PublicKey() {
        return this._publicKey;
    }
    constructor(privateKey) {
        if (privateKey.length == 64) {
            privateKey = privateKey.slice(0, 32);
        }
        if (privateKey.length != PhantasmaKeys.PrivateKeyLength) {
            throw new Error(`privateKey should have length ${PhantasmaKeys.PrivateKeyLength} but has ${privateKey.length}`);
        }
        this._privateKey = new Uint8Array(PhantasmaKeys.PrivateKeyLength);
        this._privateKey.set(privateKey);
        let wif = (0, index_js_2.getWifFromPrivateKey)((0, index_js_1.bytesToHex)(privateKey));
        const privateKeyString = (0, index_js_1.bytesToHex)(this._privateKey);
        const privateKeyBuffer = Buffer.from(privateKeyString, 'hex');
        const publicKey = ed25519.keyFromSecret(privateKeyBuffer).getPublic();
        this._publicKey = publicKey;
        this.Address = Address_js_1.Address.FromKey(this);
    }
    toString() {
        return this.Address.Text;
    }
    static generate() {
        const privateKey = Entropy_js_1.Entropy.GetRandomBytes(PhantasmaKeys.PrivateKeyLength);
        const pair = new PhantasmaKeys(privateKey);
        return pair;
    }
    static fromWIF(wif) {
        if (!wif) {
            throw new Error('WIF required');
        }
        let data = bs58_1.default.decode(wif); // checkdecode
        if (data.length == 38) {
            data = data.slice(0, 34);
        }
        if (data.length != 34 || data[0] != 0x80 || data[33] != 0x01) {
            throw new Error('Invalid WIF format');
        }
        const privateKey = data.slice(1, 33);
        return new PhantasmaKeys(privateKey);
    }
    toWIF() {
        const privateKeyString = (0, index_js_1.bytesToHex)(this._privateKey);
        const privatekeyBuffer = Buffer.from(privateKeyString, 'hex');
        const wif = wif_1.default.encode(128, privatekeyBuffer, true); //uint8ArrayToHex(data); // .base58CheckEncode();
        return wif;
    }
    static xor(x, y) {
        if (x.length != y.length) {
            throw new Error('x and y should have the same length');
        }
        const result = new Uint8Array(x.length);
        for (let i = 0; i < x.length; i++) {
            result[i] = x[i] ^ y[i];
        }
        return result;
    }
    Sign(msg, customSignFunction) {
        return Ed25519Signature_js_1.Ed25519Signature.Generate(this, msg);
    }
}
exports.PhantasmaKeys = PhantasmaKeys;
PhantasmaKeys.PrivateKeyLength = 32;
