// phantasma-sdk-ts/dist/cjs/core/vm/Decoder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Decoder = void 0;
const big_integer_1 = __importDefault(require("big-integer"));
const index_js_1 = require("../interfaces/index.js");
const VMType_js_1 = require("./VMType.js");
const logger_js_1 = require("../utils/logger.js");
class Decoder {
    constructor(str) {
        this.str = str;
    }
    isEnd() {
        return this.str.length == 0;
    }
    readCharPair() {
        var res = this.str.substr(0, 2);
        this.str = this.str.slice(2);
        return res;
    }
    readByte() {
        return parseInt(this.readCharPair(), 16);
    }
    read(numBytes) {
        var res = this.str.substr(0, numBytes * 2);
        this.str = this.str.slice(numBytes * 2);
        return res;
    }
    readString() {
        var len = this.readVarInt();
        return this.readStringBytes(len);
    }
    readStringBytes(numBytes) {
        var res = "";
        for (var i = 0; i < numBytes; ++i) {
            res += String.fromCharCode(this.readByte());
        }
        return res;
    }
    readByteArray() {
        var res;
        var length = this.readVarInt();
        if (length == 0)
            return [];
        res = this.read(length);
        return res;
    }
    readSignature() {
        let kind = this.readByte();
        let signature = new index_js_1.ISignature();
        let curve;
        signature.kind = kind;
        switch (kind) {
            case index_js_1.SignatureKind.None:
                return null;
            case index_js_1.SignatureKind.Ed25519:
                let len = this.readVarInt();
                signature.signature = this.read(len);
                break;
            case index_js_1.SignatureKind.ECDSA:
                curve = this.readByte();
                signature.signature = this.readString();
                break;
            default:
                throw "read signature: " + kind;
        }
        return signature;
    }
    readTimestamp() {
        //var len = this.readByte();
        let result = 0;
        let bytes = this.read(4);
        bytes
            .match(/.{1,2}/g)
            .reverse()
            .forEach((c) => (result = result * 256 + parseInt(c, 16)));
        return result;
    }
    readVarInt() {
        var len = this.readByte();
        var res = 0;
        if (len === 0xfd) {
            [...this.read(2).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        else if (len === 0xfe) {
            [...this.read(4).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        else if (len === 0xff) {
            [...this.read(8).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        return len;
    }
    readBigInt() {
        // TO DO: implement negative numbers
        var len = this.readVarInt();
        var res = 0;
        var stringBytes = this.read(len);
        [...stringBytes.match(/.{1,2}/g)]
            .reverse()
            .forEach((c) => (res = res * 256 + parseInt(c, 16)));
        return res;
    }
    readBigIntAccurate() {
        var len = this.readVarInt();
        var res = (0, big_integer_1.default)();
        var stringBytes = this.read(len);
        [...stringBytes.match(/.{1,2}/g)].reverse().forEach((c) => {
            res = res.times(256).plus(parseInt(c, 16));
        });
        return res.toString();
    }
    readVmObject() {
        const type = this.readByte();
        logger_js_1.logger.log("type", type);
        switch (type) {
            case VMType_js_1.VMType.String:
                return this.readString();
            case VMType_js_1.VMType.Number:
                return this.readBigIntAccurate();
            case VMType_js_1.VMType.Bool:
                return this.readByte() != 0;
            case VMType_js_1.VMType.Struct:
                const numFields = this.readVarInt();
                let res = {};
                for (let i = 0; i < numFields; ++i) {
                    const key = this.readVmObject();
                    logger_js_1.logger.log("  key", key);
                    const value = this.readVmObject();
                    logger_js_1.logger.log("  value", value);
                    res[key] = value;
                }
                return res;
            case VMType_js_1.VMType.Enum:
                return this.readVarInt();
            case VMType_js_1.VMType.Object:
                const numBytes = this.readVarInt();
                return this.read(numBytes);
            default:
                return "unsupported type " + type;
        }
    }
}
exports.Decoder = Decoder;
