// phantasma-sdk-ts/dist/cjs/core/vm/ScriptBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScriptBuilder = void 0;
const bs58_1 = __importDefault(require("bs58"));
const index_js_1 = require("../interfaces/index.js");
const CarbonSerialization_js_1 = require("../types/CarbonSerialization.js");
const index_js_2 = require("../types/index.js");
const index_js_3 = require("../utils/index.js");
const Opcode_js_1 = require("./Opcode.js");
const VMObject_js_1 = require("./VMObject.js");
const VMType_js_1 = require("./VMType.js");
const Contracts_js_1 = require("./Contracts.js");
const MaxRegisterCount = 32;
class ScriptBuilder {
    static ScriptBuilder() {
        return new ScriptBuilder();
    }
    constructor() {
        this._labelLocations = {};
        this._jumpLocations = {};
        this.NullAddress = 'S1111111111111111111111111111111111';
        this.str = '';
        this.writer = new index_js_2.PBinaryWriter();
    }
    BeginScript() {
        this.str = '';
        this.writer = new index_js_2.PBinaryWriter();
        return this;
    }
    GetScript() {
        return (0, index_js_3.bytesToHex)(this.writer.toUint8Array());
    }
    EndScript() {
        this.Emit(Opcode_js_1.Opcode.RET);
        return (0, index_js_3.bytesToHex)(this.writer.toUint8Array()).toUpperCase();
    }
    Emit(opcode, bytes) {
        this.AppendByte(opcode);
        if (bytes) {
            this.EmitBytes(bytes);
        }
        return this;
    }
    EmitThorw(reg) {
        this.Emit(Opcode_js_1.Opcode.THROW);
        this.AppendByte(reg);
        return this;
    }
    EmitPush(reg) {
        this.Emit(Opcode_js_1.Opcode.PUSH);
        this.AppendByte(reg);
        return this;
    }
    EmitPop(reg) {
        this.Emit(Opcode_js_1.Opcode.POP);
        this.AppendByte(reg);
        return this;
    }
    EmitExtCall(method, reg = 0) {
        this.EmitLoad(reg, method);
        this.Emit(Opcode_js_1.Opcode.EXTCALL);
        this.AppendByte(reg);
        return this;
    }
    EmitBigInteger(value) {
        let bytes = [];
        if (value == '0') {
            bytes = [0];
        }
        else if (value.startsWith('-1')) {
            throw new Error('Unsigned bigint serialization not suppoted');
        }
        else {
            let hex = BigInt(value).toString(16);
            if (hex.length % 2)
                hex = '0' + hex;
            const len = hex.length / 2;
            var i = 0;
            var j = 0;
            while (i < len) {
                bytes.unshift(parseInt(hex.slice(j, j + 2), 16)); // little endian
                i += 1;
                j += 2;
            }
            bytes.push(0); // add sign at the end
        }
        return this.EmitByteArray(bytes);
    }
    EmitAddress(textAddress) {
        const bytes = [...bs58_1.default.decode(textAddress.substring(1))];
        return this.EmitByteArray(bytes);
    }
    RawString(value) {
        //let bytes = stringToUint8Array(value);
        //console.log(Array.from(bytes))
        //return Array.from(bytes);
        var data = [];
        for (var i = 0; i < value.length; i++) {
            data.push(value.charCodeAt(i));
        }
        return data;
    }
    EmitLoadBigInt(reg, value) {
        const bytes = Array.from((0, CarbonSerialization_js_1.bigIntToTwosComplementLE)(value));
        this.EmitLoadBytes(reg, bytes, VMType_js_1.VMType.Number);
        return this;
    }
    EmitLoad(reg, obj) {
        let structType = Object.getPrototypeOf(obj).constructor.name;
        switch (typeof obj) {
            case 'string': {
                let bytes = this.RawString(obj);
                this.EmitLoadBytes(reg, bytes, VMType_js_1.VMType.String);
                break;
            }
            case 'boolean': {
                let bytes = [obj ? 1 : 0];
                this.EmitLoadBytes(reg, bytes, VMType_js_1.VMType.Bool);
                break;
            }
            case 'bigint': {
                this.EmitLoadBigInt(reg, obj);
                break;
            }
            case 'number': {
                // obj is BigInteger
                // var bytes = val.ToSignedByteArray();
                // this.emitLoadBytes(reg, bytes, VMType.Number);
                //let bytes = this.RawString(BigInt(obj).toString());
                if (Object.getPrototypeOf(structType).constructor.name == 'enum') {
                    this.AppendByte(obj);
                }
                else {
                    // this.EmitLoadVarInt(reg, obj);
                    let bytes = this.RawString(obj.toString());
                    this.EmitLoadBytes(reg, bytes, VMType_js_1.VMType.String);
                }
                break;
            }
            case 'object':
                if (obj instanceof Uint8Array) {
                    this.EmitLoadBytes(reg, Array.from(obj));
                }
                else if (obj instanceof VMObject_js_1.VMObject) {
                    this.EmitLoadVMObject(reg, obj);
                }
                else if (Array.isArray(obj)) {
                    this.EmitLoadArray(reg, obj);
                }
                else if (obj instanceof Date || obj instanceof index_js_2.Timestamp) {
                    this.EmitLoadTimestamp(reg, obj);
                }
                else if (obj instanceof index_js_2.Address) {
                    this.EmitLoadAddress(reg, obj);
                }
                else if ((typeof obj.UnserializeData === 'function' && typeof obj.SerializeData === 'function') ||
                    obj instanceof index_js_1.ISerializable) {
                    this.EmitLoadISerializable(reg, obj);
                }
                else {
                    if (Array.isArray(obj)) {
                        this.EmitLoadArray(reg, obj);
                    }
                    else {
                        throw Error('Load type ' + typeof obj + ' not supported');
                    }
                }
                break;
            default:
                throw Error('Load type ' + typeof obj + ' not supported');
        }
        return this;
    }
    EmitLoadBytes(reg, bytes, type = VMType_js_1.VMType.Bytes) {
        if (bytes.length > 0xffff)
            throw new Error('tried to load too much data');
        this.Emit(Opcode_js_1.Opcode.LOAD);
        this.AppendByte(reg);
        this.AppendByte(type);
        this.EmitVarInt(bytes.length);
        this.EmitBytes(bytes);
        return this;
    }
    EmitLoadArray(reg, obj) {
        this.Emit(Opcode_js_1.Opcode.CAST, [reg, reg, VMType_js_1.VMType.None]);
        for (let i = 0; i < obj.length; i++) {
            let element = obj[i];
            let temp_regVal = reg + 1;
            let temp_regKey = reg + 2;
            this.EmitLoad(temp_regVal, element);
            this.EmitLoad(temp_regKey, i);
            this.Emit(Opcode_js_1.Opcode.PUT, [temp_regVal, reg, temp_regKey]);
            //this.EmitLoad(reg, element);
            //this.EmitPush(reg);
            //reg++;
        }
        return this;
    }
    EmitLoadISerializable(reg, obj) {
        let writer = new index_js_2.PBinaryWriter();
        obj.SerializeData(writer);
        this.EmitLoadBytes(reg, writer.toArray(), VMType_js_1.VMType.Bytes);
        return this;
    }
    EmitLoadVMObject(reg, obj) {
        let writer = new index_js_2.PBinaryWriter();
        let result = obj.SerializeObjectCall(writer);
        this.Emit(Opcode_js_1.Opcode.LOAD);
        this.AppendByte(reg);
        this.AppendByte(obj.Type);
        if (result == undefined) {
            //console.log("enter");
            if (obj.Data instanceof Map || (obj.Data instanceof Map && obj.Data instanceof VMObject_js_1.VMObject)) {
                let resultData = obj.Data;
                this.EmitVarInt(resultData.size);
                for (let entry of resultData) {
                    //console.log(entry[0]);
                    let key = entry[0];
                    let value = entry[1];
                    this.EmitLoadVMObject(reg + 1, key);
                    this.EmitLoadVMObject(reg + 2, value);
                    this.Emit(Opcode_js_1.Opcode.PUT, [reg + 1, reg, reg + 2]);
                }
            }
            else if (obj.Data instanceof VMObject_js_1.VMObject) {
                let writerNew = new index_js_2.PBinaryWriter();
                let bytes = obj.Data.SerializeData(writerNew);
                //console.log(bytes.length);
                this.EmitVarInt(bytes.length);
                this.AppendBytes(Array.from(bytes));
            }
        }
        else {
            //console.log("reg", reg);
            var bytes = Array.from(result);
            //console.log(bytes.length);
            this.EmitVarInt(bytes.length);
            this.AppendBytes(bytes);
        }
        //this.EmitLoadBytes(reg, Array.from(result), obj.Type);
        return this;
    }
    EmitLoadEnum(reg, enumVal) {
        // var temp = Convert.ToUInt32(enumVal);
        // var bytes = BitConverter.GetBytes(temp);
        let bytes = [0, 0, 0, 0];
        for (let i = 0; i < bytes.length; i++) {
            var byte = enumVal & 0xff;
            bytes[i] = byte;
            enumVal = (enumVal - byte) / 256;
        }
        this.EmitLoadBytes(reg, bytes, VMType_js_1.VMType.Enum);
        return this;
    }
    EmitLoadAddress(reg, obj) {
        let writer = new index_js_2.PBinaryWriter();
        obj.SerializeData(writer);
        let byteArray = Array.from(writer.toUint8Array());
        this.EmitLoadBytes(reg, byteArray, VMType_js_1.VMType.Bytes);
        return this;
    }
    EmitLoadTimestamp(reg, obj) {
        if (obj instanceof index_js_2.Timestamp) {
            let bytes = Array.from(index_js_2.Serialization.Serialize(obj));
            this.EmitLoadBytes(reg, bytes, VMType_js_1.VMType.Timestamp);
        }
        else if (obj instanceof Date) {
            let num = (obj.getTime() / 1000) | 0;
            let a = (num & 0xff000000) >> 24;
            let b = (num & 0x00ff0000) >> 16;
            let c = (num & 0x0000ff00) >> 8;
            let d = num & 0x000000ff;
            let bytes = [d, c, b, a];
            this.EmitLoadBytes(reg, bytes, VMType_js_1.VMType.Timestamp);
        }
        return this;
    }
    EmitLoadVarInt(reg, val) {
        let bytes = (0, index_js_3.numberToByteArray)(val);
        this.Emit(Opcode_js_1.Opcode.LOAD);
        this.AppendByte(reg);
        this.AppendByte(VMType_js_1.VMType.Number);
        this.AppendByte(bytes.length);
        this.EmitBytes(Array.from(bytes));
        return this;
    }
    EmitMove(src_reg, dst_reg) {
        this.Emit(Opcode_js_1.Opcode.MOVE);
        this.AppendByte(src_reg);
        this.AppendByte(dst_reg);
        return this;
    }
    EmitCopy(src_reg, dst_reg) {
        this.Emit(Opcode_js_1.Opcode.COPY);
        this.AppendByte(src_reg);
        this.AppendByte(dst_reg);
        return this;
    }
    EmitLabel(label) {
        this.Emit(Opcode_js_1.Opcode.NOP);
        this._labelLocations[label] = this.str.length;
        return this;
    }
    EmitJump(opcode, label, reg = 0) {
        switch (opcode) {
            case Opcode_js_1.Opcode.JMP:
            case Opcode_js_1.Opcode.JMPIF:
            case Opcode_js_1.Opcode.JMPNOT:
                this.Emit(opcode);
                break;
            default:
                throw new Error('Invalid jump opcode: ' + opcode);
        }
        if (opcode != Opcode_js_1.Opcode.JMP) {
            this.AppendByte(reg);
        }
        var ofs = this.str.length;
        this.AppendUshort(0);
        this._jumpLocations[ofs] = label;
        return this;
    }
    EmitCall(label, regCount) {
        if (regCount < 1 || regCount > MaxRegisterCount) {
            throw new Error('Invalid number of registers');
        }
        var ofs = this.str.length; //(int)stream.Position;
        ofs += 2;
        this.Emit(Opcode_js_1.Opcode.CALL);
        this.AppendByte(regCount);
        this.AppendUshort(0);
        this._jumpLocations[ofs] = label;
        return this;
    }
    EmitConditionalJump(opcode, src_reg, label) {
        if (opcode != Opcode_js_1.Opcode.JMPIF && opcode != Opcode_js_1.Opcode.JMPNOT) {
            throw new Error('Opcode is not a conditional jump');
        }
        var ofs = this.str.length;
        ofs += 2;
        this.Emit(opcode);
        this.AppendByte(src_reg);
        this.AppendUshort(0);
        this._jumpLocations[ofs] = label;
        return this;
    }
    InsertMethodArgs(args) {
        let temp_reg = 0;
        for (let i = args.length - 1; i >= 0; i--) {
            let arg = args[i];
            this.EmitLoad(temp_reg, arg);
            this.EmitPush(temp_reg);
        }
    }
    CallInterop(method, args) {
        this.InsertMethodArgs(args);
        let dest_reg = 0;
        this.EmitLoad(dest_reg, method);
        this.Emit(Opcode_js_1.Opcode.EXTCALL, [dest_reg]);
        return this;
    }
    CallContract(contractName, method, args) {
        this.InsertMethodArgs(args);
        let temp_reg = 0;
        this.EmitLoad(temp_reg, method);
        this.EmitPush(temp_reg);
        let src_reg = 0;
        let dest_reg = 1;
        this.EmitLoad(src_reg, contractName);
        this.Emit(Opcode_js_1.Opcode.CTX, [src_reg, dest_reg]);
        this.Emit(Opcode_js_1.Opcode.SWITCH, [dest_reg]);
        return this;
    }
    //#region ScriptBuilderExtensions
    AllowGas(from, to, gasPrice, gasLimit) {
        return this.CallContract(Contracts_js_1.Contracts.GasContractName, 'AllowGas', [from, to, gasPrice, gasLimit]);
    }
    SpendGas(address) {
        return this.CallContract(Contracts_js_1.Contracts.GasContractName, 'SpendGas', [address]);
    }
    async CallRPC(methodName, params) {
        return 'bla';
    }
    async GetAddressTransactionCount(address, chainInput) {
        let params = [address, chainInput];
        return await this.CallRPC('getAddressTransactionCount', params);
    }
    //#endregion
    EmitTimestamp(obj) {
        let num = (obj.getTime() / 1000) | 0;
        let a = (num & 0xff000000) >> 24;
        let b = (num & 0x00ff0000) >> 16;
        let c = (num & 0x0000ff00) >> 8;
        let d = num & 0x000000ff;
        let bytes = [d, c, b, a];
        this.AppendBytes(bytes);
        return this;
    }
    EmitByteArray(bytes) {
        this.EmitVarInt(bytes.length);
        this.EmitBytes(bytes);
        return this;
    }
    EmitVarString(text) {
        let bytes = this.RawString(text);
        this.EmitVarInt(bytes.length);
        this.EmitBytes(bytes);
        return this;
    }
    EmitVarInt(value) {
        if (value < 0)
            throw 'negative value invalid';
        if (value < 0xfd) {
            this.AppendByte(value);
        }
        else if (value <= 0xffff) {
            let B = (value & 0x0000ff00) >> 8;
            let A = value & 0x000000ff;
            // TODO check if the endianess is correct, might have to reverse order of appends
            this.AppendByte(0xfd);
            this.AppendByte(A);
            this.AppendByte(B);
        }
        else if (value <= 0xffffffff) {
            let C = (value & 0x00ff0000) >> 16;
            let B = (value & 0x0000ff00) >> 8;
            let A = value & 0x000000ff;
            // TODO check if the endianess is correct, might have to reverse order of appends
            this.AppendByte(0xfe);
            this.AppendByte(A);
            this.AppendByte(B);
            this.AppendByte(C);
        }
        else {
            let D = (value & 0xff000000) >> 24;
            let C = (value & 0x00ff0000) >> 16;
            let B = (value & 0x0000ff00) >> 8;
            let A = value & 0x000000ff;
            // TODO check if the endianess is correct, might have to reverse order of appends
            this.AppendByte(0xff);
            this.AppendByte(A);
            this.AppendByte(B);
            this.AppendByte(C);
            this.AppendByte(D);
        }
        return this;
    }
    EmitUInt32(value) {
        if (value < 0)
            throw 'negative value invalid';
        let D = (value & 0xff000000) >> 24;
        let C = (value & 0x00ff0000) >> 16;
        let B = (value & 0x0000ff00) >> 8;
        let A = value & 0x000000ff;
        // TODO check if the endianess is correct, might have to reverse order of appends
        this.AppendByte(0xff);
        this.AppendByte(A);
        this.AppendByte(B);
        this.AppendByte(C);
        this.AppendByte(D);
        return this;
    }
    EmitBytes(bytes) {
        for (let i = 0; i < bytes.length; i++)
            this.AppendByte(bytes[i]);
        // writer.Write(bytes);
        return this;
    }
    //Custom Modified
    ByteToHex(byte) {
        let result = ('0' + (byte & 0xff).toString(16)).slice(-2);
        return result;
    }
    AppendByte(byte) {
        this.str += this.ByteToHex(byte);
        this.writer.writeByte(byte);
    }
    //Custom Modified
    AppendBytes(bytes) {
        for (let i = 0; i < bytes.length; i++) {
            this.AppendByte(bytes[i]);
        }
    }
    AppendUshort(ushort) {
        this.str += this.ByteToHex(ushort & 0xff) + this.ByteToHex((ushort >> 8) & 0xff);
        this.writer.writeUnsignedShort(ushort);
    }
    AppendHexEncoded(bytes) {
        this.str += bytes;
        this.writer.writeBytes(Array.from((0, index_js_3.stringToUint8Array)(bytes)));
        return this;
    }
}
exports.ScriptBuilder = ScriptBuilder;
