// phantasma-sdk-ts/dist/cjs/core/vm/VMObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VMObject = void 0;
const VMType_js_1 = require("./VMType.js");
const Timestamp_js_1 = require("../types/Timestamp.js");
const index_js_1 = require("../types/index.js");
const index_js_2 = require("../interfaces/index.js");
const index_js_3 = require("../utils/index.js");
class VMObject {
    get IsEmpty() {
        return this.Data == null || this.Data == undefined;
    }
    GetChildren() {
        return this.Type == VMType_js_1.VMType.Struct ? this.Data : null;
    }
    get Size() {
        let total = 0;
        if (this.Type == VMType_js_1.VMType.Object) {
            const children = this.GetChildren();
            let values = children?.values;
            for (let entry in values) {
                total += entry.length;
            }
        }
        else {
            total = this._localSize;
        }
        return total;
    }
    constructor() {
        this._localSize = 0;
        this.Type = VMType_js_1.VMType.None;
        this.Data = null;
    }
    AsTimestamp() {
        if (this.Type != VMType_js_1.VMType.Timestamp) {
            throw new Error(`Invalid cast: expected timestamp, got ${this.Type}`);
        }
        return this.Data;
    }
    AsByteArray() {
        switch (this.Type) {
            case VMType_js_1.VMType.Bytes:
                return this.Data;
            case VMType_js_1.VMType.Bool:
                return new Uint8Array([this.Data ? 1 : 0]);
            case VMType_js_1.VMType.String:
                return new TextEncoder().encode(this.AsString());
            case VMType_js_1.VMType.Number:
            // Here you will have to convert BigInteger to Uint8Array manually
            case VMType_js_1.VMType.Enum:
                var num = this.AsNumber();
                var bytes = new Uint8Array(new ArrayBuffer(4));
                new DataView(bytes.buffer).setUint32(0, num);
                return bytes;
            case VMType_js_1.VMType.Timestamp:
                var time = this.AsTimestamp();
                var bytes = new Uint8Array(new ArrayBuffer(4));
                new DataView(bytes.buffer).setUint32(0, time.value);
                return bytes;
            case VMType_js_1.VMType.Struct:
            // Here you will have to convert struct to Uint8Array manually
            case VMType_js_1.VMType.Object:
            // here you will have to convert ISerializable to Uint8Array manually
            default:
                throw new Error(`Invalid cast: expected bytes, got ${this.Type}`);
        }
    }
    AsString() {
        switch (this.Type) {
            case VMType_js_1.VMType.String:
                return this.Data?.toString();
            case VMType_js_1.VMType.Number:
                return this.Data.toString();
            case VMType_js_1.VMType.Bytes:
                return (0, index_js_3.uint8ArrayToStringDefault)(this.Data);
            case VMType_js_1.VMType.Enum:
                return this.Data.toString();
            case VMType_js_1.VMType.Object:
                if (this.Data instanceof index_js_1.Address) {
                    return this.Data.Text;
                }
                /*if (this.Data instanceof Hash) {
                      return this.Data.toString();
                  }*/
                return 'Interop:' + this.Data?.constructor.name;
            case VMType_js_1.VMType.Struct:
                const arrayType = this.GetArrayType();
                if (arrayType === VMType_js_1.VMType.Number) {
                    // convert array of unicode numbers into a string
                    const children = this.GetChildren();
                    let sb = '';
                    for (let i = 0; i < children.size; i++) {
                        const key = VMObject.FromObject(i);
                        const val = children?.get(key);
                        const ch = String.fromCharCode(val.AsNumber());
                        sb += ch;
                    }
                    return sb;
                }
                else {
                    /*const buffer = new ArrayBuffer(this.Data?.length as number);
                        const view = new DataView(buffer);
                        this.SerializeData(view);
                        return btoa(String.fromCharCode.apply(null, new Uint8Array(buffer)));*/
                }
            case VMType_js_1.VMType.Bool:
                return this.Data ? 'true' : 'false';
            case VMType_js_1.VMType.Timestamp:
                return this.Data.value.toString();
            default:
                throw new Error(`Invalid cast: expected string, got ${this.Type}`);
        }
    }
    ToString() {
        switch (this.Type) {
            case VMType_js_1.VMType.String:
                return this.Data;
            case VMType_js_1.VMType.Number:
                return this.Data.toString();
            case VMType_js_1.VMType.Bytes:
                return new TextDecoder().decode(this.Data);
            case VMType_js_1.VMType.Enum:
                return this.Data.toString();
            case VMType_js_1.VMType.Object:
                if (this.Data instanceof index_js_1.Address) {
                    return this.Data.Text;
                }
                /*if (this.Data instanceof Hash) {
                      return this.Data.toString();
                  }*/
                return 'Interop:' + this.Data.constructor.name;
            case VMType_js_1.VMType.Struct:
                const arrayType = this.GetArrayType();
                if (arrayType === VMType_js_1.VMType.Number) {
                    // convert array of unicode numbers into a string
                    const children = this.GetChildren();
                    let sb = '';
                    for (let i = 0; i < children.size; i++) {
                        const key = VMObject.FromObject(i);
                        const val = children.get(key);
                        const ch = String.fromCharCode(val.AsNumber());
                        sb += ch;
                    }
                    return sb;
                }
                else {
                    /*const stream = new PMemoryStream();
                        const writer = new PBinaryWriter(stream);
                        this.SerializeData(writer);
                        return new TextDecoder().decode(stream.toArray());*/
                }
            case VMType_js_1.VMType.Bool:
                return this.Data ? 'true' : 'false';
            case VMType_js_1.VMType.Timestamp:
                return this.Data.value.toString();
            default:
                throw new Error(`Invalid cast: expected string, got ${this.Type}`);
        }
    }
    AsNumber() {
        if ((this.Type === VMType_js_1.VMType.Object || this.Type === VMType_js_1.VMType.Timestamp) &&
            this.Data instanceof Timestamp_js_1.Timestamp) {
            return this.Data.value;
        }
        switch (this.Type) {
            case VMType_js_1.VMType.String: {
                let number = Number(this.Data);
                if (number.toString() === this.Data) {
                    return number;
                }
                else {
                    throw new Error(`Cannot convert String '${this.Data}' to BigInteger.`);
                }
            }
            case VMType_js_1.VMType.Bytes: {
                const bytes = new Uint8Array(this.Data);
                const num = BigInt(`0x${bytes.join('')}`);
                return num;
            }
            case VMType_js_1.VMType.Enum: {
                const num = Number(this.Data);
                return BigInt(num);
            }
            case VMType_js_1.VMType.Bool: {
                const val = this.Data;
                return val ? 1 : 0;
            }
            default: {
                if (this.Type !== VMType_js_1.VMType.Number) {
                    throw new Error(`Invalid cast: expected number, got ${this.Type}`);
                }
                return Number(this.Data);
            }
        }
    }
    AsEnum() {
        if (!VMObject.isEnum(this.Data)) {
            throw new Error('T must be an enumerated type');
        }
        if (this.Type !== VMType_js_1.VMType.Enum) {
            this.Data = new Number(this.AsNumber());
        }
        return this.Data;
    }
    GetArrayType() {
        if (this.Type !== VMType_js_1.VMType.Struct) {
            return VMType_js_1.VMType.None;
        }
        const children = this.GetChildren();
        let result = VMType_js_1.VMType.None;
        for (let i = 0; i < children.size; i++) {
            const key = VMObject.FromObject(i);
            if (!children.has(key)) {
                return VMType_js_1.VMType.None;
            }
            const val = children.get(key);
            if (result === VMType_js_1.VMType.None) {
                result = val.Type;
            }
            else if (val.Type !== result) {
                return VMType_js_1.VMType.None;
            }
        }
        return result;
    }
    AsType(type) {
        switch (type) {
            case VMType_js_1.VMType.Bool:
                return this.AsBool();
            case VMType_js_1.VMType.String:
                return this.AsString();
            case VMType_js_1.VMType.Bytes:
                return this.AsByteArray();
            case VMType_js_1.VMType.Number:
                return this.AsNumber();
            case VMType_js_1.VMType.Timestamp:
                return this.AsTimestamp();
            default:
                throw 'Unsupported VM cast';
        }
    }
    static isEnum(instance) {
        if (instance == null)
            return false;
        let keys = Object.keys(instance);
        let values = [];
        for (let key of keys) {
            let value = instance[key];
            if (typeof value === 'number') {
                value = value.toString();
            }
            values.push(value);
        }
        for (let key of keys) {
            if (values.indexOf(key) < 0) {
                return false;
            }
        }
        return true;
    }
    /*public AsEnum<T>(): T {
          if (isEnum(this.Data)) {
              throw new ArgumentException("T must be an enumerated type");
          }
          if (this.Type != VMType.Enum) {
              const num = this.AsNumber();
              this.Data = Number(this.Data);
            }
  
            return (T) Enum.Parse(typeof T, this.Data.toString());
      }*/
    AsBool() {
        switch (this.Type) {
            case VMType_js_1.VMType.String:
                return this.Data.toLowerCase() == 'true';
            case VMType_js_1.VMType.Number:
                return this.Data != BigInt(0);
            case VMType_js_1.VMType.Bool:
                return this.Data ? true : false;
            default:
                throw new Error(`Invalid cast: expected bool, got ${this.Type}`);
        }
    }
    static isStructOrClass(type) {
        return ((!VMObject.isPrimitive(type) && VMObject.isValueType(type) && !VMObject.isEnum(type)) ||
            VMObject.isClass(type) ||
            VMObject.isInterface(type));
    }
    static isSerializable(type) {
        return (type instanceof index_js_2.ISerializable ||
            VMObject.isPrimitive(type) ||
            VMObject.isStructOrClass(type) ||
            VMObject.isEnum(type));
    }
    static isPrimitive(type) {
        return type === String || type === Number || type === Boolean || type === BigInt;
    }
    static isValueType(type) {
        return type === Object;
    }
    static isClass(type) {
        return (type === Array ||
            type === Map ||
            type === Set ||
            type instanceof Object ||
            (typeof type).toLowerCase() === 'object');
    }
    static isInterface(type) {
        return type === Map;
    }
    static ConvertObjectInternal(fieldValue, fieldType) {
        if ((VMObject.isStructOrClass(fieldType) && fieldValue instanceof Uint8Array) ||
            fieldValue instanceof Array) {
            const bytes = fieldValue;
            fieldValue = index_js_1.Serialization.Unserialize(bytes, typeof fieldType);
        }
        else if (VMObject.isEnum(fieldType)) {
            let tempValue = fieldValue;
            fieldValue = tempValue;
        }
        return fieldValue;
    }
    ToArray(arrayElementType) {
        if (this.Type !== VMType_js_1.VMType.Struct) {
            throw new Error('not a valid source struct');
        }
        const children = this.GetChildren();
        let maxIndex = -1;
        for (const child of children) {
            if (child[0].Type !== VMType_js_1.VMType.Number) {
                throw new Error('source contains an element with invalid array index');
            }
            const temp = child[0].AsNumber();
            // TODO use a constant for VM max array size
            if (temp >= 1024) {
                throw new Error('source contains an element with a very large array index');
            }
            const index = Math.floor(temp);
            if (index < 0) {
                throw new Error('source contains an array index with negative value');
            }
            maxIndex = Math.max(index, maxIndex);
        }
        const length = maxIndex + 1;
        const array = new Array(length);
        for (const child of children) {
            const temp = child[0].AsNumber();
            const index = Math.floor(temp);
            let val = child[1].ToObjectType(arrayElementType);
            val = VMObject.ConvertObjectInternal(val, arrayElementType);
            array[index] = val;
        }
        return array;
    }
    ToObjectType(type) {
        if (this.Type === VMType_js_1.VMType.Struct) {
            if (Array.isArray(type)) {
                const elementType = typeof type;
                return this.ToArray(elementType);
            }
            else if (VMObject.isStructOrClass(type)) {
                return this.ToStruct(type);
            }
            else {
                throw new Error('some stuff still missing: eg: lists, dictionaries..');
            }
        }
        else {
            const temp = this.ToObject();
            return temp;
        }
    }
    ToObject() {
        if (this.Type === VMType_js_1.VMType.None) {
            throw new Error('not a valid object');
        }
        switch (this.Type) {
            case VMType_js_1.VMType.Bool:
                return this.AsBool();
            case VMType_js_1.VMType.Bytes:
                return this.AsByteArray();
            case VMType_js_1.VMType.String:
                return this.AsString();
            case VMType_js_1.VMType.Number:
                return this.AsNumber();
            case VMType_js_1.VMType.Timestamp:
                return this.AsTimestamp();
            case VMType_js_1.VMType.Object:
                return this.Data;
            case VMType_js_1.VMType.Enum:
                return this.Data;
            case VMType_js_1.VMType.Struct:
                let objs = this.GetChildren();
                let result = [];
                for (let obj of objs) {
                    result.push(obj[1].ToObject());
                }
                return result;
            default:
                throw new Error(`Cannot cast ${this.Type} to object`);
        }
    }
    ToStruct(structType) {
        if (this.Type !== VMType_js_1.VMType.Struct) {
            throw new Error('not a valid source struct');
        }
        if (!VMObject.isStructOrClass(structType)) {
            throw new Error('not a valid destination struct');
        }
        //let localType = Object.apply(typeof structType);
        let localType; //: typeof type;
        if (structType.name != undefined) {
            let className = structType.name;
            localType = Object.apply(className);
        }
        else {
            localType = new structType();
        }
        const dict = this.GetChildren();
        const dictKeys = this.GetChildren().keys();
        const result = new structType();
        const fields = Object.keys(typeof localType);
        let fields2 = index_js_1.Describer.describe(structType, false);
        //console.log("fields", keyof typeof structType);
        fields2 = fields2.map((x) => x.replace('this.', ''));
        for (let field of fields2) {
            field.replace('this.', '');
            const key = VMObject.FromObject(field);
            //const key = field;
            const dictKey = dictKeys.next().value;
            let val;
            if (dictKey?.Data.toString() == key.Data.toString()) {
                let localValue = dict.get(dictKey);
                if (localValue.GetChildren() != undefined) {
                    result[field] = localValue.ToArray(typeof localType[field]);
                    continue;
                }
                else {
                    val = localValue.ToObject();
                }
                //val = Serialization.Unserialize(dict.get(dictKey));
            }
            else {
                if (!VMObject.isStructOrClass(structType[field])) {
                    //console.log(`field not present in source struct: ${field}`);
                    //throw new Error(`field not present in source struct: ${field}`);
                }
                //val = null;
            }
            /*if (val !== null && localType[field] !== "Uint8Array") {
              if (VMObject.isSerializable(localType[field])) {
                const temp = new structType[field]();
                const stream = new Uint8Array(val);
                const reader = new PBinaryReader(stream);
                (temp as ISerializable).UnserializeData(reader);
                val = temp;
              }
            }*/
            if (VMObject.isEnum(typeof structType[field]) && !VMObject.isEnum(val)) {
                val = localType[field][val?.toString()];
            }
            // If field hasn't been found, val will be 'undefined'.
            // This will override constructors which initialize empty fields correctly.
            // Example: ConsensusPoll initialize empty array 'entries' but code below will convert in into undefined,
            // breaking corresponding test when using ES2020.
            // Adding check here.
            if (val !== undefined) {
                result[field] = val;
            }
        }
        return result;
    }
    static GetVMType(type) {
        if (VMObject.isEnum(type)) {
            return VMType_js_1.VMType.Enum;
        }
        if (type === Boolean || type.toLowerCase() === 'boolean') {
            return VMType_js_1.VMType.Bool;
        }
        if (typeof type == typeof String || type === String || type.toLowerCase() === 'string') {
            return VMType_js_1.VMType.String;
        }
        if (typeof type === typeof Uint8Array ||
            type === Uint8Array ||
            type.toLowerCase() === 'uint8array') {
            return VMType_js_1.VMType.Bytes;
        }
        if (type === 'BigInt' ||
            type === Number ||
            type === BigInt ||
            type.toLowerCase() === 'number') {
            return VMType_js_1.VMType.Number;
        }
        if (type === Timestamp_js_1.Timestamp || type === Number) {
            return VMType_js_1.VMType.Timestamp;
        }
        if (VMObject.isEnum(type)) {
            return VMType_js_1.VMType.Enum;
        }
        if (Array.isArray(type)) {
            return VMType_js_1.VMType.Struct;
        }
        if (VMObject.isClass(type) || VMObject.isValueType(type)) {
            return VMType_js_1.VMType.Object;
        }
        return VMType_js_1.VMType.Struct;
    }
    static IsVMType(type) {
        const result = VMObject.GetVMType(type);
        return result !== VMType_js_1.VMType.None;
    }
    SetValue(value) {
        this.Data = value;
        return this;
    }
    setValue(val, type) {
        this.Type = type;
        this._localSize = val != null ? val.length : 0;
        switch (type) {
            case VMType_js_1.VMType.Bytes:
                this.Data = val;
                break;
            case VMType_js_1.VMType.Number:
                this.Data = val == null ? BigInt(0) : val;
                break;
            case VMType_js_1.VMType.String:
                this.Data = new String(val);
                break;
            case VMType_js_1.VMType.Enum:
                this.Data = val;
                break;
            case VMType_js_1.VMType.Timestamp:
                this.Data = new Timestamp_js_1.Timestamp(val.slice(0, 4));
                break;
            case VMType_js_1.VMType.Bool:
                this.Data = new Boolean(val[0] === 1);
                break;
            default:
                if (val instanceof Uint8Array) {
                    const len = val.length;
                    switch (len) {
                        case index_js_1.Address.LengthInBytes:
                            this.Data = index_js_1.Address.FromBytes(val);
                            break;
                        /*case Hash.Length:
                                    this.Data = Hash.fromBytes(val);
                                    break;*/
                        /*default:
                                    try {
                                        this.unserializeData(val);
                                    } catch (e) {
                                        throw new Error("Cannot decode interop object from bytes with length: " + len);
                                    }
                                    break;*/
                    }
                    break;
                }
                else {
                    throw new Error('Cannot set value for vmtype: ' + type);
                }
        }
    }
    static ValidateStructKey(key) {
        if (key.Type == VMType_js_1.VMType.None || key.Type == VMType_js_1.VMType.Struct || key.Type == VMType_js_1.VMType.Object) {
            throw new Error(`Cannot use value of type ${key.Type} as key for struct field`);
        }
    }
    CastViaReflection(srcObj, level, dontConvertSerializables = true) {
        const srcType = srcObj.constructor.name;
        if (Array.isArray(srcObj)) {
            const children = new Map();
            const array = srcObj;
            for (let i = 0; i < array.length; i++) {
                const val = array[i];
                const key = new VMObject();
                key.SetValue(i);
                const vmVal = this.CastViaReflection(val, level + 1);
                children.set(key, vmVal);
            }
            const result = new VMObject();
            result.SetValue(children);
            return result;
        }
        else {
            const targetType = VMObject.GetVMType(srcType);
            let result;
            let isKnownType = srcType === VMType_js_1.VMType.Number || srcType === VMType_js_1.VMType.Timestamp;
            let localType = Object.apply(typeof srcType);
            if (!isKnownType && dontConvertSerializables && VMObject.isSerializable(localType)) {
                isKnownType = true;
            }
            if (VMObject.isStructOrClass(localType) && !isKnownType) {
                const children = new Map();
                const fields = Object.keys(srcObj);
                if (fields.length > 0) {
                    fields.forEach((field) => {
                        const key = VMObject.FromObject(field);
                        VMObject.ValidateStructKey(key);
                        const val = srcObj[field];
                        const vmVal = this.CastViaReflection(val, level + 1, true);
                        children.set(key, vmVal);
                    });
                    result = new VMObject();
                    result.SetValue(children);
                    result.Type = VMType_js_1.VMType.Struct;
                    return result;
                }
            }
            result = VMObject.FromObject(srcObj);
            if (result != null) {
                return result;
            }
            throw new Error(`invalid cast: Interop.${srcType} to vm object`);
        }
    }
    SetKey(key, obj) {
        VMObject.ValidateStructKey(key);
        let children;
        const temp = new VMObject();
        temp.Copy(key);
        key = temp;
        if (this.Type == VMType_js_1.VMType.Struct) {
            children = this.GetChildren();
        }
        else if (this.Type == VMType_js_1.VMType.None) {
            this.Type = VMType_js_1.VMType.Struct;
            children = new Map();
            this.Data = children;
            this._localSize = 0;
        }
        else {
            throw new Error(`Invalid cast from ${this.Type} to struct`);
        }
        const result = new VMObject();
        children?.set(key, result);
        result.Copy(obj);
    }
    Copy(other) {
        if (!other || other.Type == VMType_js_1.VMType.None) {
            this.Type = VMType_js_1.VMType.None;
            this.Data = null;
            return;
        }
        this.Type = other.Type;
        if (other.Type == VMType_js_1.VMType.Struct) {
            const children = new Map();
            const otherChildren = other.GetChildren();
            otherChildren?.forEach((val, key) => {
                const temp = new VMObject();
                temp.Copy(val);
                children.set(key, temp);
            });
            this.Data = children;
        }
        else {
            this.Data = other.Data;
        }
    }
    SetType(type) {
        this.Type = type;
    }
    static FromArray(array) {
        const result = new VMObject();
        for (let i = 0; i < array.length; i++) {
            const key = VMObject.FromObject(i);
            const val = VMObject.FromObject(array[i]);
            result.SetKey(key, val);
        }
        return result;
    }
    static CastTo(srcObj, type) {
        if (srcObj.Type == type) {
            let result = new VMObject();
            result.Copy(srcObj);
            return result;
        }
        switch (type) {
            case VMType_js_1.VMType.None:
                return new VMObject();
            case VMType_js_1.VMType.String: {
                let result = new VMObject();
                result.SetValue(srcObj.AsString());
                return result;
            }
            case VMType_js_1.VMType.Timestamp: {
                let result = new VMObject();
                result.SetValue(srcObj.AsTimestamp());
                return result;
            }
            case VMType_js_1.VMType.Bool: {
                let result = new VMObject();
                result.SetValue(srcObj.AsBool());
                return result;
            }
            case VMType_js_1.VMType.Bytes: {
                let result = new VMObject();
                result.SetValue(srcObj.AsByteArray());
                return result;
            }
            case VMType_js_1.VMType.Number: {
                let result = new VMObject();
                result.SetValue(srcObj.AsNumber());
                return result;
            }
            case VMType_js_1.VMType.Struct: {
                switch (srcObj.Type) {
                    case VMType_js_1.VMType.Enum: {
                        var result = new VMObject();
                        result.SetValue(srcObj.AsEnum()); // TODO does this work for all types?
                        return result;
                    }
                    case VMType_js_1.VMType.Object: {
                        var result = new VMObject();
                        result.SetValue(srcObj.CastViaReflection(srcObj.Data, 0)); // TODO does this work for all types?
                        return result;
                    }
                    default:
                        throw `invalid cast: ${srcObj.Type} to ${type}`;
                }
            }
            default:
                throw `invalid cast: ${srcObj.Type} to ${type}`;
        }
    }
    static FromObject(obj) {
        const objType = obj.constructor.name;
        const type = this.GetVMType(objType);
        if (type === VMType_js_1.VMType.None) {
            throw new Error('not a valid object');
        }
        const result = new VMObject();
        switch (type) {
            case VMType_js_1.VMType.Bool:
                result.setValue(obj, VMType_js_1.VMType.Bool);
                break;
            case VMType_js_1.VMType.Bytes:
                result.setValue(obj, VMType_js_1.VMType.Bytes);
                break;
            case VMType_js_1.VMType.String:
                result.setValue(obj, VMType_js_1.VMType.String);
                break;
            case VMType_js_1.VMType.Enum:
                result.setValue(obj, VMType_js_1.VMType.Enum);
                break;
            case VMType_js_1.VMType.Object:
                result.setValue(obj, VMType_js_1.VMType.Object);
                break;
            case VMType_js_1.VMType.Number:
                /*if (objType === Number) {
                  obj = BigInt(obj);
                }*/
                result.setValue(obj, VMType_js_1.VMType.Number);
                break;
            case VMType_js_1.VMType.Timestamp:
                /*if (objType === "Number") {
                  obj = new Timestamp(obj);
                }*/
                result.setValue(obj, VMType_js_1.VMType.Timestamp);
                break;
            case VMType_js_1.VMType.Struct:
                result.Type = VMType_js_1.VMType.Struct;
                if (Array.isArray(obj)) {
                    return this.FromArray(obj);
                }
                else {
                    return this.FromStruct(obj);
                }
                break;
            default:
                return null;
        }
        return result;
    }
    static FromEnum(obj) {
        let vm = new VMObject();
        vm.setValue(obj, VMType_js_1.VMType.Enum);
        return vm;
    }
    static FromStruct(obj) {
        let vm = new VMObject();
        return vm.CastViaReflection(obj, 0, false);
    }
    // Serialization
    static FromBytes(bytes) {
        var result = new VMObject();
        var reader = new index_js_1.PBinaryReader(bytes);
        result.UnserializeData(reader);
        return result;
    }
    SerializeData(writer) {
        writer.writeByte(this.Type);
        if (this.Type == VMType_js_1.VMType.None) {
            return;
        }
        let dataType = typeof this.Data;
        switch (this.Type) {
            case VMType_js_1.VMType.Struct: {
                let children = this.GetChildren();
                writer.writeVarInt(children.size);
                for (const child of children) {
                    child[0].SerializeData(writer);
                    //console.log(Base16.encodeUint8Array(writer.toUint8Array()));
                    child[1].SerializeData(writer);
                    ///console.log(Base16.encodeUint8Array(writer.toUint8Array()));
                }
                /*children.forEach((key, value) => {
                  key.SerializeData(writer);
                  value.SerializeData(writer);
                });*/
                break;
            }
            case VMType_js_1.VMType.Object: {
                var obj = this.Data;
                if (obj != null) {
                    var bytes = index_js_1.Serialization.Serialize(obj);
                    var uintBytes = (0, index_js_3.uint8ArrayToBytes)(bytes);
                    writer.writeByteArray(uintBytes);
                }
                else {
                    throw `Objects of type ${dataType} cannot be serialized`;
                }
                break;
            }
            case VMType_js_1.VMType.Enum: {
                let temp2 = 0;
                temp2 = this.Data;
                let uint8array = new Uint8Array(4);
                writer = new index_js_1.PBinaryWriter(uint8array);
                writer.writeByte(this.Type);
                writer.writeEnum(temp2);
                break;
            }
            default:
                let localBytes = index_js_1.Serialization.Serialize(this.Data);
                writer.writeByteArray(localBytes);
                break;
        }
        return writer.toUint8Array();
    }
    SerializeObjectCall(writer) {
        if (this.Type == VMType_js_1.VMType.None) {
            return;
        }
        let dataType = typeof this.Data;
        switch (this.Type) {
            case VMType_js_1.VMType.Struct: {
                let children = this.GetChildren();
                writer.writeVarInt(children.size);
                for (const child of children) {
                    child[0].SerializeData(writer);
                    child[1].SerializeData(writer);
                }
                /*children.forEach((key, value) => {
                  key.SerializeData(writer);
                  value.SerializeData(writer);
                });*/
                break;
            }
            case VMType_js_1.VMType.Object: {
                var obj = this.Data;
                if (obj != null) {
                    var bytes = index_js_1.Serialization.Serialize(obj);
                    var uintBytes = (0, index_js_3.uint8ArrayToBytes)(bytes);
                    writer.writeByteArray(uintBytes);
                }
                else {
                    throw `Objects of type ${dataType} cannot be serialized`;
                }
                break;
            }
            case VMType_js_1.VMType.Enum: {
                let temp2 = 0;
                temp2 = this.Data;
                let uint8array = new Uint8Array(4);
                writer = new index_js_1.PBinaryWriter(uint8array);
                writer.writeEnum(temp2);
                break;
            }
            default:
                let localBytes = index_js_1.Serialization.Serialize(this.Data);
                writer.writeByteArray((0, index_js_3.uint8ArrayToBytes)(localBytes));
                break;
        }
        return writer.toUint8Array();
    }
    /*UnserializeData(reader: PBinaryReader) {
      this.Type = reader.readByte();
      if (this.Type == VMType.None) {
        return;
      }
  
      switch (this.Type) {
        case VMType.Struct: {
          let children = new Map<VMObject, VMObject>();
          let count = reader.readVarInt();
          for (let i = 0; i < count; i++) {
            let key = new VMObject();
            key.UnserializeData(reader);
            let value = new VMObject();
            value.UnserializeData(reader);
            children.set(key, value);
          }
          this.Data = children;
          break;
        }
  
        case VMType.Object: {
          var bytes = reader.readByteArray();
          var obj = Serialization.Unserialize(bytes);
          this.Data = obj as Object;
          break;
        }
  
        case VMType.Enum: {
          let temp = reader.readVarInt();
          this.Data = temp as unknown as Enumerator;
          break;
        }
  
        default:
          this.Data = Serialization.UnserializeObject(reader, null);
          break;
      }
    }*/
    UnserializeData(reader) {
        this.Type = reader.readByte();
        switch (this.Type) {
            case VMType_js_1.VMType.Bool:
                this.Data = index_js_1.Serialization.Unserialize(reader, Boolean);
                break;
            case VMType_js_1.VMType.Bytes:
                this.Data = index_js_1.Serialization.Unserialize(reader, Uint8Array);
                break;
            case VMType_js_1.VMType.Number:
                this.Data = reader.readBigInteger();
                break;
            case VMType_js_1.VMType.Timestamp:
                this.Data = reader.readTimestamp();
                /*this.Data = Serialization.UnserializeObject<Timestamp>(
                  reader,
                  Timestamp
                );*/
                break;
            case VMType_js_1.VMType.String:
                this.Data = reader.readVarString();
                break;
            case VMType_js_1.VMType.Struct:
                let childCount = reader.readVarInt();
                let children = new Map();
                while (childCount > 0) {
                    let key = new VMObject();
                    key.UnserializeData(reader);
                    VMObject.ValidateStructKey(key);
                    let val = new VMObject();
                    val.UnserializeData(reader);
                    children.set(key, val);
                    childCount--;
                }
                this.Data = children;
                break;
            case VMType_js_1.VMType.Object:
                let bytes = reader.readByteArray();
                if (bytes.length == 35) {
                    let addr = index_js_1.Serialization.Unserialize(bytes, index_js_1.Address);
                    this.Data = addr;
                    this.Type = VMType_js_1.VMType.Object;
                }
                else {
                    this.Type = VMType_js_1.VMType.Bytes;
                    this.Data = bytes;
                }
                break;
            case VMType_js_1.VMType.Enum:
                this.Type = VMType_js_1.VMType.Enum;
                this.Data = reader.readVarInt();
                break;
            case VMType_js_1.VMType.None:
                this.Type = VMType_js_1.VMType.None;
                this.Data = null;
                break;
            default:
                throw new Error(`invalid unserialize: type ${this.Type}`);
        }
    }
}
exports.VMObject = VMObject;
VMObject.TimeFormat = 'MM/dd/yyyy HH:mm:ss';
