// phantasma-sdk-ts/dist/esm/core/link/easyScript.js
import { ScriptBuilder } from "../vm/index.js";
export var Nexus;
(function (Nexus) {
    Nexus["Mainnet"] = "mainnet";
    Nexus["Simnet"] = "simnet";
    Nexus["Testnet"] = "testnet";
})(Nexus || (Nexus = {}));
export class EasyScript {
    constructor(nexus = Nexus.Mainnet) {
        this.sb = new ScriptBuilder();
        this.nexus = nexus;
    }
    buildScript(_type, _options = [null]) {
        this.sb = new ScriptBuilder();
        switch (_type) {
            case "interact":
                let contractNameInteract = _options[0];
                let methodNameInteract = _options[1];
                let inputArgumentsInteract = _options[2];
                return this.sb
                    .CallContract("gas", "AllowGas", [])
                    .CallContract(contractNameInteract, methodNameInteract, inputArgumentsInteract) //The Meat of the Script
                    .CallContract("gas", "SpendGas", [])
                    .EndScript();
                break;
            case "invoke":
                let contractNameInvoke = _options[0];
                let methodNameInvoke = _options[1];
                let inputArgumentsInvoke = _options[2];
                return this.sb
                    .CallContract(contractNameInvoke, methodNameInvoke, inputArgumentsInvoke) //The Meat of the Script
                    .EndScript();
                break;
            case "interop":
                let interopNameInterop = _options[0];
                let inputArgumentsInterop = _options[1];
                return this.sb
                    .CallContract("gas", "AllowGas", [])
                    .CallInterop(interopNameInterop, inputArgumentsInterop)
                    .CallContract("gas", "SpendGas", [])
                    .EndScript();
                break;
        }
    }
}
