// phantasma-sdk-ts/dist/cjs/core/types/Carbon/TxTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxTypes = void 0;
var TxTypes;
(function (TxTypes) {
    TxTypes[TxTypes["Call"] = 0] = "Call";
    TxTypes[TxTypes["Call_Multi"] = 1] = "Call_Multi";
    TxTypes[TxTypes["Trade"] = 2] = "Trade";
    TxTypes[TxTypes["TransferFungible"] = 3] = "TransferFungible";
    TxTypes[TxTypes["TransferFungible_GasPayer"] = 4] = "TransferFungible_GasPayer";
    TxTypes[TxTypes["TransferNonFungible_Single"] = 5] = "TransferNonFungible_Single";
    TxTypes[TxTypes["TransferNonFungible_Single_GasPayer"] = 6] = "TransferNonFungible_Single_GasPayer";
    TxTypes[TxTypes["TransferNonFungible_Multi"] = 7] = "TransferNonFungible_Multi";
    TxTypes[TxTypes["TransferNonFungible_Multi_GasPayer"] = 8] = "TransferNonFungible_Multi_GasPayer";
    TxTypes[TxTypes["MintFungible"] = 9] = "MintFungible";
    TxTypes[TxTypes["BurnFungible"] = 10] = "BurnFungible";
    TxTypes[TxTypes["BurnFungible_GasPayer"] = 11] = "BurnFungible_GasPayer";
    TxTypes[TxTypes["MintNonFungible"] = 12] = "MintNonFungible";
    TxTypes[TxTypes["BurnNonFungible"] = 13] = "BurnNonFungible";
    TxTypes[TxTypes["BurnNonFungible_GasPayer"] = 14] = "BurnNonFungible_GasPayer";
    TxTypes[TxTypes["Phantasma"] = 15] = "Phantasma";
    TxTypes[TxTypes["Phantasma_Raw"] = 16] = "Phantasma_Raw";
})(TxTypes = exports.TxTypes || (exports.TxTypes = {}));
