// phantasma-sdk-ts/dist/esm/core/types/Carbon/Blockchain/Modules/TokenInfo.js
import { IntX } from '../../IntX.js';
import { Bytes32 } from '../../Bytes32.js';
import { SmallString } from '../../SmallString.js';
import { CarbonTokenFlags } from '../CarbonTokenFlags.js';
export class TokenInfo {
    constructor(init) {
        this.maxSupply = new IntX();
        this.flags = CarbonTokenFlags.None;
        this.decimals = 0;
        this.owner = new Bytes32();
        this.symbol = new SmallString();
        this.metadata = new Uint8Array();
        this.tokenSchemas = undefined;
        Object.assign(this, init);
    }
    write(w) {
        this.maxSupply.write(w);
        w.write1(this.flags & 0xff);
        w.write1(this.decimals & 0xff);
        this.owner.write(w);
        this.symbol.write(w);
        w.writeArray(this.metadata);
        if ((this.flags & CarbonTokenFlags.NonFungible) !== 0) {
            w.writeArray(this.tokenSchemas ?? new Uint8Array());
        }
    }
    read(r) {
        this.maxSupply = r.readBlob(IntX);
        this.flags = r.read1();
        this.decimals = r.read1();
        this.owner = Bytes32.read(r);
        this.symbol = r.readBlob(SmallString);
        this.metadata = r.readArray();
        if ((this.flags & CarbonTokenFlags.NonFungible) !== 0) {
            this.tokenSchemas = r.readArray();
        }
        else {
            this.tokenSchemas = undefined;
        }
    }
    static read(r) {
        const v = new TokenInfo();
        v.read(r);
        return v;
    }
}
