// phantasma-sdk-ts/dist/esm/core/types/Carbon/Blockchain/Modules/TokenListing.js
import { Bytes32 } from '../../Bytes32.js';
import { IntX } from '../../IntX.js';
export var ListingType;
(function (ListingType) {
    ListingType[ListingType["FixedPrice"] = 0] = "FixedPrice";
})(ListingType || (ListingType = {}));
export class TokenListing {
    constructor(type = ListingType.FixedPrice, seller = new Bytes32(), quoteTokenId = 0n, price = new IntX(), startDate = 0n, endDate = 0n) {
        this.type = type;
        this.seller = seller;
        this.quoteTokenId = quoteTokenId;
        this.price = price;
        this.startDate = startDate;
        this.endDate = endDate;
    }
    write(w) {
        w.write1(this.type);
        this.seller.write(w);
        w.write8u(this.quoteTokenId);
        this.price.write(w);
        w.write8(this.startDate);
        w.write8(this.endDate);
    }
    read(r) {
        this.type = r.read1();
        this.seller = Bytes32.read(r);
        this.quoteTokenId = r.read8u();
        this.price = r.readBlob(IntX);
        this.startDate = r.read8();
        this.endDate = r.read8();
    }
    static read(r) {
        const v = new TokenListing();
        v.read(r);
        return v;
    }
}
