// phantasma-sdk-ts/dist/esm/core/types/Carbon/Blockchain/Modules/TokenContract_Methods.js
export var TokenContract_Methods;
(function (TokenContract_Methods) {
    TokenContract_Methods[TokenContract_Methods["TransferFungible"] = 0] = "TransferFungible";
    TokenContract_Methods[TokenContract_Methods["TransferNonFungible"] = 1] = "TransferNonFungible";
    TokenContract_Methods[TokenContract_Methods["CreateToken"] = 2] = "CreateToken";
    TokenContract_Methods[TokenContract_Methods["MintFungible"] = 3] = "MintFungible";
    TokenContract_Methods[TokenContract_Methods["BurnFungible"] = 4] = "BurnFungible";
    TokenContract_Methods[TokenContract_Methods["GetBalance"] = 5] = "GetBalance";
    TokenContract_Methods[TokenContract_Methods["CreateTokenSeries"] = 6] = "CreateTokenSeries";
    TokenContract_Methods[TokenContract_Methods["DeleteTokenSeries"] = 7] = "DeleteTokenSeries";
    TokenContract_Methods[TokenContract_Methods["MintNonFungible"] = 8] = "MintNonFungible";
    TokenContract_Methods[TokenContract_Methods["BurnNonFungible"] = 9] = "BurnNonFungible";
    TokenContract_Methods[TokenContract_Methods["GetInstances"] = 10] = "GetInstances";
    TokenContract_Methods[TokenContract_Methods["GetNonFungibleInfo"] = 11] = "GetNonFungibleInfo";
    TokenContract_Methods[TokenContract_Methods["GetNonFungibleInfoByRomId"] = 12] = "GetNonFungibleInfoByRomId";
    TokenContract_Methods[TokenContract_Methods["GetSeriesInfo"] = 13] = "GetSeriesInfo";
    TokenContract_Methods[TokenContract_Methods["GetSeriesInfoByMetaId"] = 14] = "GetSeriesInfoByMetaId";
    TokenContract_Methods[TokenContract_Methods["GetTokenInfo"] = 15] = "GetTokenInfo";
    TokenContract_Methods[TokenContract_Methods["GetTokenInfoBySymbol"] = 16] = "GetTokenInfoBySymbol";
    TokenContract_Methods[TokenContract_Methods["GetTokenSupply"] = 17] = "GetTokenSupply";
    TokenContract_Methods[TokenContract_Methods["GetSeriesSupply"] = 18] = "GetSeriesSupply";
    TokenContract_Methods[TokenContract_Methods["GetTokenIdBySymbol"] = 19] = "GetTokenIdBySymbol";
    TokenContract_Methods[TokenContract_Methods["GetBalances"] = 20] = "GetBalances";
    TokenContract_Methods[TokenContract_Methods["CreateMintedTokenSeries"] = 21] = "CreateMintedTokenSeries";
    TokenContract_Methods[TokenContract_Methods["ApplyInflation"] = 22] = "ApplyInflation";
    TokenContract_Methods[TokenContract_Methods["UpdateTokenMetadata"] = 23] = "UpdateTokenMetadata";
    TokenContract_Methods[TokenContract_Methods["GetNextTokenInflation"] = 24] = "GetNextTokenInflation";
    TokenContract_Methods[TokenContract_Methods["SetTokensConfig"] = 25] = "SetTokensConfig";
    TokenContract_Methods[TokenContract_Methods["UpdateSeriesMetadata"] = 26] = "UpdateSeriesMetadata";
    TokenContract_Methods[TokenContract_Methods["MintPhantasmaNonFungible"] = 27] = "MintPhantasmaNonFungible";
})(TokenContract_Methods || (TokenContract_Methods = {}));
