// phantasma-sdk-ts/dist/esm/core/types/Carbon/Blockchain/Modules/Builders/TokenSchemasBuilder.js
import { bytesToHex } from '../../../../../utils/index.js';
import { CarbonBinaryWriter } from '../../../../CarbonSerialization.js';
import { VmNamedVariableSchema } from '../../Vm/VmNamedVariableSchema.js';
import { VmStructSchema } from '../../Vm/VmStructSchema.js';
import { vmTypeFromString, VmType } from '../../Vm/VmType.js';
import { TokenSchemas } from '../TokenSchemas.js';
import { nftDefaultMetadataFields, seriesDefaultMetadataFields, standardMetadataFields } from './MetadataHelper.js';
import { logger } from '../../../../../utils/logger.js';
export class TokenSchemasJson {
}
export function parseTokenSchemasJson(json) {
    const raw = JSON.parse(json);
    const pickArray = (k) => {
        const arr = raw[k];
        if (!Array.isArray(arr))
            throw new Error(`${k} must be an array`);
        return arr.map((it) => {
            if (typeof it?.name !== 'string')
                throw new Error(`${k} field name must be string`);
            if (typeof it?.type !== 'string')
                throw new Error(`${k} field type must be string`);
            return { name: it.name, type: vmTypeFromString(it.type) };
        });
    };
    return {
        seriesMetadata: pickArray('seriesMetadata'),
        rom: pickArray('rom'),
        ram: pickArray('ram'),
    };
}
export class TokenSchemasBuilder {
    static assertMetadataField(schemas, fieldTypes) {
        for (const fieldType of fieldTypes) {
            let fieldIsFound = false;
            let caseMismatch = null;
            for (const schema of schemas) {
                const found = schema.fields.find(x => x.name.data === fieldType.name);
                if (found != undefined) {
                    if (found.schema.type != fieldType.type) {
                        const actualType = VmType[found.schema.type] ?? found.schema.type;
                        const expectedType = VmType[fieldType.type] ?? fieldType.type;
                        return {
                            ok: false,
                            error: `Type mismatch for ${fieldType.name} field, must be ${actualType} instead of ${expectedType}`
                        };
                    }
                    fieldIsFound = true;
                    break;
                }
                if (!caseMismatch) {
                    const found2 = schema.fields.find(x => x.name.data.toLowerCase() === fieldType.name.toLowerCase());
                    if (found2 != undefined) {
                        caseMismatch = found2;
                    }
                }
            }
            if (!fieldIsFound) {
                if (caseMismatch) {
                    return {
                        ok: false,
                        error: `Case mismatch for ${fieldType.name} field, must be ${caseMismatch.name.data}`
                    };
                }
                return { ok: false, error: `Mandatory metadata field not found: ${fieldType.name}` };
            }
        }
        return { ok: true, error: null };
    }
    static defaultSeriesSchema(sharedMetadata) {
        // Series metadata schema
        const schema = new VmStructSchema();
        schema.fields = [];
        {
            // Default fields
            seriesDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
            });
            if (sharedMetadata) {
                standardMetadataFields.forEach(f => {
                    schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
                });
            }
        }
        return schema;
    }
    static defaultNftRomSchema(sharedMetadata) {
        // NFT ROM schema
        const schema = new VmStructSchema();
        schema.fields = [];
        {
            // Default fields
            nftDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
            });
            if (!sharedMetadata) {
                // They are default for all NFTs
                standardMetadataFields.forEach(f => {
                    schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
                });
            }
        }
        return schema;
    }
    static seriesSchemaFromFieldTypes(fieldTypes) {
        // Series metadata schema
        const schema = new VmStructSchema();
        schema.fields = [];
        {
            // Default fields. Adding them always
            seriesDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
            });
            fieldTypes.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
            });
        }
        return schema;
    }
    static nftRomSchemaFromFieldTypes(fieldTypes) {
        // NFT ROM schema
        const schema = new VmStructSchema();
        schema.fields = [];
        {
            // Default fields. Adding them always
            nftDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
            });
            fieldTypes.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
            });
        }
        return schema;
    }
    static nftRamSchemaFromFieldTypes(fieldTypes) {
        if (fieldTypes.length == 0) {
            return new VmStructSchema([], VmStructSchema.Flags.DynamicExtras);
        }
        // RAM schema
        const schema = new VmStructSchema();
        schema.fields = [];
        {
            fieldTypes.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema(f.name, f.type));
            });
        }
        return schema;
    }
    static prepareStandard(sharedMetadata) {
        const tokenSchemas = new TokenSchemas();
        // Token's series metadata schema
        tokenSchemas.seriesMetadata = this.defaultSeriesSchema(sharedMetadata);
        // NFT's ROM schema
        tokenSchemas.rom = this.defaultNftRomSchema(sharedMetadata);
        // NFT's RAM schema
        tokenSchemas.ram = new VmStructSchema([], VmStructSchema.Flags.DynamicExtras);
        return tokenSchemas;
    }
    static fromJson(json) {
        const tokenSchemasJson = parseTokenSchemasJson(json);
        const tokenSchemas = new TokenSchemas();
        tokenSchemas.seriesMetadata = this.seriesSchemaFromFieldTypes(tokenSchemasJson.seriesMetadata);
        tokenSchemas.rom = this.nftRomSchemaFromFieldTypes(tokenSchemasJson.rom);
        tokenSchemas.ram = this.nftRamSchemaFromFieldTypes(tokenSchemasJson.ram);
        const { ok, error } = this.verify(tokenSchemas);
        if (!ok) {
            throw Error(error ?? "Unknown error");
        }
        return tokenSchemas;
    }
    static verify(schemas) {
        let result = this.assertMetadataField([schemas.seriesMetadata, schemas.rom], standardMetadataFields);
        if (!result.ok) {
            return result;
        }
        result = this.assertMetadataField([schemas.seriesMetadata], seriesDefaultMetadataFields);
        if (!result.ok) {
            return result;
        }
        result = this.assertMetadataField([schemas.rom], nftDefaultMetadataFields);
        if (!result.ok) {
            return result;
        }
        return { ok: true, error: null };
    }
    static serialize(tokenSchemas) {
        const schemaBuf = new CarbonBinaryWriter();
        tokenSchemas.write(schemaBuf);
        return schemaBuf.toUint8Array();
    }
    static serializeHex(tokenSchemas) {
        return bytesToHex(this.serialize(tokenSchemas));
    }
    static logSchema(schema) {
        schema.fields.forEach(f => logger.log("Schema field: ", f.name.data));
    }
}
