// phantasma-sdk-ts/dist/esm/core/types/Carbon/Bytes16.js
import { bytesToHex } from '../../utils/Hex.js';
export class Bytes16 {
    constructor(bytes = new Uint8Array(16)) {
        this.bytes = bytes;
        if (bytes.length !== 16)
            throw new Error('Bytes16 length must be 16');
        this.bytes = new Uint8Array(bytes);
    }
    write(w) {
        w.write16(this.bytes);
    }
    read(r) {
        this.bytes = r.read16();
    }
    static read(r) {
        const v = new Bytes16();
        v.read(r);
        return v;
    }
    equals(other) {
        const a = this.bytes, b = other.bytes;
        if (a.length !== b.length)
            return false;
        for (let i = 0; i < a.length; i++)
            if (a[i] !== b[i])
                return false;
        return true;
    }
    ToHex() {
        return bytesToHex(this.bytes);
    }
    // Used by console.log / util.inspect
    [Symbol.for('nodejs.util.inspect.custom')]() {
        // Return a pretty, concise representation
        return `Bytes16(${this.ToHex()})`;
    }
}
Bytes16.Empty = new Bytes16(new Uint8Array(16));
