// phantasma-sdk-ts/dist/esm/core/types/Carbon/Bytes64.js
import { bytesToHex } from '../../utils/Hex.js';
export class Bytes64 {
    constructor(bytes = new Uint8Array(64)) {
        this.bytes = bytes;
        if (bytes.length !== 64)
            throw new Error('Bytes64 length must be 64');
        this.bytes = new Uint8Array(bytes);
    }
    write(w) {
        w.write64(this.bytes);
    }
    read(r) {
        this.bytes = r.read64();
    }
    static read(r) {
        const v = new Bytes64();
        v.read(r);
        return v;
    }
    equals(other) {
        const a = this.bytes, b = other.bytes;
        if (a.length !== b.length)
            return false;
        for (let i = 0; i < a.length; i++)
            if (a[i] !== b[i])
                return false;
        return true;
    }
    ToHex() {
        return bytesToHex(this.bytes);
    }
    // Used by console.log / util.inspect
    [Symbol.for('nodejs.util.inspect.custom')]() {
        // Return a pretty, concise representation
        return `Bytes64(${this.ToHex()})`;
    }
}
Bytes64.Empty = new Bytes64(new Uint8Array(64));
