// phantasma-sdk-ts/dist/esm/core/types/Carbon/Bytes32.js
import { bytesToHex } from '../../utils/Hex.js';
export class Bytes32 {
    constructor(bytes = new Uint8Array(32)) {
        this.bytes = bytes;
        if (bytes.length !== 32)
            throw new Error('Bytes32 length must be 32');
        this.bytes = new Uint8Array(bytes);
    }
    write(w) {
        w.write32(this.bytes);
    }
    read(r) {
        this.bytes = r.read32();
    }
    static read(r) {
        const v = new Bytes32();
        v.read(r);
        return v;
    }
    equals(other) {
        const a = this.bytes, b = other.bytes;
        if (a.length !== b.length)
            return false;
        for (let i = 0; i < a.length; i++)
            if (a[i] !== b[i])
                return false;
        return true;
    }
    ToHex() {
        return bytesToHex(this.bytes);
    }
    // Used by console.log / util.inspect
    [Symbol.for('nodejs.util.inspect.custom')]() {
        // Return a pretty, concise representation
        return `Bytes32(${this.ToHex()})`;
    }
}
Bytes32.Empty = new Bytes32(new Uint8Array(32));
