// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/TxHelpers/FeeOptions.js
"use strict";
// namespace PhantasmaPhoenix.Protocol.Carbon.Blockchain.TxHelpers
Object.defineProperty(exports, "__esModule", { value: true });
exports.MintNftFeeOptions = exports.CreateSeriesFeeOptions = exports.CreateTokenFeeOptions = exports.FeeOptions = void 0;
/** Base fee options with sensible defaults. */
class FeeOptions {
    constructor(gasFeeBase = 10000n, feeMultiplier = 1000n) {
        this.gasFeeBase = gasFeeBase;
        this.feeMultiplier = feeMultiplier;
    }
    calculateMaxGas(..._args) {
        return this.gasFeeBase * this.feeMultiplier;
    }
}
exports.FeeOptions = FeeOptions;
/** Fee options for token creation transactions. */
class CreateTokenFeeOptions extends FeeOptions {
    constructor(gasFeeBase = 10000n, gasFeeCreateTokenBase = 10000000000n, gasFeeCreateTokenSymbol = 10000000000n, feeMultiplier = 10000n) {
        super(gasFeeBase, feeMultiplier);
        this.gasFeeCreateTokenBase = gasFeeCreateTokenBase;
        this.gasFeeCreateTokenSymbol = gasFeeCreateTokenSymbol;
    }
    calculateMaxGas(...args) {
        let symbol = '';
        if (args.length > 0 && typeof args[0] === 'object' && args[0] !== null && 'data' in args[0]) {
            const s = args[0];
            symbol = s.data;
        }
        const symbolLen = symbol.length;
        const shift = symbolLen > 0 ? BigInt(symbolLen - 1) : 0n;
        const symbolCost = this.gasFeeCreateTokenSymbol >> shift;
        return (this.gasFeeBase + this.gasFeeCreateTokenBase + symbolCost) * this.feeMultiplier;
    }
}
exports.CreateTokenFeeOptions = CreateTokenFeeOptions;
/** Fee options for creating a new series on an NFT token. */
class CreateSeriesFeeOptions extends FeeOptions {
    constructor(gasFeeBase = 10000n, gasFeeCreateSeriesBase = 2500000000n, feeMultiplier = 10000n) {
        super(gasFeeBase, feeMultiplier);
        this.gasFeeCreateSeriesBase = gasFeeCreateSeriesBase;
    }
    calculateMaxGas(..._args) {
        return (this.gasFeeBase + this.gasFeeCreateSeriesBase) * this.feeMultiplier;
    }
}
exports.CreateSeriesFeeOptions = CreateSeriesFeeOptions;
/** Fee options for minting non-fungible tokens (NFT instances). */
class MintNftFeeOptions extends FeeOptions {
    constructor(gasFeeBase = 10000n, feeMultiplier = 1000n) {
        super(gasFeeBase, feeMultiplier);
    }
    calculateMaxGas(..._args) {
        return this.gasFeeBase * this.feeMultiplier;
    }
}
exports.MintNftFeeOptions = MintNftFeeOptions;
