// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/TxHelpers/MintPhantasmaNonFungibleTxHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MintPhantasmaNonFungibleTxHelper = void 0;
const index_js_1 = require("../../../../utils/index.js");
const CarbonSerialization_js_1 = require("../../../CarbonSerialization.js");
const Bytes32_js_1 = require("../../Bytes32.js");
const IntX_js_1 = require("../../IntX.js");
const SmallString_js_1 = require("../../SmallString.js");
const TxTypes_js_1 = require("../../TxTypes.js");
const TxMsgSigner_js_1 = require("../Extensions/TxMsgSigner.js");
const ModuleId_js_1 = require("../ModuleId.js");
const index_js_2 = require("../Modules/index.js");
const TxMsg_js_1 = require("../TxMsg.js");
const TxMsgCall_js_1 = require("../TxMsgCall.js");
const FeeOptions_js_1 = require("./FeeOptions.js");
class MintPhantasmaNonFungibleTxHelper {
    static buildTx(tokenId, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8) {
        if (typeof arg1 === 'bigint') {
            if (!(arg3 instanceof Bytes32_js_1.Bytes32)) {
                throw new Error('receiverPublicKey is required for single-item MintPhantasmaNonFungible tx');
            }
            const publicRom = arg4;
            const ram = arg5;
            if (!(publicRom instanceof Uint8Array)) {
                throw new Error('publicRom must be a Uint8Array for single-item MintPhantasmaNonFungible tx');
            }
            if (!(ram === undefined || ram instanceof Uint8Array)) {
                throw new Error('ram must be a Uint8Array when provided for single-item MintPhantasmaNonFungible tx');
            }
            const safeRam = ram instanceof Uint8Array ? ram : new Uint8Array();
            return this.buildTxCore(tokenId, arg2, arg3, [
                new index_js_2.PhantasmaNftMintInfo({
                    phantasmaSeriesId: IntX_js_1.IntX.fromBigInt(arg1),
                    rom: publicRom,
                    ram: safeRam,
                }),
            ], arg6 instanceof FeeOptions_js_1.MintNftFeeOptions ? arg6 : undefined, typeof arg7 === 'bigint' ? arg7 : undefined, typeof arg8 === 'bigint' ? arg8 : undefined);
        }
        if (arg3 instanceof Bytes32_js_1.Bytes32 || !Array.isArray(arg3)) {
            throw new Error('tokens array is required for multi-item MintPhantasmaNonFungible tx');
        }
        return this.buildTxCore(tokenId, arg1, arg2, arg3, arg4 instanceof FeeOptions_js_1.MintNftFeeOptions ? arg4 : undefined, typeof arg5 === 'bigint' ? arg5 : undefined, typeof arg6 === 'bigint' ? arg6 : undefined);
    }
    static buildTxCore(tokenId, senderPublicKey, receiverPublicKey, tokens, feeOptions, maxData, expiry) {
        const fees = feeOptions ?? new FeeOptions_js_1.MintNftFeeOptions();
        const maxGas = fees.calculateMaxGas();
        // This helper only packages the Token.Call ABI surface.
        const args = new index_js_2.MintPhantasmaNonFungibleArgs({
            tokenId,
            address: receiverPublicKey,
            tokens,
        });
        const msg = new TxMsg_js_1.TxMsg();
        msg.type = TxTypes_js_1.TxTypes.Call;
        msg.expiry = expiry || BigInt(Date.now() + 60000);
        msg.maxGas = maxGas;
        msg.maxData = maxData;
        msg.gasFrom = senderPublicKey;
        msg.payload = SmallString_js_1.SmallString.empty;
        const call = new TxMsgCall_js_1.TxMsgCall();
        call.moduleId = ModuleId_js_1.ModuleId.Token;
        call.methodId = index_js_2.TokenContract_Methods.MintPhantasmaNonFungible;
        const argsWriter = new CarbonSerialization_js_1.CarbonBinaryWriter();
        args.write(argsWriter);
        call.args = argsWriter.toUint8Array();
        msg.msg = call;
        return msg;
    }
    static buildTxAndSign(tokenId, arg1, signer, receiverPublicKey, arg4, arg5, arg6, arg7, arg8) {
        const senderPub = new Bytes32_js_1.Bytes32(signer.PublicKey);
        if (typeof arg1 === 'bigint') {
            const tx = this.buildTx(tokenId, arg1, senderPub, receiverPublicKey, arg4, arg5 instanceof Uint8Array ? arg5 : undefined, arg6 instanceof FeeOptions_js_1.MintNftFeeOptions ? arg6 : undefined, typeof arg7 === 'bigint' ? arg7 : undefined, typeof arg8 === 'bigint' ? arg8 : undefined);
            return TxMsgSigner_js_1.TxMsgSigner.signAndSerialize(tx, signer);
        }
        const tx = this.buildTx(tokenId, senderPub, receiverPublicKey, arg1, arg4 instanceof FeeOptions_js_1.MintNftFeeOptions ? arg4 : undefined, typeof arg5 === 'bigint' ? arg5 : undefined, typeof arg6 === 'bigint' ? arg6 : undefined);
        return TxMsgSigner_js_1.TxMsgSigner.signAndSerialize(tx, signer);
    }
    static buildTxAndSignHex(tokenId, arg1, signer, receiverPublicKey, arg4, arg5, arg6, arg7, arg8) {
        if (typeof arg1 === 'bigint') {
            return (0, index_js_1.bytesToHex)(this.buildTxAndSign(tokenId, arg1, signer, receiverPublicKey, arg4, arg5 instanceof Uint8Array ? arg5 : undefined, arg6 instanceof FeeOptions_js_1.MintNftFeeOptions ? arg6 : undefined, typeof arg7 === 'bigint' ? arg7 : undefined, typeof arg8 === 'bigint' ? arg8 : undefined));
        }
        return (0, index_js_1.bytesToHex)(this.buildTxAndSign(tokenId, arg1, signer, receiverPublicKey, arg4 instanceof FeeOptions_js_1.MintNftFeeOptions ? arg4 : undefined, typeof arg5 === 'bigint' ? arg5 : undefined, typeof arg6 === 'bigint' ? arg6 : undefined));
    }
    static parseResult(resultHex) {
        const r = new CarbonSerialization_js_1.CarbonBinaryReader((0, index_js_1.hexToBytes)(resultHex));
        return r.readArrayBlob(index_js_2.PhantasmaNftMintResult);
    }
}
exports.MintPhantasmaNonFungibleTxHelper = MintPhantasmaNonFungibleTxHelper;
