// phantasma-sdk-ts/dist/esm/core/types/Carbon/Blockchain/Vm/VmType.js
export var VmType;
(function (VmType) {
    VmType[VmType["Dynamic"] = 0] = "Dynamic";
    VmType[VmType["Array"] = 1] = "Array";
    VmType[VmType["Bytes"] = 2] = "Bytes";
    VmType[VmType["Struct"] = 4] = "Struct";
    VmType[VmType["Int8"] = 6] = "Int8";
    VmType[VmType["Int16"] = 8] = "Int16";
    VmType[VmType["Int32"] = 10] = "Int32";
    VmType[VmType["Int64"] = 12] = "Int64";
    VmType[VmType["Int256"] = 14] = "Int256";
    VmType[VmType["Bytes16"] = 16] = "Bytes16";
    VmType[VmType["Bytes32"] = 18] = "Bytes32";
    VmType[VmType["Bytes64"] = 20] = "Bytes64";
    VmType[VmType["String"] = 22] = "String";
    VmType[VmType["Array_Dynamic"] = 1] = "Array_Dynamic";
    VmType[VmType["Array_Bytes"] = 3] = "Array_Bytes";
    VmType[VmType["Array_Struct"] = 5] = "Array_Struct";
    VmType[VmType["Array_Int8"] = 7] = "Array_Int8";
    VmType[VmType["Array_Int16"] = 9] = "Array_Int16";
    VmType[VmType["Array_Int32"] = 11] = "Array_Int32";
    VmType[VmType["Array_Int64"] = 13] = "Array_Int64";
    VmType[VmType["Array_Int256"] = 15] = "Array_Int256";
    VmType[VmType["Array_Bytes16"] = 17] = "Array_Bytes16";
    VmType[VmType["Array_Bytes32"] = 19] = "Array_Bytes32";
    VmType[VmType["Array_Bytes64"] = 21] = "Array_Bytes64";
    VmType[VmType["Array_String"] = 23] = "Array_String";
})(VmType || (VmType = {}));
export function vmTypeFromString(type) {
    const map = {
        Dynamic: VmType.Dynamic,
        Array: VmType.Array,
        Bytes: VmType.Bytes,
        Struct: VmType.Struct,
        Int8: VmType.Int8,
        Int16: VmType.Int16,
        Int32: VmType.Int32,
        Int64: VmType.Int64,
        Int256: VmType.Int256,
        Bytes16: VmType.Bytes16,
        Bytes32: VmType.Bytes32,
        Bytes64: VmType.Bytes64,
        String: VmType.String,
        Array_Dynamic: VmType.Array_Dynamic,
        Array_Bytes: VmType.Array_Bytes,
        Array_Struct: VmType.Array_Struct,
        Array_Int8: VmType.Array_Int8,
        Array_Int16: VmType.Array_Int16,
        Array_Int32: VmType.Array_Int32,
        Array_Int64: VmType.Array_Int64,
        Array_Int256: VmType.Array_Int256,
        Array_Bytes16: VmType.Array_Bytes16,
        Array_Bytes32: VmType.Array_Bytes32,
        Array_Bytes64: VmType.Array_Bytes64,
        Array_String: VmType.Array_String,
    };
    const value = map[type.trim()];
    if (value === undefined) {
        throw new Error(`Unknown VmType: ${type}`);
    }
    return value;
}
