// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/Modules/Builders/MetadataHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pushMetadataField = exports.findMetadataField = exports.standardMetadataFields = exports.nftDefaultMetadataFields = exports.seriesDefaultMetadataFields = exports.FieldType = exports.MetadataField = void 0;
const index_js_1 = require("../../../../../utils/index.js");
const Bytes16_js_1 = require("../../../Bytes16.js");
const Bytes32_js_1 = require("../../../Bytes32.js");
const Bytes64_js_1 = require("../../../Bytes64.js");
const index_js_2 = require("../../Vm/index.js");
const StandardMeta_js_1 = require("../StandardMeta.js");
const INT64_MIN = -(1n << 63n);
const INT64_MAX = (1n << 63n) - 1n;
const UINT64_MAX = (1n << 64n) - 1n;
const INT256_MIN = -(1n << 255n);
const INT256_MAX = (1n << 255n) - 1n;
const UINT256_MAX = (1n << 256n) - 1n;
class MetadataField {
}
exports.MetadataField = MetadataField;
class FieldType {
}
exports.FieldType = FieldType;
exports.seriesDefaultMetadataFields = [
    { name: StandardMeta_js_1.StandardMeta.id.data, type: index_js_2.VmType.Int256 },
    { name: 'mode', type: index_js_2.VmType.Int8 },
    { name: 'rom', type: index_js_2.VmType.Bytes }
];
exports.nftDefaultMetadataFields = [
    { name: StandardMeta_js_1.StandardMeta.id.data, type: index_js_2.VmType.Int256 },
    { name: 'rom', type: index_js_2.VmType.Bytes }
];
exports.standardMetadataFields = [
    { name: 'name', type: index_js_2.VmType.String },
    { name: 'description', type: index_js_2.VmType.String },
    { name: 'imageURL', type: index_js_2.VmType.String },
    { name: 'infoURL', type: index_js_2.VmType.String },
    { name: 'royalties', type: index_js_2.VmType.Int32 }
];
function findMetadataField(fields, name) {
    return fields.find(f => f.name.toLowerCase() === name.toLowerCase());
}
exports.findMetadataField = findMetadataField;
function pushMetadataField(fieldSchema, metadata, metadataFields) {
    let found = metadataFields.find(f => f.name === fieldSchema.name.data);
    if (!found) {
        found = metadataFields.find(f => f.name.toLowerCase() === fieldSchema.name.data.toLowerCase());
        if (found) {
            throw Error(`Metadata field '${fieldSchema.name.data}' provided in incorrect case: '${found.name}'`);
        }
        throw Error(`Metadata field '${fieldSchema.name.data}' is mandatory`);
    }
    const normalizedValue = normalizeMetadataValue(fieldSchema.schema, fieldSchema.name.data, found.value);
    metadata.fields.push(index_js_2.VmNamedDynamicVariable.from(fieldSchema.name, fieldSchema.schema.type, normalizedValue));
}
exports.pushMetadataField = pushMetadataField;
function normalizeMetadataValue(schema, fieldName, value) {
    if (value === null || value === undefined) {
        throw Error(`Metadata field '${fieldName}' is mandatory`);
    }
    const isArray = (schema.type & index_js_2.VmType.Array) === index_js_2.VmType.Array;
    const baseType = (isArray ? (schema.type & ~index_js_2.VmType.Array) : schema.type);
    if (isArray) {
        if (!Array.isArray(value)) {
            throw Error(`Metadata field '${fieldName}' must be provided as an array`);
        }
        return normalizeArrayValue(baseType, fieldName, value, schema.structure);
    }
    return normalizeScalarValue(baseType, fieldName, value, schema.structure);
}
function normalizeScalarValue(type, fieldName, value, structSchema) {
    switch (type) {
        case index_js_2.VmType.String:
            return ensureNonEmptyString(fieldName, value);
        case index_js_2.VmType.Int8:
            return ensureIntegerInRange(fieldName, value, -0x80, 0x7f, 0xff);
        case index_js_2.VmType.Int16:
            return ensureIntegerInRange(fieldName, value, -0x8000, 0x7fff, 0xffff);
        case index_js_2.VmType.Int32:
            return ensureIntegerInRange(fieldName, value, -0x80000000, 0x7fffffff, 0xffffffff);
        case index_js_2.VmType.Int64:
            return ensureBigInt(fieldName, value, INT64_MIN, INT64_MAX, 'Int64', UINT64_MAX);
        case index_js_2.VmType.Int256:
            return ensureBigInt(fieldName, value, INT256_MIN, INT256_MAX, 'Int256', UINT256_MAX);
        case index_js_2.VmType.Bytes:
            return ensureBytes(fieldName, value);
        case index_js_2.VmType.Bytes16:
            return ensureBytes16(fieldName, value);
        case index_js_2.VmType.Bytes32:
            return ensureBytes32(fieldName, value);
        case index_js_2.VmType.Bytes64:
            return ensureBytes64(fieldName, value);
        case index_js_2.VmType.Struct:
            return normalizeStructValue(fieldName, structSchema, value);
        default:
            throw Error(`Metadata field '${fieldName}' has unsupported type '${index_js_2.VmType[type] ?? type}'`);
    }
}
function normalizeArrayValue(type, fieldName, values, structSchema) {
    const elementPath = (index) => `${fieldName}[${index}]`;
    switch (type) {
        case index_js_2.VmType.String:
            return values.map((val, idx) => ensureNonEmptyString(elementPath(idx), val));
        case index_js_2.VmType.Int8:
            return Uint8Array.from(values.map((val, idx) => ensureIntegerInRange(elementPath(idx), val, -0x80, 0x7f, 0xff) & 0xff));
        case index_js_2.VmType.Int16:
            return values.map((val, idx) => ensureIntegerInRange(elementPath(idx), val, -0x8000, 0x7fff, 0xffff));
        case index_js_2.VmType.Int32:
            return values.map((val, idx) => ensureIntegerInRange(elementPath(idx), val, -0x80000000, 0x7fffffff, 0xffffffff));
        case index_js_2.VmType.Int64:
            return values.map((val, idx) => ensureBigInt(elementPath(idx), val, INT64_MIN, INT64_MAX, 'Int64', UINT64_MAX));
        case index_js_2.VmType.Int256:
            return values.map((val, idx) => ensureBigInt(elementPath(idx), val, INT256_MIN, INT256_MAX, 'Int256', UINT256_MAX));
        case index_js_2.VmType.Bytes:
            return values.map((val, idx) => ensureBytes(elementPath(idx), val));
        case index_js_2.VmType.Bytes16:
            return values.map((val, idx) => ensureBytes16(elementPath(idx), val));
        case index_js_2.VmType.Bytes32:
            return values.map((val, idx) => ensureBytes32(elementPath(idx), val));
        case index_js_2.VmType.Bytes64:
            return values.map((val, idx) => ensureBytes64(elementPath(idx), val));
        case index_js_2.VmType.Struct: {
            if (!structSchema) {
                throw Error(`Metadata field '${fieldName}' is missing schema for struct elements`);
            }
            const sa = new index_js_2.VmStructArray();
            sa.schema = structSchema;
            sa.structs = values.map((val, idx) => normalizeStructValue(elementPath(idx), structSchema, val));
            return sa;
        }
        default:
            throw Error(`Metadata field '${fieldName}' with type '${index_js_2.VmType[type] ?? type}' does not support array values`);
    }
}
function normalizeStructValue(fieldName, structSchema, value) {
    if (!structSchema) {
        throw Error(`Metadata field '${fieldName}' is missing struct schema`);
    }
    const struct = new index_js_2.VmDynamicStruct();
    struct.fields = [];
    const providedFields = metadataStructInputToFields(fieldName, value);
    structSchema.fields.forEach(childSchema => {
        let childValue = providedFields.find(f => f.name === childSchema.name.data);
        if (!childValue) {
            childValue = providedFields.find(f => f.name.toLowerCase() === childSchema.name.data.toLowerCase());
            if (childValue) {
                throw Error(`Metadata field '${childSchema.name.data}' provided in incorrect case inside '${fieldName}': '${childValue.name}'`);
            }
            throw Error(`Metadata field '${fieldName}.${childSchema.name.data}' is mandatory`);
        }
        const normalized = normalizeMetadataValue(childSchema.schema, `${fieldName}.${childSchema.name.data}`, childValue.value);
        struct.fields.push(index_js_2.VmNamedDynamicVariable.from(childSchema.name, childSchema.schema.type, normalized));
    });
    const allowedNames = new Set(structSchema.fields.map(f => f.name.data.toLowerCase()));
    for (const provided of providedFields) {
        if (!allowedNames.has(provided.name.toLowerCase())) {
            throw Error(`Metadata field '${fieldName}' received unknown property '${provided.name}'`);
        }
    }
    return struct;
}
function metadataStructInputToFields(fieldName, value) {
    if (Array.isArray(value)) {
        return value;
    }
    if (value instanceof Uint8Array || typeof value !== "object") {
        throw Error(`Metadata field '${fieldName}' must be provided as an object or array of fields`);
    }
    return Object.entries(value).map(([name, val]) => ({
        name,
        value: val
    }));
}
function ensureNonEmptyString(fieldName, value) {
    if (typeof value !== "string") {
        throw Error(`Metadata field '${fieldName}' must be a string`);
    }
    if (value.trim().length === 0) {
        throw Error(`Metadata field '${fieldName}' is mandatory`);
    }
    return value;
}
function ensureIntegerInRange(fieldName, value, min, max, unsignedMax) {
    if (typeof value !== "number" || !Number.isInteger(value)) {
        const rangeText = unsignedMax !== undefined
            ? `an integer between ${min} and ${max} or between 0 and ${unsignedMax}`
            : `an integer between ${min} and ${max}`;
        throw Error(`Metadata field '${fieldName}' must be ${rangeText}`);
    }
    const fitsSigned = value >= min && value <= max;
    const fitsUnsigned = unsignedMax !== undefined && value >= 0 && value <= unsignedMax;
    if (!fitsSigned && !fitsUnsigned) {
        const rangeText = unsignedMax !== undefined
            ? `between ${min} and ${max} or between 0 and ${unsignedMax}`
            : `between ${min} and ${max}`;
        throw Error(`Metadata field '${fieldName}' must be ${rangeText}`);
    }
    return value;
}
function ensureBigInt(fieldName, value, min, max, label, unsignedMax) {
    let bigintValue;
    if (typeof value === "bigint") {
        bigintValue = value;
    }
    else if (typeof value === "number" && Number.isInteger(value)) {
        if (!Number.isSafeInteger(value)) {
            throw Error(`Metadata field '${fieldName}' must be provided as a bigint when it exceeds safe integer range`);
        }
        bigintValue = BigInt(value);
    }
    else {
        throw Error(`Metadata field '${fieldName}' must be a bigint or a safe integer number`);
    }
    const hasSignedRange = min !== undefined && max !== undefined;
    const fitsSigned = hasSignedRange ? bigintValue >= min && bigintValue <= max : false;
    const hasUnsignedRange = unsignedMax !== undefined;
    const fitsUnsigned = hasUnsignedRange ? bigintValue >= 0 && bigintValue <= unsignedMax : false;
    if (!fitsSigned && !fitsUnsigned) {
        const signedPart = hasSignedRange ? `between ${min.toString()} and ${max.toString()}` : null;
        const unsignedPart = hasUnsignedRange ? `between 0 and ${unsignedMax.toString()}` : null;
        const rangeText = [signedPart, unsignedPart].filter(Boolean).join(' or ');
        const labelSuffix = label ? ` (${label})` : '';
        throw Error(`Metadata field '${fieldName}' must be ${rangeText}${labelSuffix}`);
    }
    return bigintValue;
}
function ensureBytes(fieldName, value) {
    if (value instanceof Uint8Array) {
        return value;
    }
    if (typeof value === "string") {
        const trimmed = value.trim();
        if (trimmed.length === 0) {
            throw Error(`Metadata field '${fieldName}' must be a byte array or hex string`);
        }
        try {
            return (0, index_js_1.hexToBytes)(trimmed);
        }
        catch {
            throw Error(`Metadata field '${fieldName}' must be a byte array or hex string`);
        }
    }
    throw Error(`Metadata field '${fieldName}' must be a byte array or hex string`);
}
function ensureFixedBytes(fieldName, value, expectedLength) {
    const bytes = ensureBytes(fieldName, value);
    if (bytes.length !== expectedLength) {
        throw Error(`Metadata field '${fieldName}' must be exactly ${expectedLength} bytes`);
    }
    return bytes;
}
function ensureBytes16(fieldName, value) {
    return new Bytes16_js_1.Bytes16(ensureFixedBytes(fieldName, value, 16));
}
function ensureBytes32(fieldName, value) {
    return new Bytes32_js_1.Bytes32(ensureFixedBytes(fieldName, value, 32));
}
function ensureBytes64(fieldName, value) {
    return new Bytes64_js_1.Bytes64(ensureFixedBytes(fieldName, value, 64));
}
