// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/Modules/Builders/TokenSchemasBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenSchemasBuilder = exports.parseTokenSchemasJson = exports.TokenSchemasJson = void 0;
const index_js_1 = require("../../../../../utils/index.js");
const CarbonSerialization_js_1 = require("../../../../CarbonSerialization.js");
const VmNamedVariableSchema_js_1 = require("../../Vm/VmNamedVariableSchema.js");
const VmStructSchema_js_1 = require("../../Vm/VmStructSchema.js");
const VmType_js_1 = require("../../Vm/VmType.js");
const TokenSchemas_js_1 = require("../TokenSchemas.js");
const MetadataHelper_js_1 = require("./MetadataHelper.js");
const logger_js_1 = require("../../../../../utils/logger.js");
class TokenSchemasJson {
}
exports.TokenSchemasJson = TokenSchemasJson;
function parseTokenSchemasJson(json) {
    const raw = JSON.parse(json);
    const pickArray = (k) => {
        const arr = raw[k];
        if (!Array.isArray(arr))
            throw new Error(`${k} must be an array`);
        return arr.map((it) => {
            if (typeof it?.name !== 'string')
                throw new Error(`${k} field name must be string`);
            if (typeof it?.type !== 'string')
                throw new Error(`${k} field type must be string`);
            return { name: it.name, type: (0, VmType_js_1.vmTypeFromString)(it.type) };
        });
    };
    return {
        seriesMetadata: pickArray('seriesMetadata'),
        rom: pickArray('rom'),
        ram: pickArray('ram'),
    };
}
exports.parseTokenSchemasJson = parseTokenSchemasJson;
class TokenSchemasBuilder {
    static assertMetadataField(schemas, fieldTypes) {
        for (const fieldType of fieldTypes) {
            let fieldIsFound = false;
            let caseMismatch = null;
            for (const schema of schemas) {
                const found = schema.fields.find(x => x.name.data === fieldType.name);
                if (found != undefined) {
                    if (found.schema.type != fieldType.type) {
                        const actualType = VmType_js_1.VmType[found.schema.type] ?? found.schema.type;
                        const expectedType = VmType_js_1.VmType[fieldType.type] ?? fieldType.type;
                        return {
                            ok: false,
                            error: `Type mismatch for ${fieldType.name} field, must be ${actualType} instead of ${expectedType}`
                        };
                    }
                    fieldIsFound = true;
                    break;
                }
                if (!caseMismatch) {
                    const found2 = schema.fields.find(x => x.name.data.toLowerCase() === fieldType.name.toLowerCase());
                    if (found2 != undefined) {
                        caseMismatch = found2;
                    }
                }
            }
            if (!fieldIsFound) {
                if (caseMismatch) {
                    return {
                        ok: false,
                        error: `Case mismatch for ${fieldType.name} field, must be ${caseMismatch.name.data}`
                    };
                }
                return { ok: false, error: `Mandatory metadata field not found: ${fieldType.name}` };
            }
        }
        return { ok: true, error: null };
    }
    static defaultSeriesSchema(sharedMetadata) {
        // Series metadata schema
        const schema = new VmStructSchema_js_1.VmStructSchema();
        schema.fields = [];
        {
            // Default fields
            MetadataHelper_js_1.seriesDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
            });
            if (sharedMetadata) {
                MetadataHelper_js_1.standardMetadataFields.forEach(f => {
                    schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
                });
            }
        }
        return schema;
    }
    static defaultNftRomSchema(sharedMetadata) {
        // NFT ROM schema
        const schema = new VmStructSchema_js_1.VmStructSchema();
        schema.fields = [];
        {
            // Default fields
            MetadataHelper_js_1.nftDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
            });
            if (!sharedMetadata) {
                // They are default for all NFTs
                MetadataHelper_js_1.standardMetadataFields.forEach(f => {
                    schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
                });
            }
        }
        return schema;
    }
    static seriesSchemaFromFieldTypes(fieldTypes) {
        // Series metadata schema
        const schema = new VmStructSchema_js_1.VmStructSchema();
        schema.fields = [];
        {
            // Default fields. Adding them always
            MetadataHelper_js_1.seriesDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
            });
            fieldTypes.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
            });
        }
        return schema;
    }
    static nftRomSchemaFromFieldTypes(fieldTypes) {
        // NFT ROM schema
        const schema = new VmStructSchema_js_1.VmStructSchema();
        schema.fields = [];
        {
            // Default fields. Adding them always
            MetadataHelper_js_1.nftDefaultMetadataFields.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
            });
            fieldTypes.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
            });
        }
        return schema;
    }
    static nftRamSchemaFromFieldTypes(fieldTypes) {
        if (fieldTypes.length == 0) {
            return new VmStructSchema_js_1.VmStructSchema([], VmStructSchema_js_1.VmStructSchema.Flags.DynamicExtras);
        }
        // RAM schema
        const schema = new VmStructSchema_js_1.VmStructSchema();
        schema.fields = [];
        {
            fieldTypes.forEach(f => {
                schema.fields.push(new VmNamedVariableSchema_js_1.VmNamedVariableSchema(f.name, f.type));
            });
        }
        return schema;
    }
    static prepareStandard(sharedMetadata) {
        const tokenSchemas = new TokenSchemas_js_1.TokenSchemas();
        // Token's series metadata schema
        tokenSchemas.seriesMetadata = this.defaultSeriesSchema(sharedMetadata);
        // NFT's ROM schema
        tokenSchemas.rom = this.defaultNftRomSchema(sharedMetadata);
        // NFT's RAM schema
        tokenSchemas.ram = new VmStructSchema_js_1.VmStructSchema([], VmStructSchema_js_1.VmStructSchema.Flags.DynamicExtras);
        return tokenSchemas;
    }
    static fromJson(json) {
        const tokenSchemasJson = parseTokenSchemasJson(json);
        const tokenSchemas = new TokenSchemas_js_1.TokenSchemas();
        tokenSchemas.seriesMetadata = this.seriesSchemaFromFieldTypes(tokenSchemasJson.seriesMetadata);
        tokenSchemas.rom = this.nftRomSchemaFromFieldTypes(tokenSchemasJson.rom);
        tokenSchemas.ram = this.nftRamSchemaFromFieldTypes(tokenSchemasJson.ram);
        const { ok, error } = this.verify(tokenSchemas);
        if (!ok) {
            throw Error(error ?? "Unknown error");
        }
        return tokenSchemas;
    }
    static verify(schemas) {
        let result = this.assertMetadataField([schemas.seriesMetadata, schemas.rom], MetadataHelper_js_1.standardMetadataFields);
        if (!result.ok) {
            return result;
        }
        result = this.assertMetadataField([schemas.seriesMetadata], MetadataHelper_js_1.seriesDefaultMetadataFields);
        if (!result.ok) {
            return result;
        }
        result = this.assertMetadataField([schemas.rom], MetadataHelper_js_1.nftDefaultMetadataFields);
        if (!result.ok) {
            return result;
        }
        return { ok: true, error: null };
    }
    static serialize(tokenSchemas) {
        const schemaBuf = new CarbonSerialization_js_1.CarbonBinaryWriter();
        tokenSchemas.write(schemaBuf);
        return schemaBuf.toUint8Array();
    }
    static serializeHex(tokenSchemas) {
        return (0, index_js_1.bytesToHex)(this.serialize(tokenSchemas));
    }
    static logSchema(schema) {
        schema.fields.forEach(f => logger_js_1.logger.log("Schema field: ", f.name.data));
    }
}
exports.TokenSchemasBuilder = TokenSchemasBuilder;
