// phantasma-sdk-ts/dist/cjs/core/link/easyConnect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EasyConnect = void 0;
const phantasmaLink_js_1 = require("./phantasmaLink.js");
const ProofOfWork_js_1 = require("./interfaces/ProofOfWork.js");
const easyScript_js_1 = require("./easyScript.js");
const logger_js_1 = require("../utils/logger.js");
class EasyConnect {
    constructor(_options = null) {
        this.platform = 'phantasma';
        this.providerHint = 'poltergeist';
        this.script = new easyScript_js_1.EasyScript();
        this.link = new phantasmaLink_js_1.PhantasmaLink('easyConnect', false);
        this.connected = false;
        this.requiredVersion = 4;
        this.nexus = null;
        if (_options == null) {
            this.setConfig('auto');
        }
        else {
            try {
                this.requiredVersion = Number(_options[0]);
                this.platform = _options[1];
                this.providerHint = _options[2];
            }
            catch (error) {
                logger_js_1.logger.log(error);
            }
        }
        this.script = new easyScript_js_1.EasyScript();
    }
    setConfig(_provider) {
        this.requiredVersion = 4;
        this.platform = 'phantasma';
        switch (_provider) {
            case 'auto':
                // @ts-ignore
                if (!!window.PhantasmaLinkSocket == true) {
                    this.setConfig('ecto');
                }
                else {
                    this.providerHint = '';
                }
                break;
            case 'ecto':
                this.providerHint = 'ecto';
                break;
            case 'poltergeist':
                this.providerHint = 'poltergeist';
                break;
        }
    }
    connect(onSuccess = (data) => { }, onFail = (data) => {
        logger_js_1.logger.log('%cError: ' + data, 'color:red');
    }) {
        let that = this;
        this.link.login(function (data) {
            //Console Logging for Debugging Purposes
            if (data) {
                that.connected = true;
                that.nexus =
                    that.link.nexus === easyScript_js_1.Nexus.Mainnet ||
                        that.link.nexus === easyScript_js_1.Nexus.Simnet ||
                        that.link.nexus === easyScript_js_1.Nexus.Testnet
                        ? that.link.nexus
                        : null;
                onSuccess(data);
                logger_js_1.logger.log('%c[EasyConnect Connected]', 'color:green');
                logger_js_1.logger.log("Wallet Address '" + that.link.account.address + "' connected via " + that.link.wallet);
            }
            else {
                onFail();
                logger_js_1.logger.log('EasyConnect could not connect to wallet');
            }
        }, onFail, this.requiredVersion, this.platform, this.providerHint);
    }
    disconnect(_message = 'Graceful Disconect') {
        this.link.disconnect(_message);
        this.connected = false;
        this.nexus = null;
    }
    async query(_type = null, _arguments = null, _callback = (data) => {
        logger_js_1.logger.log(data);
    }) {
        if (this.connected == true) {
            switch (_type) {
                case 'account':
                    let account = this.link.account;
                    _callback(account);
                    return account;
                    break;
                case 'name':
                    let name = this.link.account.name;
                    _callback(name);
                    return name;
                    break;
                case 'balances':
                    let balances = this.link.account.balances;
                    _callback(balances);
                    return balances;
                    break;
                case 'walletAddress':
                    let walletAddress = this.link.account.address;
                    _callback(walletAddress);
                    return walletAddress;
                    break;
                case 'avatar':
                    let avatar = this.link.account.avatar;
                    _callback(avatar);
                    return avatar;
                    break;
                case 'tokenBalance':
                    //let token = _arguments[0];
                    //return this.link.accounts[]
                    break;
            }
        }
        else {
            logger_js_1.logger.log('%cWallet is not connected', 'color:red');
        }
    }
    async action(_type = null, _arguments = null, onSuccess = (data) => { }, onFail = (data) => {
        logger_js_1.logger.log('%cError: ' + data, 'color:red');
    }) {
        if (this.connected == true) {
            switch (_type) {
                case 'sendFT':
                    let sendFTScript = await this.script.buildScript('interop', [
                        'Runtime.SendTokens',
                        [_arguments[0], _arguments[1], _arguments[2], _arguments[3]],
                    ]);
                    this.signTransaction(sendFTScript, null, onSuccess, onFail);
                    break;
                case 'sendNFT':
                    let sendNFTScript = await this.script.buildScript('interop', [
                        'Runtime.SendTokens',
                        [_arguments[0], _arguments[1], _arguments[2], _arguments[3]],
                    ]);
                    this.signTransaction(sendNFTScript, null, onSuccess, onFail);
                    break;
            }
        }
        else {
            logger_js_1.logger.log('%cWallet is not connected', 'color:red');
        }
    }
    signTransaction(script, payload = null, onSuccess = (data) => { }, onFail = (data) => {
        logger_js_1.logger.log('%cError: ' + data, 'color:red');
    }) {
        this.link.signTx(script, payload, onSuccess, onFail);
    }
    signData(data, onSuccess = (data) => { }, onFail = (data) => {
        logger_js_1.logger.log('%cError: ' + data, 'color:red');
    }) {
        this.link.signData(data, onSuccess, onFail);
    }
    signCarbonTransaction(txMsg, onSuccess = (data) => { }, onFail = (data) => {
        logger_js_1.logger.log('%cError: ' + data, 'color:red');
    }) {
        if (this.connected == true) {
            this.link.signCarbonTxAndBroadcast(txMsg, onSuccess, onFail);
        }
        else {
            const message = 'Wallet is not connected';
            logger_js_1.logger.log('%c' + message, 'color:red');
            onFail(message);
        }
    }
    signPrebuiltTransaction(tx, onSuccess = (data) => { }, onFail = (data) => {
        logger_js_1.logger.log('%cError: ' + data, 'color:red');
    }) {
        if (this.connected == true) {
            this.link.signPrebuiltTransaction(tx, onSuccess, onFail);
        }
        else {
            const message = 'Wallet is not connected';
            logger_js_1.logger.log('%c' + message, 'color:red');
            onFail(message);
        }
    }
    invokeScript(script, _callback) {
        this.link.invokeScript(script, _callback);
    }
    deployContract(script, payload = null, proofOfWork = ProofOfWork_js_1.ProofOfWork.Minimal, onSuccess = (data) => { }, onFail = (data) => {
        logger_js_1.logger.log('%cError: ' + data, 'color:red');
    }) {
        this.link.signTx(script, payload, onSuccess, onFail, proofOfWork);
    }
}
exports.EasyConnect = EasyConnect;
