// phantasma-sdk-ts/dist/cjs/core/link/phantasmaLink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhantasmaLink = void 0;
const Transaction_js_1 = require("../tx/Transaction.js");
const index_js_1 = require("../vm/index.js");
const ProofOfWork_js_1 = require("./interfaces/ProofOfWork.js");
const CarbonBlob_js_1 = require("../types/Carbon/CarbonBlob.js");
const Hex_js_1 = require("../utils/Hex.js");
const index_js_2 = require("../types/index.js");
const logger_js_1 = require("../utils/logger.js");
class PhantasmaLink {
    //Construct The Link
    constructor(dappID, logging = true) {
        this.lastSocketErrorMessage = null;
        this.socketTransport = null;
        this.socketOpen = false;
        this.requestID = 0;
        //Message Logging
        this.onMessage = (msg) => {
            if (this.messageLogging == true) {
                logger_js_1.logger.log(msg);
            }
        };
        this.version = 4;
        this.nexus = '';
        this.chain = 'main';
        this.platform = 'poltergeist';
        //Turn On|Off Console Logging
        if (logging == false) {
            this.messageLogging = false;
        }
        else {
            this.messageLogging = true;
            logger_js_1.logger.log('%cPhantasmaLink created', 'color:green');
        }
        this.requestID = 0;
        //Standard Sets
        this.host = 'localhost:7090';
        this.dapp = dappID;
        this.onLogin = function (succ) { }; //Does Nothing for Now
        this.onError = function (message) { }; //Does Nothing for Now
    }
    // Preserve wallet-side failure details whenever the transport provides them.
    describeFailure(result, fallback) {
        if (typeof result === 'string' && result.length > 0) {
            return result;
        }
        if (result && typeof result.error === 'string' && result.error.length > 0) {
            return result.error;
        }
        if (result && typeof result.message === 'string' && result.message.length > 0) {
            return result.message;
        }
        return fallback;
    }
    //Connect To Wallet
    login(onLoginCallback, onErrorCallback, version = 4, platform = 'phantasma', providerHint = 'poltergeist') {
        this.providerHint = providerHint;
        this.onLogin = onLoginCallback;
        this.onError = onErrorCallback;
        this.version = version;
        this.platform = platform;
        this.providerHint = providerHint;
        this.createSocket();
    }
    //Script Invoking With Wallet Connection
    invokeScript(script, callback) {
        this.onMessage('Relaying transaction to wallet...');
        if (!this.socket) {
            callback('not logged in');
            return;
        }
        if (script.length >= 8192) {
            callback('script too big, sorry :(');
            return;
        }
        let requestStr = this.chain + '/' + script;
        if (this.version >= 2) {
            requestStr = requestStr;
        }
        else {
            requestStr = this.nexus + '/' + requestStr;
        }
        let invokeScriptRequest = 'invokeScript/' + requestStr;
        var that = this;
        this.sendLinkRequest(invokeScriptRequest, function (result) {
            if (result.success) {
                that.onMessage('Invoke successful, hash: ' + result + '...');
                if (callback) {
                    callback(result);
                }
            }
        });
    }
    //Wallet Transaction Signing + Sending
    signTx(script, payload, callback, onErrorCallback, pow = ProofOfWork_js_1.ProofOfWork.None, signature = 'Ed25519') {
        //Overload Protection
        if (script.length >= 65536) {
            this.onMessage('Error: script is too big!');
            if (onErrorCallback) {
                onErrorCallback();
            }
            return;
        }
        //Check Payload
        if (payload == null) {
            payload = '7068616e7461736d612d7473'; //Says 'Phantasma-ts' in hex
        }
        else if (typeof payload === 'string') {
            //Turn String Payload -> Bytes -> Hex
            let sb = new index_js_1.ScriptBuilder();
            let bytes = sb.RawString(payload);
            sb.AppendBytes(bytes);
            payload = sb.EndScript();
        }
        else {
            this.onMessage('Error: Invalid Payload');
            if (onErrorCallback) {
                onErrorCallback();
            }
            return;
        }
        this.onError = onErrorCallback; //Sets Error Callback Function
        let that = this; //Allows the use of 'this' inside sendLinkRequest Object
        let request = 'signTx/' +
            this.chain +
            '/' +
            script +
            '/' +
            payload +
            '/' +
            signature +
            '/' +
            this.platform +
            '/' +
            pow;
        if (this.version == 1) {
            request = 'signTx/' + this.nexus + '/' + this.chain + '/' + script + '/' + payload;
        }
        //Sends Signiture Request To Connected Wallet For Script
        this.sendLinkRequest(request, function (result) {
            if (result.success) {
                if (result.hash.error) {
                    that.onMessage('Error: ' + result.hash.error);
                    return;
                }
                that.onMessage('Transaction successful, hash: ' + result.hash.substr(0, 15) + '...');
                if (callback) {
                    callback(result);
                }
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback();
                }
            }
        });
    }
    // Wallet Transaction Signing
    signCarbonTxAndBroadcast(txMsg, callback = () => { }, onErrorCallback = () => { }) {
        if (!txMsg) {
            const message = 'Error: Invalid Carbon transaction message';
            this.onMessage(message);
            onErrorCallback(message);
            return;
        }
        if (this.version < 4) {
            const message = 'Carbon transactions require a wallet that supports Phantasma Link v4 or higher. Please reconnect with version 4+.';
            this.onMessage(message);
            if (onErrorCallback) {
                onErrorCallback(message);
            }
            return;
        }
        let txHex;
        try {
            txHex = this.serializeCarbonTx(txMsg);
        }
        catch (err) {
            const message = 'Error: Unable to serialize Carbon transaction';
            this.onMessage(message + (err?.message ? ` (${err.message})` : ''));
            onErrorCallback(message);
            return;
        }
        const txLengthLimit = 65536;
        if (txHex.length >= txLengthLimit) {
            const message = `Error: Carbon transaction message is too big (${txHex.length} > ${txLengthLimit})!`;
            this.onMessage(message);
            onErrorCallback(message);
            return;
        }
        this.onError = onErrorCallback;
        const request = 'signCarbonTxAndBroadcast/' + txHex;
        this.sendLinkRequest(request, (result) => {
            if (result.success) {
                this.onMessage('Carbon transaction signed');
                callback(result);
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback(result.error || 'Carbon transaction signing failed');
                }
            }
        });
    }
    signTxSignature(tx, callback, onErrorCallback, signature = 'Ed25519') {
        if (!this.socket) {
            this.onMessage('not logged in');
            if (onErrorCallback) {
                onErrorCallback('Wallet is not connected');
            }
            return;
        }
        if (tx == null) {
            this.onMessage('invalid data, sorry :(');
            if (onErrorCallback) {
                onErrorCallback('Invalid transaction data');
            }
            return;
        }
        if (tx.length >= 65536) {
            this.onMessage('data too big, sorry :(');
            if (onErrorCallback) {
                onErrorCallback('Transaction data is too big');
            }
            return;
        }
        this.onError = onErrorCallback;
        let signDataStr = 'signTxSignature/' + tx + '/' + signature + '/' + this.platform;
        var that = this; //Allows the use of 'this' inside sendLinkRequest Object
        this.sendLinkRequest(signDataStr, function (result) {
            if (result.success) {
                that.onMessage('Data successfully signed');
                if (callback) {
                    callback(result);
                }
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback(that.describeFailure(result, 'Wallet rejected transaction signature'));
                }
            }
        });
    }
    decodeWalletSignatureBytes(signatureHex, signature) {
        if (signature !== 'Ed25519') {
            throw new Error(`Unsupported transaction signature type: ${signature}`);
        }
        const reader = new index_js_2.PBinaryReader((0, Hex_js_1.hexToBytes)(signatureHex));
        const rawSignatureHex = reader.readByteArray();
        if (typeof rawSignatureHex !== 'string' || rawSignatureHex.length === 0) {
            throw new Error('Wallet returned an empty transaction signature');
        }
        return (0, Hex_js_1.hexToBytes)(rawSignatureHex);
    }
    signPrebuiltTransaction(tx, callback, onErrorCallback, signature = 'Ed25519') {
        if (!tx) {
            const message = 'Error: Invalid transaction';
            this.onMessage(message);
            if (onErrorCallback) {
                onErrorCallback(message);
            }
            return;
        }
        let unsignedTxHex;
        try {
            unsignedTxHex = tx.ToStringEncoded(false).toUpperCase();
        }
        catch (err) {
            const message = 'Error: Unable to encode unsigned transaction';
            this.onMessage(message + (err?.message ? ` (${err.message})` : ''));
            if (onErrorCallback) {
                onErrorCallback(message);
            }
            return;
        }
        this.signTxSignature(unsignedTxHex, (result) => {
            if (!result?.success || typeof result.signature !== 'string' || result.signature.length === 0) {
                const failure = this.describeFailure(result, 'Wallet rejected transaction signature');
                if (onErrorCallback) {
                    onErrorCallback(failure);
                }
                return;
            }
            try {
                const signedTx = Transaction_js_1.Transaction.FromBytes(unsignedTxHex);
                signedTx.signatures = [
                    new index_js_2.Ed25519Signature(this.decodeWalletSignatureBytes(result.signature, signature)),
                ];
                if (this.account?.address && !signedTx.VerifySignature(this.account.address)) {
                    throw new Error('Wallet returned a signature that does not match the connected account');
                }
                callback({
                    success: true,
                    signature: result.signature,
                    signedTx: signedTx.ToStringEncoded(true).toUpperCase(),
                });
            }
            catch (err) {
                const message = err?.message || 'Unable to assemble signed transaction';
                if (onErrorCallback) {
                    onErrorCallback(message);
                }
            }
        }, (message) => {
            const failure = message || this.lastSocketErrorMessage || 'Wallet rejected transaction signature';
            if (onErrorCallback) {
                onErrorCallback(failure);
            }
        }, signature);
    }
    multiSig(subject, callback, onErrorCallback, signature = 'Ed25519') {
        if (!this.socket) {
            this.onMessage('not logged in');
            return;
        }
        if (subject == null) {
            this.onMessage('invalid data, sorry :(');
            return;
        }
        if (subject.length >= 1024) {
            this.onMessage('data too big, sorry :(');
            if (onErrorCallback) {
                onErrorCallback();
            }
            return;
        }
        let signDataStr = 'multiSig/' + subject + '/' + signature + '/' + this.platform;
        var that = this; //Allows the use of 'this' inside sendLinkRequest Object
        this.sendLinkRequest(signDataStr, function (result) {
            if (result.success) {
                that.onMessage('Data successfully signed');
                if (callback) {
                    callback(result);
                }
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback();
                }
            }
        });
    }
    getPeer(callback, onErrorCallback) {
        this.onError = onErrorCallback; //Sets Error Callback Function
        let that = this; //Allows the use of 'this' inside sendLinkRequest Object
        //Sends Signiture Request To Connected Wallet For Script
        this.sendLinkRequest('getPeer/', function (result) {
            if (result.success) {
                that.onMessage('Peer Query,: ' + result);
                if (callback) {
                    callback(result);
                }
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback();
                }
            }
        });
    }
    fetchWallet(callback, onErrorCallback) {
        let that = this; //Allows the use of 'this' inside sendLinkRequest Object
        let getAccountRequest = 'getAccount/' + this.platform;
        this.sendLinkRequest(getAccountRequest, function (result) {
            if (result.success) {
                that.account = result;
                callback(result);
            }
            else {
                onErrorCallback('Could not obtain account info... Make sure you have an account currently open in ' +
                    that.wallet +
                    '...');
                //that.disconnect("Unable to optain Account Info");
            }
            //that.onLogin(result.success);
            //that.onLogin = null;
        });
    }
    getNexus(callback, onErrorCallback) {
        this.onError = onErrorCallback; //Sets Error Callback Function
        let that = this; //Allows the use of 'this' inside sendLinkRequest Object
        //Sends Signiture Request To Connected Wallet For Script
        this.sendLinkRequest('getNexus/', function (result) {
            if (result.success) {
                if (typeof result.nexus === 'string') {
                    that.nexus = result.nexus;
                }
                that.onMessage('Nexus Query,: ' + result);
                if (callback) {
                    callback(result);
                }
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback('Error: ' + result.error);
                }
            }
        });
    }
    getWalletVersion(callback, onErrorCallback) {
        this.onError = onErrorCallback; //Sets Error Callback Function
        let that = this; //Allows the use of 'this' inside sendLinkRequest Object
        //Sends Signiture Request To Connected Wallet For Script
        this.sendLinkRequest('getWalletVersion/', function (result) {
            if (result.success) {
                that.onMessage('Wallet Version Query,: ' + result);
                if (callback) {
                    callback(result);
                }
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback('Error: ' + result.error);
                }
            }
        });
    }
    //Uses Wallet To Sign Data With Signiture
    // Data needs to be in Base16 encode.
    signData(data, callback, onErrorCallback, signature = 'Ed25519') {
        if (!this.socket) {
            this.onMessage('not logged in');
            return;
        }
        if (data == null) {
            this.onMessage('invalid data, sorry :(');
            return;
        }
        if (data.length >= 1024) {
            this.onMessage('data too big, sorry :(');
            if (onErrorCallback) {
                onErrorCallback('data too big, sorry :(');
            }
            return;
        }
        let signDataStr = 'signData/' + data + '/' + signature + '/' + this.platform;
        var that = this; //Allows the use of 'this' inside sendLinkRequest Object
        this.sendLinkRequest(signDataStr, function (result) {
            if (result.success) {
                that.onMessage('Data successfully signed');
                if (callback) {
                    callback(result);
                }
            }
            else {
                if (onErrorCallback) {
                    onErrorCallback(result);
                }
            }
        });
    }
    //Wallet Socket Connection Creation
    createSocket(isResume = false) {
        let path = 'ws://' + this.host + '/phantasma';
        this.onMessage('Phantasma Link connecting...');
        if (this.socket) {
            this.socket.close();
        }
        const useInjectedSocket = 
        // @ts-ignore
        !!window.PhantasmaLinkSocket && this.providerHint !== 'poltergeist';
        this.socketTransport = useInjectedSocket ? 'injected' : 'websocket';
        this.socketOpen = false;
        this.onMessage(useInjectedSocket
            ? 'Using injected PhantasmaLinkSocket transport'
            : `Using raw WebSocket transport: ${path}`);
        //@ts-ignore
        this.socket = useInjectedSocket
            ? // @ts-ignore
                new PhantasmaLinkSocket()
            : new WebSocket(path);
        this.requestCallback = null;
        this.lastSocketErrorMessage = null;
        this.token = null;
        this.account = null;
        this.nexus = '';
        this.requestID = 0;
        let authorizeRequest = 'authorize/' + this.dapp + '/' + this.version;
        let getAccountRequest = 'getAccount/' + this.platform;
        let that = this;
        //Once Socket Opened
        this.socket.onopen = function (e) {
            that.socketOpen = true;
            that.onMessage('Connection established, authorizing dapp in wallet...');
            if (isResume) {
                that.fetchWallet(undefined, undefined);
            }
            else {
                that.sendLinkRequest(authorizeRequest, function (result) {
                    //Set Global Variables With Successful Account Query
                    if (result.success) {
                        that.token = result.token;
                        that.wallet = result.wallet;
                        that.nexus = typeof result.nexus === 'string' ? result.nexus : '';
                        that.onMessage('Authorized, obtaining account info...');
                        that.sendLinkRequest(getAccountRequest, function (result) {
                            if (result.success) {
                                that.account = result;
                            }
                            else {
                                that.onError('Could not obtain account info... Make sure you have an account currently open in ' +
                                    that.wallet +
                                    '...');
                                that.disconnect('Unable to optain Account Info');
                            }
                            that.onLogin(result.success);
                            that.onLogin = null;
                        });
                    }
                    else {
                        that.onError(that.describeFailure(result, 'Authorization failed...'));
                        that.disconnect('Auth Failure');
                    }
                });
            }
        };
        //Retrieves Message From Socket and Processes It
        this.socket.onmessage = function (event) {
            const obj = JSON.parse(event.data);
            if (that.messageLogging == true) {
                logger_js_1.logger.log('%c' + event.data, 'color:blue');
            }
            //Checks What To Do Based On Message
            switch (obj.message) {
                case 'Wallet is Closed':
                    that.onError('Could not obtain account info... Make sure you have an account currently open in ' +
                        that.wallet);
                    that.disconnect(true);
                    break;
                case 'not logged in':
                    that.onError('Could not obtain account info... Make sure you have an account currently logged in');
                    that.disconnect(true);
                    break;
                case 'A previouus request is still pending':
                case 'A previous request is still pending':
                    that.onError(that.describeFailure(obj, 'You have a pending action in your wallet'));
                    break;
                case 'user rejected':
                    that.onError('Transaction cancelled by user in ' + that.wallet);
                    break;
                case 'user rejected':
                    that.onError('Transaction cancelled by user in ' + that.wallet);
                    break;
                default:
                    if (obj.message && obj.message.startsWith('nexus mismatch')) {
                        that.onError(obj.message);
                    }
                    else {
                        let temp = that.requestCallback;
                        if (temp == null) {
                            that.onError('Something bad happened');
                            return;
                        }
                        that.requestCallback = null;
                        temp(obj);
                    }
                    break;
            }
        };
        //Cleanup After Socket Closes
        this.socket.onclose = function (event) {
            that.socketOpen = false;
            const reason = event.reason && event.reason.length > 0
                ? event.reason
                : that.lastSocketErrorMessage || (event.wasClean ? 'Wallet connection closed' : 'Connection terminated unexpectedly');
            that.lastSocketErrorMessage = null;
            if (that.requestCallback) {
                that.handleSocketFailure(reason);
            }
            else if (!event.wasClean) {
                that.handleSocketFailure(reason);
            }
        };
        //Error Callback When Socket Has Error
        this.socket.onerror = function (error) {
            const errMsg = error && typeof error.message === 'string' && error.message.length > 0
                ? error.message
                : 'WebSocket error';
            that.lastSocketErrorMessage = errMsg;
            that.onMessage('Error: ' + errMsg);
        };
    }
    //Message Logging Util
    toggleMessageLogging() {
        if (this.messageLogging == true) {
            this.messageLogging = false;
        }
        else {
            this.messageLogging = true;
        }
    }
    resume(token) {
        this.token = token;
        this.retry();
    }
    //Retry Util
    retry() {
        this.createSocket();
    }
    //Get Dapp ID Name Util
    set dappID(dapp) {
        this.dapp = dapp;
    }
    get dappID() {
        return this.dapp;
    }
    //Build Request and Send To Wallet Via Socket
    sendLinkRequest(request, callback) {
        this.onMessage('Sending Phantasma Link request: ' + request);
        this.requestCallback = callback;
        const socket = this.socket;
        const openState = typeof WebSocket !== 'undefined' && typeof WebSocket.OPEN === 'number' ? WebSocket.OPEN : 1;
        const hasSend = socket && typeof socket.send === 'function';
        const hasReadyState = socket && typeof socket.readyState === 'number';
        const isSocketOpen = hasSend &&
            (hasReadyState ? socket.readyState === openState : this.socketOpen);
        if (!socket || !hasSend || !isSocketOpen) {
            this.handleSocketFailure('Wallet connection is closed. Please reconnect to your wallet.');
            return;
        }
        if (this.token != null) {
            request = request + '/' + this.dapp + '/' + this.token;
        }
        this.requestID++; //Object Nonce Increase?
        request = this.requestID + ',' + request;
        try {
            socket.send(request);
        }
        catch (err) {
            const errMessage = err && typeof err.message === 'string' && err.message.length > 0
                ? err.message
                : 'Failed to send request to wallet';
            this.handleSocketFailure(errMessage);
        }
    }
    handleSocketFailure(message) {
        const callback = this.requestCallback;
        this.requestCallback = null;
        const errorMessage = message || 'Connection lost with Phantasma Link wallet';
        if (callback) {
            callback({ success: false, error: errorMessage });
            return;
        }
        if (this.onError) {
            this.onError(errorMessage);
        }
    }
    //Disconnect The Wallet Connection Socket
    disconnect(triggered) {
        this.onMessage('Disconnecting Phantasma Link: ' + triggered);
        this.socketOpen = false;
        if (this.socket)
            this.socket.close();
    }
    serializeCarbonTx(txMsg) {
        const serialized = CarbonBlob_js_1.CarbonBlob.Serialize(txMsg);
        return (0, Hex_js_1.bytesToHex)(serialized);
    }
}
exports.PhantasmaLink = PhantasmaLink;
