// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/TxMsg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMsg = void 0;
const Bytes32_js_1 = require("../Bytes32.js");
const SmallString_js_1 = require("../SmallString.js");
const TxTypes_js_1 = require("../TxTypes.js");
const TxMsgCall_js_1 = require("./TxMsgCall.js");
const TxMsgCallMulti_js_1 = require("./TxMsgCallMulti.js");
const TxMsgMintFungible_js_1 = require("./TxMsgMintFungible.js");
const TxMsgMintNonFungible_js_1 = require("./TxMsgMintNonFungible.js");
const TxMsgTrade_js_1 = require("./TxMsgTrade.js");
const TxMsgTransferFungible_js_1 = require("./TxMsgTransferFungible.js");
const TxMsgTransferFungibleGasPayer_js_1 = require("./TxMsgTransferFungibleGasPayer.js");
const TxMsgTransferNonFungibleMulti_js_1 = require("./TxMsgTransferNonFungibleMulti.js");
const TxMsgTransferNonFungibleMultiGasPayer_js_1 = require("./TxMsgTransferNonFungibleMultiGasPayer.js");
const TxMsgTransferNonFungibleSingle_js_1 = require("./TxMsgTransferNonFungibleSingle.js");
const TxMsgTransferNonFungibleSingleGasPayer_js_1 = require("./TxMsgTransferNonFungibleSingleGasPayer.js");
const TxMsgBurnFungible_js_1 = require("./TxMsgBurnFungible.js");
const TxMsgBurnFungibleGasPayer_js_1 = require("./TxMsgBurnFungibleGasPayer.js");
const TxMsgBurnNonFungible_js_1 = require("./TxMsgBurnNonFungible.js");
const TxMsgBurnNonFungibleGasPayer_js_1 = require("./TxMsgBurnNonFungibleGasPayer.js");
const TxMsgPhantasma_js_1 = require("./TxMsgPhantasma.js");
const TxMsgPhantasmaRaw_js_1 = require("./TxMsgPhantasmaRaw.js");
class TxMsg {
    constructor(type = TxTypes_js_1.TxTypes.Call, expiry = 0n, // int64
    maxGas = 0n, // uint64
    maxData = 0n, // uint64
    gasFrom = new Bytes32_js_1.Bytes32(), payload = new SmallString_js_1.SmallString(''), msg) {
        this.type = type;
        this.expiry = expiry;
        this.maxGas = maxGas;
        this.maxData = maxData;
        this.gasFrom = gasFrom;
        this.payload = payload;
        this.msg = msg;
    }
    write(w) {
        w.write1(this.type);
        w.write8(this.expiry);
        w.write8u(this.maxGas);
        w.write8u(this.maxData);
        this.gasFrom.write(w);
        this.payload.write(w);
        switch (this.type) {
            case TxTypes_js_1.TxTypes.Call:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.Call_Multi:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.Trade:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.TransferFungible:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.TransferFungible_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Single:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Single_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Multi:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Multi_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.MintFungible:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.BurnFungible:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.BurnFungible_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.MintNonFungible:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.BurnNonFungible:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.BurnNonFungible_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.Phantasma:
                this.msg.write(w);
                break;
            case TxTypes_js_1.TxTypes.Phantasma_Raw:
                this.msg.write(w);
                break;
            default:
                throw new Error('Unsupported transaction type');
        }
    }
    read(r) {
        this.type = r.read1();
        this.expiry = r.read8();
        this.maxGas = r.read8u();
        this.maxData = r.read8u();
        this.gasFrom = Bytes32_js_1.Bytes32.read(r);
        this.payload = SmallString_js_1.SmallString.read(r);
        switch (this.type) {
            case TxTypes_js_1.TxTypes.Call:
                this.msg = TxMsgCall_js_1.TxMsgCall.read(r);
                break;
            case TxTypes_js_1.TxTypes.Call_Multi:
                this.msg = TxMsgCallMulti_js_1.TxMsgCallMulti.read(r);
                break;
            case TxTypes_js_1.TxTypes.Trade:
                this.msg = TxMsgTrade_js_1.TxMsgTrade.read(r);
                break;
            case TxTypes_js_1.TxTypes.TransferFungible:
                this.msg = TxMsgTransferFungible_js_1.TxMsgTransferFungible.read(r);
                break;
            case TxTypes_js_1.TxTypes.TransferFungible_GasPayer:
                this.msg = TxMsgTransferFungibleGasPayer_js_1.TxMsgTransferFungibleGasPayer.read(r);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Single:
                this.msg = TxMsgTransferNonFungibleSingle_js_1.TxMsgTransferNonFungibleSingle.read(r);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Single_GasPayer:
                this.msg = TxMsgTransferNonFungibleSingleGasPayer_js_1.TxMsgTransferNonFungibleSingleGasPayer.read(r);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Multi:
                this.msg = TxMsgTransferNonFungibleMulti_js_1.TxMsgTransferNonFungibleMulti.read(r);
                break;
            case TxTypes_js_1.TxTypes.TransferNonFungible_Multi_GasPayer:
                this.msg = TxMsgTransferNonFungibleMultiGasPayer_js_1.TxMsgTransferNonFungibleMultiGasPayer.read(r);
                break;
            case TxTypes_js_1.TxTypes.MintFungible:
                this.msg = TxMsgMintFungible_js_1.TxMsgMintFungible.read(r);
                break;
            case TxTypes_js_1.TxTypes.BurnFungible:
                this.msg = TxMsgBurnFungible_js_1.TxMsgBurnFungible.read(r);
                break;
            case TxTypes_js_1.TxTypes.BurnFungible_GasPayer:
                this.msg = TxMsgBurnFungibleGasPayer_js_1.TxMsgBurnFungibleGasPayer.read(r);
                break;
            case TxTypes_js_1.TxTypes.MintNonFungible:
                this.msg = TxMsgMintNonFungible_js_1.TxMsgMintNonFungible.read(r);
                break;
            case TxTypes_js_1.TxTypes.BurnNonFungible:
                this.msg = TxMsgBurnNonFungible_js_1.TxMsgBurnNonFungible.read(r);
                break;
            case TxTypes_js_1.TxTypes.BurnNonFungible_GasPayer:
                this.msg = TxMsgBurnNonFungibleGasPayer_js_1.TxMsgBurnNonFungibleGasPayer.read(r);
                break;
            case TxTypes_js_1.TxTypes.Phantasma:
                this.msg = TxMsgPhantasma_js_1.TxMsgPhantasma.read(r);
                break;
            case TxTypes_js_1.TxTypes.Phantasma_Raw:
                this.msg = TxMsgPhantasmaRaw_js_1.TxMsgPhantasmaRaw.read(r);
                break;
            default:
                throw new Error('Unsupported transaction type');
        }
    }
    static read(r) {
        const v = new TxMsg();
        v.read(r);
        return v;
    }
}
exports.TxMsg = TxMsg;
