// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/TxMsgTransferNonFungibleMultiGasPayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMsgTransferNonFungibleMultiGasPayer = void 0;
const Bytes32_js_1 = require("../Bytes32.js");
class TxMsgTransferNonFungibleMultiGasPayer {
    constructor(init) {
        this.to = new Bytes32_js_1.Bytes32();
        this.from = new Bytes32_js_1.Bytes32();
        this.tokenId = 0n;
        this.instanceIds = [];
        Object.assign(this, init);
    }
    write(w) {
        w.write32(this.to);
        w.write32(this.from);
        w.write8u(this.tokenId);
        w.writeArray64u(this.instanceIds);
    }
    read(r) {
        this.to = Bytes32_js_1.Bytes32.read(r);
        this.from = Bytes32_js_1.Bytes32.read(r);
        this.tokenId = r.read8u();
        this.instanceIds = r.readArray64u();
    }
    static read(r) {
        const v = new TxMsgTransferNonFungibleMultiGasPayer();
        v.read(r);
        return v;
    }
}
exports.TxMsgTransferNonFungibleMultiGasPayer = TxMsgTransferNonFungibleMultiGasPayer;
