// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/TxMsgTrade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMsgTrade = void 0;
const TxMsgBurnFungibleGasPayer_js_1 = require("./TxMsgBurnFungibleGasPayer.js");
const TxMsgBurnNonFungibleGasPayer_js_1 = require("./TxMsgBurnNonFungibleGasPayer.js");
const TxMsgMintFungible_js_1 = require("./TxMsgMintFungible.js");
const TxMsgMintNonFungible_js_1 = require("./TxMsgMintNonFungible.js");
const TxMsgTransferFungibleGasPayer_js_1 = require("./TxMsgTransferFungibleGasPayer.js");
const TxMsgTransferNonFungibleSingleGasPayer_js_1 = require("./TxMsgTransferNonFungibleSingleGasPayer.js");
class TxMsgTrade {
    constructor(init) {
        this.transferF = [];
        this.transferN = [];
        this.mintF = [];
        this.burnF = [];
        this.mintN = [];
        this.burnN = [];
        Object.assign(this, init);
    }
    write(w) {
        w.writeArrayBlob(this.transferF);
        w.writeArrayBlob(this.transferN);
        w.writeArrayBlob(this.mintF);
        w.writeArrayBlob(this.burnF);
        w.writeArrayBlob(this.mintN);
        w.writeArrayBlob(this.burnN);
    }
    read(r) {
        this.transferF = r.readArrayBlob(TxMsgTransferFungibleGasPayer_js_1.TxMsgTransferFungibleGasPayer);
        this.transferN = r.readArrayBlob(TxMsgTransferNonFungibleSingleGasPayer_js_1.TxMsgTransferNonFungibleSingleGasPayer);
        this.mintF = r.readArrayBlob(TxMsgMintFungible_js_1.TxMsgMintFungible);
        this.burnF = r.readArrayBlob(TxMsgBurnFungibleGasPayer_js_1.TxMsgBurnFungibleGasPayer);
        this.mintN = r.readArrayBlob(TxMsgMintNonFungible_js_1.TxMsgMintNonFungible);
        this.burnN = r.readArrayBlob(TxMsgBurnNonFungibleGasPayer_js_1.TxMsgBurnNonFungibleGasPayer);
    }
    static read(r) {
        const v = new TxMsgTrade();
        v.read(r);
        return v;
    }
}
exports.TxMsgTrade = TxMsgTrade;
