// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/TxMsgTransferNonFungibleSingleGasPayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxMsgTransferNonFungibleSingleGasPayer = void 0;
const Bytes32_js_1 = require("../Bytes32.js");
class TxMsgTransferNonFungibleSingleGasPayer {
    constructor(init) {
        this.to = new Bytes32_js_1.Bytes32();
        this.from = new Bytes32_js_1.Bytes32();
        this.tokenId = 0n;
        this.instanceId = 0n;
        Object.assign(this, init);
    }
    write(w) {
        w.write32(this.to);
        w.write32(this.from);
        w.write8u(this.tokenId);
        w.write8u(this.instanceId);
    }
    read(r) {
        this.to = Bytes32_js_1.Bytes32.read(r);
        this.from = Bytes32_js_1.Bytes32.read(r);
        this.tokenId = r.read8u();
        this.instanceId = r.read8u();
    }
    static read(r) {
        const v = new TxMsgTransferNonFungibleSingleGasPayer();
        v.read(r);
        return v;
    }
}
exports.TxMsgTransferNonFungibleSingleGasPayer = TxMsgTransferNonFungibleSingleGasPayer;
