// phantasma-sdk-ts/dist/cjs/core/types/Extensions/PBinaryReader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBinaryReader = void 0;
const big_integer_1 = __importDefault(require("big-integer"));
const csharp_binary_stream_1 = require("csharp-binary-stream");
const index_js_1 = require("../../interfaces/index.js");
const index_js_2 = require("../../utils/index.js");
const index_js_3 = require("../../vm/index.js");
const Ed25519Signature_js_1 = require("../Ed25519Signature.js");
const Timestamp_js_1 = require("../Timestamp.js");
class PBinaryReader {
    get length() {
        return this.reader.length;
    }
    get position() {
        return this.reader.position;
    }
    set position(value) {
        this.reader.position = value;
    }
    get isEndOfStream() {
        return this.reader.isEndOfStream;
    }
    readBoolean() {
        return this.reader.readBoolean();
    }
    readByte() {
        return this.reader.readByte();
    }
    readBytes(bytesToRead) {
        return this.reader.readBytes(bytesToRead);
    }
    readSignedByte() {
        return this.reader.readSignedByte();
    }
    readShort() {
        return this.reader.readShort();
    }
    readUnsignedShort() {
        return this.reader.readUnsignedShort();
    }
    readInt() {
        return this.reader.readInt();
    }
    readUnsignedInt() {
        return this.reader.readUnsignedInt();
    }
    readLongString() {
        return this.reader.readLongString();
    }
    readLong() {
        return this.reader.readLong();
    }
    readUnsignedLongString() {
        return this.reader.readUnsignedLongString();
    }
    readUnsignedLong() {
        return this.reader.readUnsignedLong();
    }
    readFloat() {
        return this.reader.readFloat();
    }
    readDouble() {
        return this.reader.readDouble();
    }
    readChar(encoding) {
        return this.reader.readChar(encoding);
    }
    readChars(charactersToRead, encoding) {
        return this.reader.readChars(charactersToRead, encoding);
    }
    readCharBytes(bytesToRead, encoding) {
        return this.reader.readCharBytes(bytesToRead, encoding);
    }
    constructor(arg1) {
        this.reader = new csharp_binary_stream_1.BinaryReader(arg1);
    }
    read(numBytes) {
        var res = (0, index_js_2.bytesToHex)(this.readBytes(numBytes)).substr(0, numBytes * 2);
        //this.position += numBytes;
        return res;
    }
    readString() {
        var len = this.readVarInt();
        return this.readStringBytes(len);
    }
    readStringBytes(numBytes) {
        var res = '';
        for (var i = 0; i < numBytes; ++i) {
            res += String.fromCharCode(this.readByte());
        }
        return res;
    }
    readBigInteger() {
        // TO DO: implement negative numbers
        var len = this.readVarInt();
        var res = 0;
        var stringBytes = this.read(len);
        [...stringBytes.match(/.{1,2}/g)]
            .reverse()
            .forEach((c) => (res = res * 256 + parseInt(c, 16)));
        let bigInt = BigInt(res);
        return bigInt;
    }
    readBigIntAccurate() {
        var len = this.readVarInt();
        var res = (0, big_integer_1.default)();
        var stringBytes = this.read(len);
        [...stringBytes.match(/.{1,2}/g)].reverse().forEach((c) => {
            res = res.times(256).plus(parseInt(c, 16));
        });
        return res.toString();
    }
    readSignatureV2() {
        let kind = this.readByte();
        let curve;
        let signature = new Ed25519Signature_js_1.Ed25519Signature();
        switch (kind) {
            case index_js_1.SignatureKind.None:
                return null;
            case index_js_1.SignatureKind.Ed25519:
                let len = this.readVarInt();
                signature.Bytes = new Uint8Array(this.readBytes(len));
                break;
            case index_js_1.SignatureKind.ECDSA:
                curve = this.readByte();
                signature.Bytes = (0, index_js_2.stringToUint8Array)(this.readString());
                break;
            default:
                throw 'read signature: ' + kind;
        }
        return signature;
    }
    readSignature() {
        let kind = this.readByte();
        let signature = new Ed25519Signature_js_1.Ed25519Signature();
        let curve;
        signature.Kind = kind;
        switch (kind) {
            case index_js_1.SignatureKind.None:
                return null;
            case index_js_1.SignatureKind.Ed25519:
                let len = this.readVarInt();
                signature.Bytes = (0, index_js_2.stringToUint8Array)(this.read(len));
                break;
            case index_js_1.SignatureKind.ECDSA:
                curve = this.readByte();
                signature.Bytes = (0, index_js_2.stringToUint8Array)(this.readString());
                break;
            default:
                throw 'read signature: ' + kind;
        }
        return signature;
    }
    readByteArray() {
        var res;
        var length = this.readVarInt();
        if (length == 0)
            return [];
        res = this.read(length);
        return res;
    }
    readTimestamp() {
        //var len = this.readByte();
        let result = 0;
        let bytes = this.read(4);
        //[...(bytes.match(/.{1,2}/g) as any)];
        bytes
            .match(/.{1,2}/g)
            .reverse()
            .forEach((c) => (result = result * 256 + parseInt(c, 16)));
        let timestamp = new Timestamp_js_1.Timestamp(result);
        return timestamp;
    }
    readVarInt() {
        var len = this.readByte();
        var res = 0;
        if (len === 0xfd) {
            [...this.read(2).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        else if (len === 0xfe) {
            [...this.read(4).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        else if (len === 0xff) {
            [...this.read(8).match(/.{1,2}/g)]
                .reverse()
                .forEach((c) => (res = res * 256 + parseInt(c, 16)));
            return res;
        }
        return len;
    }
    readVarString() {
        var len = this.readVarInt();
        if (len == 0)
            return '';
        return this.readStringBytes(len);
    }
    readVmObject() {
        const type = this.readByte();
        switch (type) {
            case index_js_3.VMType.String:
                return this.readString();
            case index_js_3.VMType.Number:
                return this.readBigIntAccurate();
            case index_js_3.VMType.Bool:
                return this.readByte() != 0;
            case index_js_3.VMType.Struct:
                const numFields = this.readVarInt();
                let res = {};
                for (let i = 0; i < numFields; ++i) {
                    const key = this.readVmObject();
                    const value = this.readVmObject();
                    res[key] = value;
                }
                return res;
            case index_js_3.VMType.Enum:
                return this.readVarInt();
            case index_js_3.VMType.Object:
                const numBytes = this.readVarInt();
                return this.read(numBytes);
            default:
                return 'unsupported type ' + type;
        }
    }
}
exports.PBinaryReader = PBinaryReader;
