// phantasma-sdk-ts/dist/cjs/core/types/Extensions/PBinaryWriter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBinaryWriter = void 0;
const csharp_binary_stream_1 = require("csharp-binary-stream");
const index_js_1 = require("../../utils/index.js");
const index_js_2 = require("../../interfaces/index.js");
class PBinaryWriter {
    constructor(arg1) {
        this.writer = new csharp_binary_stream_1.BinaryWriter(arg1);
    }
    get length() {
        return this.writer.length;
    }
    get position() {
        return this.writer.position;
    }
    set position(value) {
        this.writer.position = value;
    }
    writeBoolean(value) {
        this.writer.writeBoolean(value);
    }
    writeByte(value) {
        this.writer.writeByte(value);
    }
    writeSameByte(value, repeats) {
        this.writer.writeSameByte(value, repeats);
    }
    writeSignedByte(value) {
        this.writer.writeSignedByte(value);
    }
    writeShort(value) {
        this.writer.writeShort(value);
    }
    writeUnsignedShort(value) {
        this.writer.writeUnsignedShort(value);
    }
    writeInt(value) {
        this.writer.writeInt(value);
    }
    writeUnsignedInt(value) {
        this.writer.writeUnsignedInt(value);
    }
    writeLong(value) {
        this.writer.writeLong(value);
    }
    writeUnsignedLong(value) {
        this.writer.writeUnsignedLong(value);
    }
    writeFloat(value) {
        this.writer.writeFloat(value);
    }
    writeDouble(value) {
        this.writer.writeDouble(value);
    }
    writeChar(character, encoding) {
        this.writer.writeChar(character, encoding);
    }
    writeChars(characters, encoding) {
        this.writer.writeChars(characters, encoding);
    }
    clear() {
        this.writer.clear();
    }
    toArray() {
        return this.writer.toArray();
    }
    toUint8Array() {
        return this.writer.toUint8Array();
    }
    appendByte(value) {
        this.writeByte(value);
        return this;
    }
    appendBytes(bytes) {
        for (let i = 0; i < bytes.length; i++) {
            this.appendByte(bytes[i]);
        }
    }
    writeEnum(value) {
        const bytes = new Uint8Array(4);
        for (let i = 0; i < bytes.length; i++) {
            const byte = value & 0xff;
            bytes[i] = byte;
            value = (value - byte) / 256;
        }
        this.appendBytes(bytes);
        return this;
    }
    writeBytes(bytes) {
        for (let i = 0; i < bytes.length; i++)
            this.appendByte(bytes[i]);
        // writer.Write(bytes);
        return this;
    }
    writeVarInt(value) {
        if (value < 0)
            throw 'negative value invalid';
        if (value < 0xfd) {
            this.appendByte(value);
        }
        else if (value <= 0xffff) {
            let B = (value & 0x0000ff00) >> 8;
            let A = value & 0x000000ff;
            // TODO check if the endianess is correct, might have to reverse order of appends
            this.appendByte(0xfd);
            this.appendByte(A);
            this.appendByte(B);
        }
        else if (value <= 0xffffffff) {
            let C = (value & 0x00ff0000) >> 16;
            let B = (value & 0x0000ff00) >> 8;
            let A = value & 0x000000ff;
            // TODO check if the endianess is correct, might have to reverse order of appends
            this.appendByte(0xfe);
            this.appendByte(A);
            this.appendByte(B);
            this.appendByte(C);
        }
        else {
            let D = (value & 0xff000000) >> 24;
            let C = (value & 0x00ff0000) >> 16;
            let B = (value & 0x0000ff00) >> 8;
            let A = value & 0x000000ff;
            // TODO check if the endianess is correct, might have to reverse order of appends
            this.appendByte(0xff);
            this.appendByte(A);
            this.appendByte(B);
            this.appendByte(C);
            this.appendByte(D);
        }
        return this;
    }
    writeTimestamp(obj) {
        let num = obj.value;
        let a = (num & 0xff000000) >> 24;
        let b = (num & 0x00ff0000) >> 16;
        let c = (num & 0x0000ff00) >> 8;
        let d = num & 0x000000ff;
        let bytes = Uint8Array.from([d, c, b, a]);
        this.appendBytes(bytes);
        return this;
    }
    writeDateTime(obj) {
        let num = (obj.getTime() / 1000) | 0;
        let a = (num & 0xff000000) >> 24;
        let b = (num & 0x00ff0000) >> 16;
        let c = (num & 0x0000ff00) >> 8;
        let d = num & 0x000000ff;
        let bytes = Uint8Array.from([d, c, b, a]);
        this.appendBytes(bytes);
        return this;
    }
    rawString(value) {
        var data = [];
        for (var i = 0; i < value.length; i++) {
            data.push(value.charCodeAt(i));
        }
        return data;
    }
    writeByteArray(bytes) {
        if (bytes instanceof Uint8Array) {
            bytes = Array.from(bytes);
        }
        this.writeVarInt(bytes.length);
        this.writeBytes(bytes);
        return this;
    }
    writeString(text) {
        let bytes = this.rawString(text);
        this.writeVarInt(bytes.length);
        this.writeBytes(bytes);
        return this;
    }
    emitUInt32(value) {
        if (value < 0)
            throw 'negative value invalid';
        let D = (value & 0xff000000) >> 24;
        let C = (value & 0x00ff0000) >> 16;
        let B = (value & 0x0000ff00) >> 8;
        let A = value & 0x000000ff;
        // TODO check if the endianess is correct, might have to reverse order of appends
        this.appendByte(0xff);
        this.appendByte(A);
        this.appendByte(B);
        this.appendByte(C);
        this.appendByte(D);
        return this;
    }
    writeBigInteger(value) {
        return this.writeBigIntegerString(value.toString());
    }
    writeBigIntegerString(value) {
        let bytes = [];
        if (value == '0') {
            bytes = [0];
        }
        else if (value.startsWith('-1')) {
            throw new Error('Unsigned bigint serialization not suppoted');
        }
        else {
            let hex = BigInt(value).toString(16);
            if (hex.length % 2)
                hex = '0' + hex;
            const len = hex.length / 2;
            var i = 0;
            var j = 0;
            while (i < len) {
                bytes.unshift(parseInt(hex.slice(j, j + 2), 16)); // little endian
                i += 1;
                j += 2;
            }
            bytes.push(0); // add sign at the end
        }
        return this.writeByteArray(bytes);
    }
    writeSignature(signature) {
        if (!signature) {
            this.writeByte(index_js_2.SignatureKind.None);
            return this;
        }
        this.writeByte(signature.Kind);
        signature.SerializeData(this);
        return this;
    }
    AppendHexEncoded(bytesHex) {
        let bytes = (0, index_js_1.hexToBytes)(bytesHex);
        this.writeVarInt(bytes.length);
        this.appendBytes(bytes);
        return this;
    }
}
exports.PBinaryWriter = PBinaryWriter;
