// phantasma-sdk-ts/dist/esm/core/types/Carbon/Blockchain/TxMsg.js
import { Bytes32 } from '../Bytes32.js';
import { SmallString } from '../SmallString.js';
import { TxTypes } from '../TxTypes.js';
import { TxMsgCall } from './TxMsgCall.js';
import { TxMsgCallMulti } from './TxMsgCallMulti.js';
import { TxMsgMintFungible } from './TxMsgMintFungible.js';
import { TxMsgMintNonFungible } from './TxMsgMintNonFungible.js';
import { TxMsgTrade } from './TxMsgTrade.js';
import { TxMsgTransferFungible } from './TxMsgTransferFungible.js';
import { TxMsgTransferFungibleGasPayer } from './TxMsgTransferFungibleGasPayer.js';
import { TxMsgTransferNonFungibleMulti } from './TxMsgTransferNonFungibleMulti.js';
import { TxMsgTransferNonFungibleMultiGasPayer } from './TxMsgTransferNonFungibleMultiGasPayer.js';
import { TxMsgTransferNonFungibleSingle } from './TxMsgTransferNonFungibleSingle.js';
import { TxMsgTransferNonFungibleSingleGasPayer } from './TxMsgTransferNonFungibleSingleGasPayer.js';
import { TxMsgBurnFungible } from './TxMsgBurnFungible.js';
import { TxMsgBurnFungibleGasPayer } from './TxMsgBurnFungibleGasPayer.js';
import { TxMsgBurnNonFungible } from './TxMsgBurnNonFungible.js';
import { TxMsgBurnNonFungibleGasPayer } from './TxMsgBurnNonFungibleGasPayer.js';
import { TxMsgPhantasma } from './TxMsgPhantasma.js';
import { TxMsgPhantasmaRaw } from './TxMsgPhantasmaRaw.js';
export class TxMsg {
    constructor(type = TxTypes.Call, expiry = 0n, // int64
    maxGas = 0n, // uint64
    maxData = 0n, // uint64
    gasFrom = new Bytes32(), payload = new SmallString(''), msg) {
        this.type = type;
        this.expiry = expiry;
        this.maxGas = maxGas;
        this.maxData = maxData;
        this.gasFrom = gasFrom;
        this.payload = payload;
        this.msg = msg;
    }
    write(w) {
        w.write1(this.type);
        w.write8(this.expiry);
        w.write8u(this.maxGas);
        w.write8u(this.maxData);
        this.gasFrom.write(w);
        this.payload.write(w);
        switch (this.type) {
            case TxTypes.Call:
                this.msg.write(w);
                break;
            case TxTypes.Call_Multi:
                this.msg.write(w);
                break;
            case TxTypes.Trade:
                this.msg.write(w);
                break;
            case TxTypes.TransferFungible:
                this.msg.write(w);
                break;
            case TxTypes.TransferFungible_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes.TransferNonFungible_Single:
                this.msg.write(w);
                break;
            case TxTypes.TransferNonFungible_Single_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes.TransferNonFungible_Multi:
                this.msg.write(w);
                break;
            case TxTypes.TransferNonFungible_Multi_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes.MintFungible:
                this.msg.write(w);
                break;
            case TxTypes.BurnFungible:
                this.msg.write(w);
                break;
            case TxTypes.BurnFungible_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes.MintNonFungible:
                this.msg.write(w);
                break;
            case TxTypes.BurnNonFungible:
                this.msg.write(w);
                break;
            case TxTypes.BurnNonFungible_GasPayer:
                this.msg.write(w);
                break;
            case TxTypes.Phantasma:
                this.msg.write(w);
                break;
            case TxTypes.Phantasma_Raw:
                this.msg.write(w);
                break;
            default:
                throw new Error('Unsupported transaction type');
        }
    }
    read(r) {
        this.type = r.read1();
        this.expiry = r.read8();
        this.maxGas = r.read8u();
        this.maxData = r.read8u();
        this.gasFrom = Bytes32.read(r);
        this.payload = SmallString.read(r);
        switch (this.type) {
            case TxTypes.Call:
                this.msg = TxMsgCall.read(r);
                break;
            case TxTypes.Call_Multi:
                this.msg = TxMsgCallMulti.read(r);
                break;
            case TxTypes.Trade:
                this.msg = TxMsgTrade.read(r);
                break;
            case TxTypes.TransferFungible:
                this.msg = TxMsgTransferFungible.read(r);
                break;
            case TxTypes.TransferFungible_GasPayer:
                this.msg = TxMsgTransferFungibleGasPayer.read(r);
                break;
            case TxTypes.TransferNonFungible_Single:
                this.msg = TxMsgTransferNonFungibleSingle.read(r);
                break;
            case TxTypes.TransferNonFungible_Single_GasPayer:
                this.msg = TxMsgTransferNonFungibleSingleGasPayer.read(r);
                break;
            case TxTypes.TransferNonFungible_Multi:
                this.msg = TxMsgTransferNonFungibleMulti.read(r);
                break;
            case TxTypes.TransferNonFungible_Multi_GasPayer:
                this.msg = TxMsgTransferNonFungibleMultiGasPayer.read(r);
                break;
            case TxTypes.MintFungible:
                this.msg = TxMsgMintFungible.read(r);
                break;
            case TxTypes.BurnFungible:
                this.msg = TxMsgBurnFungible.read(r);
                break;
            case TxTypes.BurnFungible_GasPayer:
                this.msg = TxMsgBurnFungibleGasPayer.read(r);
                break;
            case TxTypes.MintNonFungible:
                this.msg = TxMsgMintNonFungible.read(r);
                break;
            case TxTypes.BurnNonFungible:
                this.msg = TxMsgBurnNonFungible.read(r);
                break;
            case TxTypes.BurnNonFungible_GasPayer:
                this.msg = TxMsgBurnNonFungibleGasPayer.read(r);
                break;
            case TxTypes.Phantasma:
                this.msg = TxMsgPhantasma.read(r);
                break;
            case TxTypes.Phantasma_Raw:
                this.msg = TxMsgPhantasmaRaw.read(r);
                break;
            default:
                throw new Error('Unsupported transaction type');
        }
    }
    static read(r) {
        const v = new TxMsg();
        v.read(r);
        return v;
    }
}
