// phantasma-sdk-ts/dist/cjs/core/types/Carbon/Blockchain/Vm/VmDynamicStruct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VmDynamicStruct = void 0;
const VmNamedDynamicVariable_js_1 = require("./VmNamedDynamicVariable.js");
const VmDynamicVariable_js_1 = require("./VmDynamicVariable.js");
const VmStructSchema_js_1 = require("./VmStructSchema.js");
class VmDynamicStruct {
    constructor() {
        // NOTE: fields must be sorted by name
        this.fields = [];
    }
    getValue(key) {
        const found = this.fields.find((x) => x.name.data === key.data);
        return found ? found.value : undefined;
    }
    write(w) {
        if (this.fields.length > 1) {
            this.fields.sort((a, b) => a.name.compareTo(b.name));
        }
        w.writeArrayBlob(this.fields);
    }
    read(r) {
        this.fields = r.readArrayBlob(VmNamedDynamicVariable_js_1.VmNamedDynamicVariable);
        this.fields.sort((a, b) => a.name.compareTo(b.name));
    }
    static read(r) {
        const v = new VmDynamicStruct();
        v.read(r);
        return v;
    }
    writeWithSchema(schema, w) {
        let ok = true;
        let fieldsFound = 0;
        for (const f of schema.fields) {
            const inStruct = this.fields.find((x) => x.name.data === f.name.data);
            if (inStruct) {
                inStruct.value.writeWithSchema(f.schema, w);
                fieldsFound++;
            }
            else {
                const error = VmDynamicVariable_js_1.VmDynamicVariable.fromType(f.schema.type);
                error.writeWithSchema(f.schema, w);
                ok = false;
            }
        }
        if ((schema.flags & VmStructSchema_js_1.VmStructSchema.Flags.DynamicExtras) === 0) {
            return ok;
        }
        if (fieldsFound === schema.fields.length && schema.fields.length === this.fields.length) {
            w.write4u(0);
            return ok;
        }
        const extras = [];
        for (const f of this.fields) {
            const inSchema = schema.fields.find((x) => x.name.data === f.name.data);
            if (!inSchema)
                extras.push(f);
        }
        w.writeArrayBlob(extras);
        return ok;
    }
    readWithSchema(schema, r) {
        if (schema.fields.length === 0) {
            this.fields = [];
        }
        else {
            this.fields = new Array(schema.fields.length)
                .fill(null)
                .map(() => new VmNamedDynamicVariable_js_1.VmNamedDynamicVariable());
            for (let i = 0; i < schema.fields.length; i++) {
                this.fields[i].name = schema.fields[i].name;
                this.fields[i].value = new VmDynamicVariable_js_1.VmDynamicVariable();
                this.fields[i].value.readWithSchema(schema.fields[i].schema, r);
            }
        }
        if ((schema.flags & VmStructSchema_js_1.VmStructSchema.Flags.DynamicExtras) === 0) {
            if ((schema.flags & VmStructSchema_js_1.VmStructSchema.Flags.IsSorted) === 0) {
                this.fields.sort((a, b) => a.name.compareTo(b.name));
            }
            return;
        }
        const extras = r.readArrayBlob(VmNamedDynamicVariable_js_1.VmNamedDynamicVariable);
        this.fields = this.fields.concat(extras);
        this.fields.sort((a, b) => a.name.compareTo(b.name));
    }
}
exports.VmDynamicStruct = VmDynamicStruct;
